// @videojs/html/dist/dev/ui/position-controller.js
import { isEventWithinElement } from "@videojs/core/dom";
//#region src/ui/position-controller.ts
/**
* Reactive controller that manages JS-fallback positioning for floating
* popup elements (tooltips, popovers). Tracks scroll, resize, and
* ResizeObserver events to keep the popup aligned with its trigger.
*
* Native CSS Anchor Positioning still uses this controller for dynamic
* sizing variables such as available width/height.
*/
var PositionController = class {
	#host;
	#abort = null;
	#frame = 0;
	#resizeObserver = null;
	#trigger = null;
	#boundary = null;
	constructor(host) {
		this.#host = host;
		host.addController(this);
	}
	/** Discover a trigger element linked via `commandfor` attribute. */
	findTrigger() {
		if (!this.#host.id) return null;
		return this.#host.getRootNode().querySelector(`[commandfor="${this.#host.id}"]`);
	}
	/** Start or update position tracking for the given trigger. */
	sync(trigger, boundary = null) {
		if (!trigger) return;
		if (this.#abort && this.#trigger === trigger && this.#boundary === boundary) return;
		this.cleanup();
		this.#abort = new AbortController();
		this.#trigger = trigger;
		this.#boundary = boundary;
		const { signal } = this.#abort;
		const reposition = (event) => {
			if (event && isEventWithinElement(event, this.#host)) return;
			cancelAnimationFrame(this.#frame);
			this.#frame = requestAnimationFrame(() => {
				if (signal.aborted) return;
				this.#host.requestUpdate();
			});
		};
		window.addEventListener("scroll", reposition, {
			capture: true,
			passive: true,
			signal
		});
		window.addEventListener("resize", reposition, { signal });
		if (typeof ResizeObserver === "function") {
			this.#resizeObserver = new ResizeObserver(() => {
				reposition();
			});
			this.#resizeObserver.observe(trigger);
			this.#resizeObserver.observe(this.#host);
			if (boundary) this.#resizeObserver.observe(boundary);
		}
		reposition();
	}
	/** Stop all position tracking. */
	cleanup() {
		this.#abort?.abort();
		this.#abort = null;
		this.#trigger = null;
		this.#boundary = null;
		cancelAnimationFrame(this.#frame);
		this.#frame = 0;
		this.#resizeObserver?.disconnect();
		this.#resizeObserver = null;
	}
	hostDisconnected() {
		this.cleanup();
	}
	hostDestroyed() {
		this.cleanup();
	}
};
//#endregion
export { PositionController };
