// @videojs/core/dist/default/dom/media/media-host.js
//#region src/dom/media/media-host.ts
const EMPTY_TIME_RANGES = Object.freeze({
	length: 0,
	start() {
		return 0;
	},
	end() {
		return 0;
	}
});
var HTMLMediaElementHost = class extends EventTarget {
	#target = null;
	#types = /* @__PURE__ */ new Set();
	get target() {
		return this.#target;
	}
	attach(target) {
		if (!target || this.#target === target) return;
		this.#target = target;
		for (const type of this.#types) target.addEventListener(type, this.#forwardEvent);
	}
	detach() {
		if (!this.#target) return;
		for (const type of this.#types) this.#target.removeEventListener(type, this.#forwardEvent);
		this.#target = null;
	}
	querySelectorAll(selectors) {
		return this.target?.querySelectorAll(selectors) ?? [];
	}
	querySelector(selectors) {
		return this.target?.querySelector(selectors) ?? null;
	}
	addEventListener(type, listener, options) {
		if (!this.#types.has(type)) {
			this.#types.add(type);
			this.target?.addEventListener(type, this.#forwardEvent);
		}
		super.addEventListener(type, listener, options);
	}
	removeEventListener(type, listener, options) {
		super.removeEventListener(type, listener, options);
	}
	#forwardEvent = (event) => {
		this.dispatchEvent(new event.constructor(event.type, event));
	};
	get title() {
		return this.target?.title ?? "";
	}
	set title(value) {
		if (this.target) this.target.title = value;
	}
	get paused() {
		return this.target?.paused ?? true;
	}
	get ended() {
		return this.target?.ended ?? false;
	}
	get loop() {
		return this.target?.loop ?? false;
	}
	set loop(value) {
		if (this.target) this.target.loop = value;
	}
	play() {
		return this.target?.play() ?? Promise.reject();
	}
	pause() {
		this.target?.pause();
	}
	get currentTime() {
		return this.target?.currentTime ?? 0;
	}
	set currentTime(value) {
		if (this.target) this.target.currentTime = value;
	}
	get duration() {
		return this.target?.duration ?? NaN;
	}
	get seeking() {
		return this.target?.seeking ?? false;
	}
	get src() {
		return this.target?.src ?? "";
	}
	set src(value) {
		if (this.target) this.target.src = value;
	}
	get currentSrc() {
		return this.target?.currentSrc ?? "";
	}
	get readyState() {
		return this.target?.readyState ?? 0;
	}
	load() {
		this.target?.load();
	}
	get volume() {
		return this.target?.volume ?? 1;
	}
	set volume(value) {
		if (this.target) this.target.volume = value;
	}
	get muted() {
		return this.target?.muted ?? false;
	}
	set muted(value) {
		if (this.target) this.target.muted = value;
	}
	get playbackRate() {
		return this.target?.playbackRate ?? 1;
	}
	set playbackRate(value) {
		if (this.target) this.target.playbackRate = value;
	}
	get buffered() {
		return this.target?.buffered ?? EMPTY_TIME_RANGES;
	}
	get seekable() {
		return this.target?.seekable ?? EMPTY_TIME_RANGES;
	}
	get error() {
		return this.target?.error ?? null;
	}
	get textTracks() {
		return this.target?.textTracks ?? [];
	}
	get remote() {
		return this.target?.remote;
	}
	get disableRemotePlayback() {
		return this.target?.disableRemotePlayback ?? false;
	}
	set disableRemotePlayback(value) {
		if (this.target) this.target.disableRemotePlayback = value;
	}
};
//#endregion
export { HTMLMediaElementHost };
