// @videojs/spf/dist/dev/playback/behaviors/select-tracks.js
import { computed } from "../../core/signals/primitives.js";
import { defineBehavior } from "../../core/composition/create-composition.js";
import { createMachineReactor } from "../../core/reactors/create-machine-reactor.js";
import { isResolvedPresentation } from "../../media/types/index.js";
import { AUDIO_TYPE_CONFIG, TEXT_TYPE_CONFIG, VIDEO_TYPE_CONFIG } from "./track-types.js";
import { pickFirstTrackId, pickTextTrack } from "../../media/primitives/select-tracks.js";
//#region src/playback/behaviors/select-tracks.ts
/**
* **Default audio/text track selection on src load / unselect on src unload.**
* When a presentation is resolved, sets `selectedAudioTrackId` /
* `selectedTextTrackId` to a per-type-picker default if no selection already
* exists. When the presentation is unset/reset (transitions back to
* unresolved), clears the selection so a stale id from the previous source
* doesn't persist.
*
* Lifecycle-driven: each transition fires its work once. Does not police
* the selection between transitions; external writes (user picks) are left
* alone.
*
* Picker is config-driven: each per-type export wires a sensible default
* (`pickFirstTrackId` for audio, `pickTextTrack` for text) and the caller
* can supply their own via `config.picker` for custom selection logic
* (language preferences, default-track handling, etc.). The behavior's
* `config` is forwarded to the picker as its second argument, so options
* like `preferredAudioLanguage` / `preferredSubtitleLanguage` reach the
* picker without an intermediate wrapping layer.
*
* Compose `selectVideoTrack` for the simple "pick a default video track"
* behavior, or `switchVideoQuality` (`./quality-switching.ts`) for the
* ABR-driven variant — they're alternatives, not stackable (both write
* `selectedVideoTrackId`). The simple variant tree-shakes out the
* ABR machinery (bandwidth-estimator, quality-selection algorithms).
*/
function setupTrackSelection({ state, config: { selectedKey, picker, pickerConfig } }) {
	const derivedStateSignal = computed(() => isResolvedPresentation(state.presentation.get()) ? "presentation-resolved" : "presentation-unresolved");
	return createMachineReactor({
		initial: "presentation-unresolved",
		monitor: () => derivedStateSignal.get(),
		states: {
			"presentation-unresolved": {},
			"presentation-resolved": { entry: () => {
				if (!state[selectedKey].get()) {
					const id = picker(state.presentation.get(), pickerConfig);
					if (id) state[selectedKey].set(id);
				}
				return () => state[selectedKey].set(void 0);
			} }
		}
	});
}
/** Default video picker: first track in the video selection set. */
const defaultVideoPicker = (presentation) => pickFirstTrackId(presentation, "video");
const VIDEO_TRACK_SELECTION_CONFIG = {
	...VIDEO_TYPE_CONFIG,
	picker: defaultVideoPicker
};
/** Default audio picker: first track in the audio selection set. */
const defaultAudioPicker = (presentation) => pickFirstTrackId(presentation, "audio");
const AUDIO_TRACK_SELECTION_CONFIG = {
	...AUDIO_TYPE_CONFIG,
	picker: defaultAudioPicker
};
const TEXT_TRACK_SELECTION_CONFIG = {
	...TEXT_TYPE_CONFIG,
	picker: pickTextTrack
};
defineBehavior({
	stateKeys: ["presentation", "selectedVideoTrackId"],
	contextKeys: [],
	setup: ({ state, config }) => setupTrackSelection({
		state,
		config: {
			...VIDEO_TRACK_SELECTION_CONFIG,
			picker: config?.picker ?? VIDEO_TRACK_SELECTION_CONFIG.picker,
			pickerConfig: config
		}
	})
});
/**
* Select an audio track when a presentation loads. Clears the selection
* on src unload.
*
* @example
* const reactor = selectAudioTrack.setup({ state });
*
* @example
* // Custom picker with language preference
* const reactor = selectAudioTrack.setup({
*   state,
*   config: { preferredAudioLanguage: 'en', picker: myLanguageAwarePicker },
* });
*/
const selectAudioTrack = defineBehavior({
	stateKeys: ["presentation", "selectedAudioTrackId"],
	contextKeys: [],
	setup: ({ state, config }) => setupTrackSelection({
		state,
		config: {
			...AUDIO_TRACK_SELECTION_CONFIG,
			picker: config?.picker ?? AUDIO_TRACK_SELECTION_CONFIG.picker,
			pickerConfig: config
		}
	})
});
/**
* Select a text track based on user preferences (preferred language,
* default-track auto-select, forced-track filtering). Clears the selection
* on src unload.
*
* Unlike audio selection, the default text picker returns `undefined` when
* no preference matches, leaving the selection unset — text-track
* selection is user opt-in.
*
* @example
* const reactor = selectTextTrack.setup({ state, config: { preferredSubtitleLanguage: 'en' } });
*/
const selectTextTrack = defineBehavior({
	stateKeys: ["presentation", "selectedTextTrackId"],
	contextKeys: [],
	setup: ({ state, config }) => setupTrackSelection({
		state,
		config: {
			...TEXT_TRACK_SELECTION_CONFIG,
			picker: config?.picker ?? TEXT_TRACK_SELECTION_CONFIG.picker,
			pickerConfig: config
		}
	})
});
//#endregion
export { selectAudioTrack, selectTextTrack };
