// @videojs/spf/dist/default/playback/behaviors/dom/setup-mediasource.js
import { computed } from "../../../core/signals/primitives.js";
import { defineBehavior } from "../../../core/composition/create-composition.js";
import { createMachineReactor } from "../../../core/reactors/create-machine-reactor.js";
import { isResolvedPresentation } from "../../../media/types/index.js";
import { attachMediaSource, createMediaSource, waitForMediaSourceOpen } from "../../../media/dom/mse/mediasource-setup.js";
//#region src/playback/behaviors/dom/setup-mediasource.ts
/**
* **Own the MediaSource lifecycle for the current source.** When a resolved
* presentation and a mediaElement are both in scope, creates a MediaSource,
* attaches it to the element, waits for `'open'`, and publishes it on
* `context.mediaSource`. On source change or behavior destroy, detaches the
* MediaSource and clears the slot so the next source starts fresh.
*
* Single-positive-state reactor (`'preconditions-unmet'` Ōåö `'mediasource-attached'`):
* state derivation gates on `mediaElement + isResolvedPresentation`. Riding the
* resolver's resolved/unresolved lifecycle makes direct URL replacement
* structural ŌĆö `resolvePresentation` routes the presentation back through
* unresolved on URL change, which drives this reactor through
* `'preconditions-unmet'` so the entry's state-exit cleanup detaches the old
* MediaSource before the new one is built.
*
* The entry resolves preconditions in sequence before publishing:
*
* 1. **Create + attach** ŌĆö `createMediaSource` + `attachMediaSource` run
*    synchronously on entry. The `detach` closure returned by
*    `attachMediaSource` is captured for state-exit cleanup, so the cleanup
*    is always bound to its setup even if the wait below is aborted.
* 2. **Wait for `'open'`** ŌĆö `waitForMediaSourceOpen` defers until the first
*    `sourceopen` event (or any readyState transition out of `'closed'`).
* 3. **Publish on `'open'`** ŌĆö re-check `readyState === 'open'` after the
*    await (covers `'ended'` / `'closed'` race) before writing to
*    `context.mediaSource`. Downstream `setupVideoBufferActors` /
*    `setupAudioBufferActors` call `addSourceBuffer` directly, which
*    throws on non-open, so publish-only-when-open is the load-bearing
*    contract.
*
* State-exit cleanup aborts the in-flight wait, detaches the MediaSource,
* and clears `context.mediaSource`. Order: abort first (prevents a late
* publish racing the slot clear), then detach, then clear.
*
* Sole writer of `context.mediaSource`; other MSE behaviors
* (`setupVideoBufferActors`, `setupAudioBufferActors`,
* `updateMediaSourceDuration`, `endOfStream`, `loadVideoSegments`) only
* read.
*/
function deriveState(presentation, mediaElement) {
	if (!mediaElement || !isResolvedPresentation(presentation)) return "preconditions-unmet";
	return "mediasource-attached";
}
function setupMediaSourceSetup({ state, context }) {
	const derivedStateSignal = computed(() => deriveState(state.presentation.get(), context.mediaElement.get()));
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			"mediasource-attached": { entry: () => {
				const mediaElement = context.mediaElement.get();
				const controller = new AbortController();
				const mediaSource = createMediaSource({ preferManaged: true });
				const { detach } = attachMediaSource(mediaSource, mediaElement);
				const publishWhenOpen = async () => {
					await waitForMediaSourceOpen(mediaSource, controller.signal);
					if (controller.signal.aborted) return;
					if (mediaSource.readyState !== "open") {
						console.warn(`[setupMediaSource] MediaSource transitioned to '${mediaSource.readyState}' before first 'sourceopen' ŌĆö slot left unpublished; recoverable on next source reset.`);
						return;
					}
					context.mediaSource.set(mediaSource);
				};
				publishWhenOpen().catch((err) => console.error("[setupMediaSource] failed to publish MediaSource:", err));
				return () => {
					controller.abort();
					detach();
					context.mediaSource.set(void 0);
				};
			} }
		}
	});
}
const setupMediaSource = defineBehavior({
	stateKeys: ["presentation"],
	contextKeys: ["mediaElement", "mediaSource"],
	setup: setupMediaSourceSetup
});
//#endregion
export { setupMediaSource };
