// @videojs/core/dist/default/dom/media/predicate.js
import { isFunction, isObject } from "@videojs/utils/predicate";
//#region src/dom/media/predicate.ts
function hasMetadata(media) {
	return media.readyState >= 1;
}
function isMediaPauseCapable(value) {
	return isObject(value) && "paused" in value && "ended" in value && isFunction(value.pause);
}
function isMediaSeekCapable(value) {
	return isObject(value) && "currentTime" in value && "duration" in value && "seeking" in value;
}
function isMediaSourceCapable(value) {
	return isObject(value) && "src" in value && "currentSrc" in value && "readyState" in value && isFunction(value.load);
}
function isMediaVolumeCapable(value) {
	return isObject(value) && "volume" in value && "muted" in value;
}
function isMediaPlaybackRateCapable(value) {
	return isObject(value) && "playbackRate" in value;
}
function isMediaBufferCapable(value) {
	return isObject(value) && "buffered" in value && "seekable" in value;
}
function isMediaErrorCapable(value) {
	return isObject(value) && "error" in value;
}
function isMediaTextTrackCapable(value) {
	return isObject(value) && "textTracks" in value;
}
function isMediaRemotePlaybackCapable(value) {
	return isObject(value) && "remote" in value && isObject(value.remote);
}
function isMediaStreamTypeCapable(value) {
	return isObject(value) && "streamType" in value;
}
function isMediaLiveCapable(value) {
	return isObject(value) && "liveEdgeStart" in value && "targetLiveWindow" in value;
}
function isQuerySelectorAllCapable(value) {
	return isObject(value) && "querySelectorAll" in value && isFunction(value.querySelectorAll);
}
//#endregion
export { hasMetadata, isMediaBufferCapable, isMediaErrorCapable, isMediaLiveCapable, isMediaPauseCapable, isMediaPlaybackRateCapable, isMediaRemotePlaybackCapable, isMediaSeekCapable, isMediaSourceCapable, isMediaStreamTypeCapable, isMediaTextTrackCapable, isMediaVolumeCapable, isQuerySelectorAllCapable };
