// @videojs/html/dist/default/ui/playback-rate-menu/playback-rate-menu-element.js
import { playerContext } from "../../player/context.js";
import { PlayerController } from "../../player/player-controller.js";
import { MenuElement } from "../menu/menu-element.js";
import { applyElementProps, applyStateDataAttrs, selectPlaybackRate } from "@videojs/core/dom";
import { PlaybackRateMenuCore, PlaybackRateMenuDataAttrs } from "@videojs/core";
//#region src/ui/playback-rate-menu/playback-rate-menu-element.ts
var PlaybackRateMenuElement = class extends MenuElement {
	constructor(..._args) {
		super(..._args);
		this.label = "";
		this.disabled = false;
		this.align = "center";
		this.formatRate = PlaybackRateMenuCore.defaultProps.formatRate;
		this.#core = new PlaybackRateMenuCore();
		this.#mediaState = new PlayerController(this, playerContext, selectPlaybackRate);
	}
	static {
		this.tagName = "media-playback-rate-menu";
	}
	static {
		this.properties = {
			...MenuElement.properties,
			label: { type: String },
			disabled: { type: Boolean }
		};
	}
	#core;
	#mediaState;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
	}
	update(changed) {
		super.update(changed);
		const media = this.#mediaState.value;
		if (!media) return;
		this.#core.setProps(this);
		this.#core.setMedia(media);
		const state = this.#core.getState();
		applyElementProps(this, this.#core.getAttrs(state));
		applyStateDataAttrs(this, state, PlaybackRateMenuDataAttrs);
	}
};
//#endregion
export { PlaybackRateMenuElement };
