// @videojs/html/dist/default/ui/playback-rate-menu/playback-rate-options-element.js
import { playerContext } from "../../player/context.js";
import { PlayerController } from "../../player/player-controller.js";
import { MenuItemIndicatorElement } from "../menu/menu-item-indicator-element.js";
import { MenuRadioGroupElement } from "../menu/menu-radio-group-element.js";
import { MenuRadioItemElement } from "../menu/menu-radio-item-element.js";
import { applyStateDataAttrs, selectPlaybackRate } from "@videojs/core/dom";
import { PlaybackRateMenuCore, PlaybackRateMenuDataAttrs } from "@videojs/core";
//#region src/ui/playback-rate-menu/playback-rate-options-element.ts
var PlaybackRateOptionsElement = class extends MenuRadioGroupElement {
	constructor(..._args) {
		super(..._args);
		this.disabled = false;
		this.formatRate = PlaybackRateMenuCore.defaultProps.formatRate;
		this.#core = new PlaybackRateMenuCore();
		this.#mediaState = new PlayerController(this, playerContext, selectPlaybackRate);
		this.#ratesKey = "";
		this.#disconnect = null;
		this.#handleValueChange = (event) => {
			if (event.target !== this) return;
			const media = this.#mediaState.value;
			if (!media) return;
			const { value } = event.detail;
			this.#core.selectValue(media, value);
		};
	}
	static {
		this.tagName = "media-playback-rate-options";
	}
	static {
		this.properties = {
			...MenuRadioGroupElement.properties,
			disabled: { type: Boolean }
		};
	}
	#core;
	#mediaState;
	#ratesKey;
	#disconnect;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		this.#disconnect = new AbortController();
		this.addEventListener("value-change", this.#handleValueChange, { signal: this.#disconnect.signal });
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#disconnect?.abort();
		this.#disconnect = null;
	}
	update(changed) {
		const media = this.#mediaState.value;
		let state = null;
		if (media) {
			this.#core.setProps({
				formatRate: this.formatRate,
				disabled: this.disabled
			});
			this.#core.setMedia(media);
			state = this.#core.getState();
			this.value = this.#core.getRateValue(state.rate);
			this.label = this.label || "Playback rate";
			this.#syncContent(state);
		}
		super.update(changed);
		if (state) applyStateDataAttrs(this, state, PlaybackRateMenuDataAttrs);
	}
	#syncContent(state) {
		const template = this.#getTemplate();
		const templateKey = template?.innerHTML ?? "";
		const ratesKey = `${state.rates.join("|")}::${templateKey}`;
		if (ratesKey !== this.#ratesKey) {
			this.#ratesKey = ratesKey;
			for (const child of [...this.children]) {
				if (child instanceof HTMLTemplateElement) continue;
				child.remove();
			}
			this.append(...state.rates.map((rate) => this.#createItem(rate, template)));
		}
		for (const item of this.querySelectorAll(MenuRadioItemElement.tagName)) {
			const checked = item.value === this.value;
			item.disabled = state.disabled;
			for (const indicator of item.querySelectorAll(MenuItemIndicatorElement.tagName)) indicator.checked = checked;
		}
	}
	#createItem(rate, template) {
		const item = this.#createItemFromTemplate(template);
		const value = this.#core.getRateValue(rate);
		item.value = value;
		item.setAttribute("data-rate", value);
		this.#setLabel(item, this.#core.getRateLabel(rate));
		return item;
	}
	#createItemFromTemplate(template) {
		if (!template) return document.createElement(MenuRadioItemElement.tagName);
		const root = template.content.cloneNode(true).firstElementChild;
		if (!root || root.localName !== MenuRadioItemElement.tagName || root.nextElementSibling) return document.createElement(MenuRadioItemElement.tagName);
		return root;
	}
	#setLabel(item, label) {
		const labelPart = item.querySelector("[data-part~=\"label\"]");
		if (labelPart) labelPart.textContent = label;
		else item.textContent = label;
	}
	#getTemplate() {
		for (const child of this.children) if (child instanceof HTMLTemplateElement) return child;
		return null;
	}
	#handleValueChange;
};
//#endregion
export { PlaybackRateOptionsElement };
