// @videojs/core/dist/dev/dom/ui/input-action.js
import { IndicatorVisibilityCoordinator } from "../../core/ui/input-feedback/indicator-lifecycle.js";
import { getGestureCoordinator } from "../gesture/coordinator.js";
import { selectFullscreen, selectPiP, selectPlayback, selectTextTrack, selectTime, selectVolume } from "../store/selectors.js";
import { getHotkeyCoordinator } from "../hotkey/hotkey.js";
//#region src/dom/ui/input-action.ts
function toInputActionEvent(event) {
	return {
		action: event.action,
		value: event.value,
		source: event.source,
		key: "key" in event.event ? event.event.key : void 0
	};
}
function getMediaSnapshot(store) {
	if (!store) return {};
	const state = store.state;
	const time = selectTime(state);
	return {
		paused: selectPlayback(state)?.paused,
		volume: selectVolume(state)?.volume,
		muted: selectVolume(state)?.muted,
		fullscreen: selectFullscreen(state)?.fullscreen,
		subtitlesShowing: selectTextTrack(state)?.subtitlesShowing,
		pip: selectPiP(state)?.pip,
		currentTime: time?.currentTime,
		duration: time?.duration
	};
}
function subscribeToInputActions(container, callback) {
	const handleEvent = (event) => callback(toInputActionEvent(event));
	const gestureUnsubscribe = getGestureCoordinator(container).subscribe(handleEvent);
	const hotkeyUnsubscribe = getHotkeyCoordinator(container).subscribe(handleEvent);
	return () => {
		gestureUnsubscribe();
		hotkeyUnsubscribe();
	};
}
const indicatorVisibilityCoordinators = /* @__PURE__ */ new WeakMap();
function getIndicatorVisibilityCoordinator(container) {
	let coordinator = indicatorVisibilityCoordinators.get(container);
	if (!coordinator) {
		coordinator = new IndicatorVisibilityCoordinator();
		indicatorVisibilityCoordinators.set(container, coordinator);
	}
	return coordinator;
}
//#endregion
export { getIndicatorVisibilityCoordinator, getMediaSnapshot, subscribeToInputActions, toInputActionEvent };
