// @videojs/core/dist/dev/dom/media/video-host.js
import { HTMLMediaElementHost } from "./media-host.js";
import { isFunction } from "@videojs/utils/predicate";
//#region src/dom/media/video-host.ts
var HTMLVideoElementHost = class extends HTMLMediaElementHost {
	get poster() {
		return this.target?.poster ?? "";
	}
	set poster(value) {
		if (this.target) this.target.poster = value;
	}
	get webkitPresentationMode() {
		return this.target?.webkitPresentationMode;
	}
	get webkitSetPresentationMode() {
		const target = this.target;
		const fn = target?.webkitSetPresentationMode;
		return isFunction(fn) ? fn.bind(target) : void 0;
	}
	get isPictureInPicture() {
		return !!this.target && globalThis.document?.pictureInPictureElement === this.target || this.webkitPresentationMode === "picture-in-picture";
	}
	get isFullscreen() {
		if (!this.target) return false;
		if (this.webkitPresentationMode === "fullscreen") return true;
		const doc = globalThis.document;
		return doc?.fullscreenElement === this.target || doc?.webkitFullscreenElement === this.target;
	}
	async requestPictureInPicture() {
		if (!this.target) return Promise.reject();
		return this.target.requestPictureInPicture();
	}
	async exitPictureInPicture() {
		if (!this.target) return Promise.reject();
		return globalThis.document?.exitPictureInPicture();
	}
	requestFullscreen() {
		if (!this.target) return Promise.reject();
		return this.target.requestFullscreen();
	}
	exitFullscreen() {
		if (!this.target) return Promise.reject();
		return globalThis.document?.exitFullscreen();
	}
};
//#endregion
export { HTMLVideoElementHost };
