// @videojs/core/dist/dev/core/ui/pip-button/pip-button-core.js
import { createState } from "@videojs/store";
import { defaults } from "@videojs/utils/object";
import { isFunction } from "@videojs/utils/predicate";
//#region src/core/ui/pip-button/pip-button-core.ts
var PiPButtonCore = class PiPButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	state = createState({
		pip: false,
		availability: "available",
		label: ""
	});
	#props = { ...PiPButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, PiPButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return state.pip ? "Exit picture-in-picture" : "Enter picture-in-picture";
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		this.state.patch({
			pip: media.pip,
			availability: media.pipAvailability
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	async toggle(media) {
		if (this.#props.disabled) return;
		if (media.pipAvailability !== "available") return;
		try {
			if (media.pip) await media.exitPictureInPicture();
			else await media.requestPictureInPicture();
		} catch {}
	}
};
//#endregion
export { PiPButtonCore };
