// @videojs/core/dist/default/core/ui/menu/menu-core.js
import { getTransitionFlags } from "../transition.js";
import { defaults } from "@videojs/utils/object";
//#region src/core/ui/menu/menu-core.ts
/** Base menu logic: ARIA attributes and open/close state computation. */
var MenuCore = class MenuCore {
	static defaultProps = {
		side: "bottom",
		align: "start",
		open: false,
		defaultOpen: false,
		closeOnEscape: true,
		closeOnOutsideClick: true,
		isSubmenu: false
	};
	#props = { ...MenuCore.defaultProps };
	#input = null;
	get props() {
		return this.#props;
	}
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, MenuCore.defaultProps);
	}
	setInput(input) {
		this.#input = input;
	}
	getState() {
		const input = this.#input;
		const isSubmenu = this.#props.isSubmenu;
		return {
			open: input.active,
			status: input.status,
			side: isSubmenu ? void 0 : this.#props.side,
			align: isSubmenu ? void 0 : this.#props.align,
			isSubmenu,
			...getTransitionFlags(input.status)
		};
	}
	getTriggerAttrs(state, contentId) {
		return {
			"aria-haspopup": "menu",
			"aria-expanded": state.open ? "true" : "false",
			"aria-controls": contentId
		};
	}
	getContentAttrs(state) {
		return {
			role: "menu",
			tabIndex: -1,
			...!state.isSubmenu && { popover: "manual" }
		};
	}
};
//#endregion
export { MenuCore };
