// @videojs/spf/dist/dev/playback/behaviors/quality-switching.js
import { computed, peek } from "../../core/signals/primitives.js";
import { defineBehavior } from "../../core/composition/create-composition.js";
import { createMachineReactor } from "../../core/reactors/create-machine-reactor.js";
import { isResolvedPresentation } from "../../media/types/index.js";
import { getTracksByType } from "../../media/utils/tracks.js";
import { DEFAULT_BANDWIDTH_CONFIG, getBandwidthEstimate } from "../../network/bandwidth-estimator.js";
import { DEFAULT_QUALITY_CONFIG, selectLowestQuality, selectQuality } from "../../media/abr/quality-selection.js";
function setupQualitySwitching({ state, config }) {
	const safetyMargin = config.quality?.safetyMargin ?? DEFAULT_QUALITY_CONFIG.safetyMargin;
	const upgradeMargin = config.quality?.upgradeMargin ?? DEFAULT_QUALITY_CONFIG.upgradeMargin;
	const initialBandwidth = config.initialBandwidth ?? 5e6;
	const bandwidthConfig = {
		...DEFAULT_BANDWIDTH_CONFIG,
		...config.bandwidth
	};
	const { selectionKey, userSelectionKey, getTracks, selectOptimal } = config;
	const derivedStateSignal = computed(() => isResolvedPresentation(state.presentation.get()) ? "presentation-resolved" : "presentation-unresolved");
	return createMachineReactor({
		initial: "presentation-unresolved",
		monitor: () => derivedStateSignal.get(),
		states: {
			"presentation-unresolved": {},
			"presentation-resolved": {
				entry: () => () => state[selectionKey].set(void 0),
				effects: [() => {
					const presentation = peek(state.presentation);
					if (!presentation) return;
					const allTracks = getTracks(presentation);
					const [firstAllTrack] = allTracks;
					if (!firstAllTrack) return;
					const userFilter = state[userSelectionKey].get();
					const matching = userFilter ? allTracks.filter((track) => {
						for (const key in userFilter) {
							const filterValue = userFilter[key];
							if (filterValue !== void 0 && track[key] !== filterValue) return false;
						}
						return true;
					}) : allTracks;
					const candidates = matching.length > 0 ? matching : allTracks;
					if (!candidates.length) return;
					const selectedId = state[selectionKey].get();
					if (candidates.length === 1) {
						if (candidates[0].id !== selectedId) state[selectionKey].set(candidates[0].id);
						return;
					}
					const bandwidth = getBandwidthEstimate(state.bandwidthState.get(), initialBandwidth, bandwidthConfig);
					if (!selectedId && config.picker) {
						const id = config.picker(presentation, config);
						if (id) {
							state[selectionKey].set(id);
							return;
						}
					}
					const optimal = selectOptimal(candidates, bandwidth, {
						safetyMargin,
						upgradeMargin,
						currentTrack: candidates.find((t) => t.id === selectedId)
					}) ?? selectLowestQuality(candidates);
					if (optimal.id !== selectedId) state[selectionKey].set(optimal.id);
				}]
			}
		}
	});
}
/**
* Manage `selectedVideoTrackId`: pick a default on src load, dynamically
* adjust based on bandwidth, clear on src unload. Honors
* `userVideoTrackSelection` as a partial-track constraint on candidates;
* short-circuits ABR when the constraint narrows to a single track.
*
* @example
* const reactor = switchVideoQuality.setup({ state });
*/
const switchVideoQuality = defineBehavior({
	stateKeys: [
		"presentation",
		"bandwidthState",
		"selectedVideoTrackId",
		"userVideoTrackSelection"
	],
	contextKeys: [],
	setup: ({ state, config }) => setupQualitySwitching({
		state,
		config: {
			...config,
			selectionKey: "selectedVideoTrackId",
			userSelectionKey: "userVideoTrackSelection",
			getTracks: (presentation) => getTracksByType(presentation, "video"),
			selectOptimal: selectQuality
		}
	})
});
//#endregion
export { switchVideoQuality };
