// @videojs/html/dist/dev/ui/status-announcer/status-announcer-element.js
import { containerContext, playerContext } from "../../player/context.js";
import { MediaElement } from "../media-element.js";
import { PlayerController } from "../../player/player-controller.js";
import { getMediaSnapshot, subscribeToInputActions } from "@videojs/core/dom";
import { ContextConsumer } from "@videojs/element/context";
import { StatusAnnouncerCore } from "@videojs/core";
//#region src/ui/status-announcer/status-announcer-element.ts
var StatusAnnouncerElement = class extends MediaElement {
	static {
		this.tagName = "media-status-announcer";
	}
	static {
		this.properties = { closeDelay: {
			type: Number,
			attribute: "close-delay"
		} };
	}
	#core = new StatusAnnouncerCore();
	#player = new PlayerController(this, playerContext);
	#container = new ContextConsumer(this, {
		context: containerContext,
		callback: () => this.#reconnect(),
		subscribe: true
	});
	#disconnect = null;
	#inputActionUnsubscribe = null;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		this.setAttribute("role", "status");
		this.#disconnect = new AbortController();
		this.#core.state.subscribe(() => this.requestUpdate(), { signal: this.#disconnect.signal });
		this.#reconnect();
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#inputActionUnsubscribe?.();
		this.#inputActionUnsubscribe = null;
		this.#disconnect?.abort();
		this.#disconnect = null;
	}
	destroyCallback() {
		this.#inputActionUnsubscribe?.();
		this.#core.destroy();
		super.destroyCallback();
	}
	willUpdate(changed) {
		super.willUpdate(changed);
		this.#core.setProps({ closeDelay: this.closeDelay });
	}
	update(changed) {
		super.update(changed);
		const label = this.#core.state.current.label;
		if (label) this.setAttribute("aria-label", label);
		else this.removeAttribute("aria-label");
	}
	#reconnect() {
		this.#inputActionUnsubscribe?.();
		this.#inputActionUnsubscribe = null;
		const container = this.#container.value?.container;
		if (!container) return;
		this.#inputActionUnsubscribe = subscribeToInputActions(container, (event) => {
			this.#core.processEvent(event, getMediaSnapshot(this.#player.value));
		});
	}
};
//#endregion
export { StatusAnnouncerElement };
