// @videojs/core/dist/dev/dom/store/features/feature.parts.js
import { __exportAll } from "../../../_virtual/_rolldown/runtime.js";
import { bufferFeature } from "./buffer.js";
import { controlsFeature } from "./controls.js";
import { fullscreenFeature } from "./fullscreen.js";
import { liveFeature } from "./live.js";
import { pipFeature } from "./pip.js";
import { playbackFeature } from "./playback.js";
import { playbackRateFeature } from "./playback-rate.js";
import { remotePlaybackFeature } from "./remote-playback.js";
import { sourceFeature } from "./source.js";
import { streamTypeFeature } from "./stream-type.js";
import { textTrackFeature } from "./text-track.js";
import { timeFeature } from "./time.js";
import { volumeFeature } from "./volume.js";
import { audioFeatures, backgroundFeatures, videoFeatures } from "./presets.js";
//#region src/dom/store/features/feature.parts.ts
var feature_parts_exports = /* @__PURE__ */ __exportAll({
	audioFeatures: () => audioFeatures,
	backgroundFeatures: () => backgroundFeatures,
	buffer: () => bufferFeature,
	controls: () => controlsFeature,
	fullscreen: () => fullscreenFeature,
	live: () => liveFeature,
	pip: () => pipFeature,
	playback: () => playbackFeature,
	playbackRate: () => playbackRateFeature,
	remotePlayback: () => remotePlaybackFeature,
	source: () => sourceFeature,
	streamType: () => streamTypeFeature,
	textTrack: () => textTrackFeature,
	time: () => timeFeature,
	videoFeatures: () => videoFeatures,
	volume: () => volumeFeature
});
//#endregion
export { feature_parts_exports };
