// @videojs/spf/dist/dev/playback/behaviors/dom/sync-text-tracks.js
import { computed, peek } from "../../../core/signals/primitives.js";
import { defineBehavior } from "../../../core/composition/create-composition.js";
import { createMachineReactor } from "../../../core/reactors/create-machine-reactor.js";
import { getTracksByType } from "../../../media/utils/tracks.js";
import { syncTextTrackModes } from "../../../media/dom/text/text-track-slots.js";
import { listen } from "@videojs/utils/dom";
//#region src/playback/behaviors/dom/sync-text-tracks.ts
/**
* **Own the text-track slots on the host media element, mirroring the SPF
* model.** When a presentation is resolved and a media element is
* available, allocate one slot in `mediaElement.textTracks` per model text
* track — via creating `<track>` children, since that's the only spec
* mechanism for adding *and* removing entries to `textTracks` (no
* `removeTextTrack` API exists). Once slots are provisioned, mirror
* `selectedTextTrackId` into their `mode`s, and propagate user-initiated
* DOM `change` events back to `selectedTextTrackId` so non-SPF consumers
* (host-page captions buttons, browser native UI, video.js store) remain
* the canonical write path for selection.
*
* Single-positive-state reactor (`'preconditions-unmet'` ↔ `'sync-active'`):
* the entry allocates the slots, applies the initial selection, attaches
* the `change` listener, and opens a brief Chromium settling-window guard —
* all transition-driven, fire-once on state entry, with paired cleanup on
* state exit. A single `effects:` mirrors subsequent
* `selectedTextTrackId` changes into `mode`s; that's the only
* continuous-reactivity concern.
*
* State-exit cleanup also sends a `'clear'` message to the
* `TextTracksActor` so its cue+segment cache (keyed by trackId) is
* dropped alongside the DOM `<track>` slots. The actor itself is owned
* by `setupTextTrackActors` and bound to mediaElement, not presentation,
* so it survives source resets; clearing its context here keeps the
* cache consistent with the DOM. Without this, a subsequent
* presentation reusing a trackId would have `getSegmentsToLoad` treat
* its segments as already-buffered and skip loading them.
*
* Multi-writer with `selectTextTrack` is intentional and orthogonal:
* `selectTextTrack` owns the *default-on-load / clear-on-unload* contract
* for `selectedTextTrackId`; this behavior writes only from DOM `change`
* events. They reflect distinct decision-making domains (config-driven
* intent vs DOM-driven user action), so this behavior never clears the
* slot on source unload — that's `selectTextTrack`'s contract. (It *does*
* write `undefined` when a user disables all tracks via native UI, since
* that's a real user action that should round-trip into SPF state.)
*/
function deriveState(presentation, mediaElement) {
	if (!mediaElement || !presentation) return "preconditions-unmet";
	return getTracksByType(presentation, "text").length > 0 ? "sync-active" : "preconditions-unmet";
}
function syncTextTracksSetup({ state, context, config }) {
	const { addSubtitlesTracksToMedia, getShowingSubtitlesTrackFromMedia, removeAllSubtitlesTracksFromMedia } = config;
	const derivedStateSignal = computed(() => deriveState(state.presentation.get(), context.mediaElement.get()));
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			"sync-active": {
				entry: () => {
					const mediaElement = context.mediaElement.get();
					addSubtitlesTracksToMedia(mediaElement, getTracksByType(state.presentation.get(), "text"));
					syncTextTrackModes(mediaElement.textTracks, state.selectedTextTrackId.get());
					let inSettlingWindow = true;
					const settlingTimeout = setTimeout(() => {
						inSettlingWindow = false;
					}, 0);
					const onChange = () => {
						if (inSettlingWindow) {
							syncTextTrackModes(mediaElement.textTracks, state.selectedTextTrackId.get());
							return;
						}
						const nextId = getShowingSubtitlesTrackFromMedia(mediaElement)?.id || void 0;
						if (nextId === state.selectedTextTrackId.get()) return;
						state.selectedTextTrackId.set(nextId);
					};
					const unlisten = listen(mediaElement.textTracks, "change", onChange);
					return () => {
						unlisten();
						clearTimeout(settlingTimeout);
						removeAllSubtitlesTracksFromMedia(mediaElement);
						peek(context.textTracksActor)?.send({ type: "clear" });
					};
				},
				effects: () => {
					syncTextTrackModes(peek(context.mediaElement).textTracks, state.selectedTextTrackId.get());
				}
			}
		}
	});
}
const syncTextTracks = defineBehavior({
	stateKeys: ["presentation", "selectedTextTrackId"],
	contextKeys: ["mediaElement", "textTracksActor"],
	setup: syncTextTracksSetup
});
//#endregion
export { syncTextTracks };
