// @videojs/html/dist/dev/ui/menu/menu-item-element.js
import { MediaElement } from "../media-element.js";
import { menuContext } from "./context.js";
import { applyElementProps, applyStateDataAttrs, completeMenuItemSelection } from "@videojs/core/dom";
import { ContextConsumer } from "@videojs/element/context";
//#region src/ui/menu/menu-item-element.ts
var MenuItemElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.disabled = false;
		this.commandfor = void 0;
		this.#ctx = new ContextConsumer(this, {
			context: menuContext,
			subscribe: true
		});
		this.#disconnect = null;
		this.#registered = false;
		this.#cleanupRegistration = null;
	}
	static {
		this.tagName = "media-menu-item";
	}
	static {
		this.properties = {
			disabled: { type: Boolean },
			commandfor: { type: String }
		};
	}
	#ctx;
	#disconnect;
	#registered;
	#cleanupRegistration;
	connectedCallback() {
		super.connectedCallback();
		this.#disconnect = new AbortController();
		this.#registered = false;
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#cleanupRegistration?.();
		this.#cleanupRegistration = null;
		this.#disconnect?.abort();
		this.#disconnect = null;
		this.#registered = false;
	}
	update(_changed) {
		super.update(_changed);
		const ctx = this.#ctx.value;
		if (!ctx || !this.#disconnect) return;
		if (!this.#registered) {
			this.#registered = true;
			this.#cleanupRegistration = ctx.menu.registerItem(this);
			applyElementProps(this, {
				onClick: (event) => {
					const currentCtx = this.#ctx.value;
					if (!currentCtx || this.disabled) return;
					const target = this.commandfor;
					if (target) currentCtx.menu.push(target, this.id);
					else {
						this.dispatchEvent(new CustomEvent("select", { bubbles: true }));
						completeMenuItemSelection(currentCtx.menu, currentCtx.parentMenu);
					}
					event.preventDefault();
				},
				onKeyDown: (event) => {
					const currentCtx = this.#ctx.value;
					if (!currentCtx || this.disabled || event.key !== "ArrowRight") return;
					const target = this.commandfor;
					if (!target) return;
					currentCtx.menu.push(target, this.id);
					event.preventDefault();
				},
				onPointerenter: () => {
					const currentCtx = this.#ctx.value;
					if (!this.disabled) currentCtx?.menu.highlight(this, { focus: false });
				}
			}, { signal: this.#disconnect.signal });
		}
		const hasSubmenu = Boolean(this.commandfor);
		const activeSubMenuId = ctx.navigation.stack[ctx.navigation.stack.length - 1]?.menuId ?? null;
		const isExpanded = hasSubmenu ? activeSubMenuId === this.commandfor : void 0;
		applyElementProps(this, {
			role: "menuitem",
			"aria-disabled": this.disabled ? "true" : void 0,
			...hasSubmenu && {
				"aria-haspopup": "menu",
				"aria-expanded": isExpanded ? "true" : "false",
				"data-has-submenu": ""
			}
		});
		applyStateDataAttrs(this, ctx.state, ctx.stateAttrMap);
	}
};
//#endregion
export { MenuItemElement };
