// @videojs/html/dist/dev/ui/menu/menu-element.js
import { containerContext } from "../../player/context.js";
import { MediaElement } from "../media-element.js";
import { menuContext } from "./context.js";
import { PositionController } from "../position-controller.js";
import { applyElementProps, applyStateDataAttrs, createMenu, createMenuViewTransition, createTransition, getAnchorNameStyle, getAnchorPositionStyle, getMenuViewTransitionAttrs, getMenuViewportAttrs, getPopupPositionRect, getPositioningBoundaryRect, getRootPositionOptions, isMenuNavigationKey, resolveOffsets, resolvePositioningBoundary, syncMenuViewRoot, syncMenuViewTransition } from "@videojs/core/dom";
import { applyStyles, supportsAnchorPositioning, tryHidePopover, tryShowPopover } from "@videojs/utils/dom";
import { ContextConsumer, ContextProvider } from "@videojs/element/context";
import { SnapshotController } from "@videojs/store/html";
import { MenuCore, MenuDataAttrs, POPUP_HOST_ATTR } from "@videojs/core";
//#region src/ui/menu/menu-element.ts
var MenuElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.open = MenuCore.defaultProps.open;
		this.defaultOpen = MenuCore.defaultProps.defaultOpen;
		this.side = MenuCore.defaultProps.side;
		this.align = MenuCore.defaultProps.align;
		this.closeOnEscape = MenuCore.defaultProps.closeOnEscape;
		this.closeOnOutsideClick = MenuCore.defaultProps.closeOnOutsideClick;
		this.boundary = "container";
		this.#core = new MenuCore();
		this.#provider = new ContextProvider(this, { context: menuContext });
		this.#position = new PositionController(this);
		this.#containerCtx = new ContextConsumer(this, {
			context: containerContext,
			subscribe: true
		});
		this.#parentCtx = new ContextConsumer(this, {
			context: menuContext,
			subscribe: true
		});
		this.#menuViewTransition = createMenuViewTransition({
			focusFirstItem: () => {
				this.#menu?.highlightFirstItem({ preventScroll: true });
			},
			restoreFocus: (triggerId) => {
				const triggerElement = triggerId ? document.getElementById(triggerId) : null;
				const fallbackTrigger = this.parentElement?.querySelector(`[data-has-submenu][commandfor="${this.id}"]`);
				(triggerElement ?? fallbackTrigger)?.focus({ preventScroll: true });
			}
		});
		this.#menu = null;
		this.#snapshot = null;
		this.#navSnapshot = null;
		this.#menuViewSnapshot = null;
		this.#navState = {
			stack: [],
			direction: "forward"
		};
		this.#disconnect = null;
		this.#triggerAbort = null;
		this.#currentTrigger = null;
		this.#handleContentKeyDown = (event) => {
			const isNavigationKey = isMenuNavigationKey(event);
			const defaultPreventedBeforeMenu = event.defaultPrevented;
			this.#menu?.contentProps.onKeyDown(event);
			const parentCtx = this.#parentCtx.value ?? null;
			if (!parentCtx) {
				if (event.key === "Escape") return;
				if (isNavigationKey) event.stopPropagation();
				return;
			}
			const stack = parentCtx.menu.navigationInput.current.stack;
			const ownsActiveSubmenu = stack[stack.length - 1]?.menuId === this.id;
			const isBackNavigationKey = event.key === "ArrowLeft" || event.key === "Escape";
			if (isBackNavigationKey && ownsActiveSubmenu && !defaultPreventedBeforeMenu) {
				event.preventDefault();
				parentCtx.menu.pop();
			}
			if (isNavigationKey && (!isBackNavigationKey || ownsActiveSubmenu)) event.stopPropagation();
		};
		this.#handleContentFocusOut = (event) => {
			this.#menu?.contentProps.onFocusOut(event);
		};
	}
	static {
		this.tagName = "media-menu";
	}
	static {
		this.properties = {
			open: { type: Boolean },
			defaultOpen: {
				type: Boolean,
				attribute: "default-open"
			},
			side: { type: String },
			align: { type: String },
			closeOnEscape: {
				type: Boolean,
				attribute: "close-on-escape"
			},
			closeOnOutsideClick: {
				type: Boolean,
				attribute: "close-on-outside-click"
			},
			boundary: { type: String }
		};
	}
	#core;
	#provider;
	#position;
	#containerCtx;
	#parentCtx;
	#menuViewTransition;
	#menu;
	#snapshot;
	#navSnapshot;
	#menuViewSnapshot;
	#navState;
	#disconnect;
	#triggerAbort;
	#currentTrigger;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		this.setAttribute(POPUP_HOST_ATTR, "");
		this.#disconnect = new AbortController();
		this.#menu = createMenu({
			transition: createTransition(),
			onOpenChange: (nextOpen, details) => {
				this.open = nextOpen;
				this.dispatchEvent(new CustomEvent("open-change", { detail: {
					open: nextOpen,
					...details
				} }));
			},
			closeOnEscape: () => this.closeOnEscape,
			closeOnOutsideClick: () => this.closeOnOutsideClick,
			group: () => this.#parentCtx.value ? void 0 : this.#containerCtx.value?.popupGroup
		});
		this.#menu.setContentElement(this);
		applyElementProps(this, {
			onKeyDown: this.#handleContentKeyDown,
			onFocusOut: this.#handleContentFocusOut
		}, { signal: this.#disconnect.signal });
		if (this.#snapshot) this.#snapshot.track(this.#menu.input);
		else this.#snapshot = new SnapshotController(this, this.#menu.input);
		if (this.#navSnapshot) this.#navSnapshot.track(this.#menu.navigationInput);
		else this.#navSnapshot = new SnapshotController(this, this.#menu.navigationInput);
		if (this.#menuViewSnapshot) this.#menuViewSnapshot.track(this.#menuViewTransition.input);
		else this.#menuViewSnapshot = new SnapshotController(this, this.#menuViewTransition.input);
	}
	firstUpdated(changed) {
		super.firstUpdated(changed);
		if (this.defaultOpen && !this.open) this.#menu?.open();
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#cleanupTrigger();
		this.#menu?.destroy();
		this.#menu = null;
		this.#disconnect?.abort();
		this.#disconnect = null;
		this.#menuViewTransition.destroy();
	}
	close(reason = "imperative-action") {
		this.#menu?.close(reason);
	}
	willUpdate(changed) {
		super.willUpdate(changed);
		const isSubmenu = (this.#parentCtx.value ?? null) !== null;
		this.#core.setProps({
			open: this.open,
			defaultOpen: this.defaultOpen,
			side: this.side,
			align: this.align,
			closeOnEscape: this.closeOnEscape,
			closeOnOutsideClick: this.closeOnOutsideClick,
			isSubmenu
		});
		if (this.#menu && changed.has("open") && !isSubmenu) {
			const { active: interactionOpen } = this.#menu.input.current;
			if (this.open !== interactionOpen) if (this.open) this.#menu.open();
			else this.#menu.close();
		}
	}
	update(_changed) {
		super.update(_changed);
		if (!this.#menu) return;
		const parentCtx = this.#parentCtx.value ?? null;
		const isSubmenu = parentCtx !== null;
		this.#navState = this.#menu.navigationInput.current;
		const input = this.#menu.input.current;
		this.#core.setInput(input);
		const state = this.#core.getState();
		if (isSubmenu && parentCtx) this.#updateAsSubmenu(parentCtx);
		else this.#updateAsRoot(state);
		const parentMenu = parentCtx?.menu ?? null;
		this.#provider.setValue({
			menu: this.#menu,
			state,
			stateAttrMap: MenuDataAttrs,
			navigation: this.#navState,
			parentMenu
		});
	}
	#updateAsRoot(state) {
		if (!this.#menu) return;
		const triggerElement = this.#position.findTrigger();
		this.#syncTrigger(triggerElement);
		applyElementProps(this, {
			...this.#core.getContentAttrs(state),
			...getMenuViewportAttrs()
		});
		applyStateDataAttrs(this, state, MenuDataAttrs);
		if (state.open) tryShowPopover(this);
		else tryHidePopover(this);
		if (this.#currentTrigger) {
			applyElementProps(this.#currentTrigger, this.#core.getTriggerAttrs(state, this.id));
			applyStyles(this.#currentTrigger, getAnchorNameStyle(this.id));
		}
		if (!state.open) {
			this.#position.cleanup();
			return;
		}
		syncMenuViewRoot(this, this.#navState.stack.length > 0);
		const positionOptions = getRootPositionOptions(state.side, state.align);
		if (!positionOptions) return;
		const boundaryElement = this.#getBoundaryElement();
		const triggerRect = this.#currentTrigger?.getBoundingClientRect();
		const boundaryRect = getPositioningBoundaryRect(boundaryElement);
		const offsets = resolveOffsets(this);
		if (supportsAnchorPositioning()) applyStyles(this, getAnchorPositionStyle(this.id, positionOptions, triggerRect, void 0, boundaryRect, offsets));
		else {
			const selfRect = getPopupPositionRect(this);
			applyStyles(this, getAnchorPositionStyle(this.id, positionOptions, triggerRect, selfRect, boundaryRect, offsets));
		}
		this.#position.sync(this.#currentTrigger, boundaryElement);
	}
	#updateAsSubmenu(parentCtx) {
		const parentNavigation = parentCtx.navigation;
		const topEntry = parentNavigation.stack[parentNavigation.stack.length - 1];
		const isActive = (topEntry?.menuId ?? null) === this.id;
		this.#menuViewTransition.setElement(this);
		this.#menuViewTransition.sync({
			active: isActive,
			direction: parentNavigation.direction,
			triggerId: topEntry?.triggerId ?? null
		});
		const transitionState = this.#menuViewTransition.input.current;
		applyElementProps(this, {
			...getMenuViewTransitionAttrs(transitionState),
			role: "menu",
			tabIndex: -1,
			"data-submenu": ""
		});
		syncMenuViewTransition(parentCtx.menu.contentElement, this, transitionState);
	}
	#handleContentKeyDown;
	#handleContentFocusOut;
	#syncTrigger(triggerElement) {
		if (triggerElement === this.#currentTrigger) return;
		this.#position.cleanup();
		this.#cleanupTrigger();
		this.#currentTrigger = triggerElement;
		this.#menu?.setTriggerElement(triggerElement);
		if (triggerElement && this.#menu) {
			this.#triggerAbort = new AbortController();
			applyElementProps(triggerElement, this.#menu.triggerProps, { signal: this.#triggerAbort.signal });
		}
	}
	#cleanupTrigger() {
		if (this.#currentTrigger) {
			applyElementProps(this.#currentTrigger, {
				"aria-expanded": void 0,
				"aria-haspopup": void 0,
				"aria-controls": void 0
			});
			this.#currentTrigger.style.removeProperty("anchor-name");
		}
		this.#triggerAbort?.abort();
		this.#triggerAbort = null;
		this.#currentTrigger = null;
	}
	#getBoundaryElement() {
		return resolvePositioningBoundary(this.boundary, {
			container: this.#containerCtx.value?.container ?? null,
			root: this.getRootNode()
		});
	}
};
//#endregion
export { MenuElement };
