// @videojs/html/dist/dev/ui/menu/menu-back-element.js
import { MediaElement } from "../media-element.js";
import { menuContext } from "./context.js";
import { applyElementProps, applyStateDataAttrs } from "@videojs/core/dom";
import { ContextConsumer } from "@videojs/element/context";
//#region src/ui/menu/menu-back-element.ts
var MenuBackElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.label = "Back";
		this.#ctx = new ContextConsumer(this, {
			context: menuContext,
			subscribe: true
		});
		this.#disconnect = null;
		this.#bound = false;
	}
	static {
		this.tagName = "media-menu-back";
	}
	static {
		this.properties = { label: { type: String } };
	}
	#ctx;
	#disconnect;
	#bound;
	connectedCallback() {
		super.connectedCallback();
		this.#disconnect = new AbortController();
		this.#bound = false;
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#disconnect?.abort();
		this.#disconnect = null;
		this.#bound = false;
	}
	update(_changed) {
		super.update(_changed);
		const ctx = this.#ctx.value;
		if (!ctx || !this.#disconnect) return;
		if (!this.#bound) {
			this.#bound = true;
			applyElementProps(this, { onClick: () => {
				ctx.parentMenu?.pop();
			} }, { signal: this.#disconnect.signal });
		}
		applyElementProps(this, {
			role: "button",
			"aria-label": this.label
		});
		if (ctx) applyStateDataAttrs(this, ctx.state, ctx.stateAttrMap);
	}
};
//#endregion
export { MenuBackElement };
