// @videojs/html/dist/default/ui/slider/slider-preview-element.js
import { MediaElement } from "../media-element.js";
import { sliderContext } from "./context.js";
import { applyStateDataAttrs, getSliderPreviewStyle } from "@videojs/core/dom";
import { applyStyles } from "@videojs/utils/dom";
import { ContextConsumer } from "@videojs/element/context";
//#region src/ui/slider/slider-preview-element.ts
var SliderPreviewElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.overflow = "clamp";
		this.#ctx = new ContextConsumer(this, {
			context: sliderContext,
			subscribe: true
		});
		this.#resizeObserver = null;
		this.#width = 0;
	}
	static {
		this.tagName = "media-slider-preview";
	}
	static {
		this.properties = { overflow: { type: String } };
	}
	#ctx;
	#resizeObserver;
	#width;
	connectedCallback() {
		super.connectedCallback();
		this.#resizeObserver = new ResizeObserver(([entry]) => {
			this.#width = entry.contentRect.width;
			this.#applyPosition();
		});
		this.#resizeObserver.observe(this);
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#resizeObserver?.disconnect();
		this.#resizeObserver = null;
	}
	#applyPosition() {
		applyStyles(this, getSliderPreviewStyle(this.#width, this.overflow));
	}
	update(_changed) {
		super.update(_changed);
		const ctx = this.#ctx.value;
		if (ctx) applyStateDataAttrs(this, ctx.state, ctx.stateAttrMap);
		this.#applyPosition();
	}
};
//#endregion
export { SliderPreviewElement };
