// vidstack/server/global/vidstack-player.js
import "../define/vidstack-player.js";
//#region src/global/layouts/default.ts
var VidstackPlayerLayout = class {
	props;
	constructor(props) {
		this.props = props;
	}
	name = "vidstack";
	async load() {
		await import("../define/vidstack-player-default-layout.js");
		await import("../define/vidstack-player-ui.js");
	}
	create() {
		const layouts = [document.createElement("media-audio-layout"), document.createElement("media-video-layout")];
		if (this.props) for (const [prop, value] of Object.entries(this.props)) for (const el of layouts) el[prop] = value;
		return layouts;
	}
};
//#endregion
//#region src/global/layouts/plyr.ts
var PlyrLayout = class {
	props;
	constructor(props) {
		this.props = props;
	}
	name = "plyr";
	async load() {
		await import("../define/plyr-layout.js");
	}
	create() {
		const layout = document.createElement("media-plyr-layout");
		if (this.props) for (const [prop, value] of Object.entries(this.props)) layout[prop] = value;
		return [layout];
	}
};
//#endregion
//#region src/global/player.ts
var VidstackPlayer = class {
	static async create({ target, layout, tracks, ...props }) {
		throw Error("[vidstack] can not create player on server.");
	}
};
//#endregion
export { PlyrLayout, VidstackPlayer, VidstackPlayerLayout };
