// @videojs/spf/dist/default/playback/engines/hls/engine.js
import { createComposition } from "../../../core/composition/create-composition.js";
import { makeShareSignals } from "../../../core/composition/share-signals.js";
import { resolveVttSegment } from "../../../media/dom/text/resolve-vtt-segment.js";
import { loadAudioSegments, loadTextTrackSegments, loadVideoSegments } from "../../behaviors/dom/load-segments.js";
import { setupTextTrackActors } from "../../behaviors/dom/setup-text-track-actors.js";
import { trackCurrentTime } from "../../behaviors/dom/track-current-time.js";
import { trackLoadTriggers } from "../../behaviors/dom/track-load-triggers.js";
import { addSubtitlesTracksToMedia, getShowingSubtitlesTrackFromMedia, removeAllSubtitlesTracksFromMedia } from "../../../media/dom/text/text-track-slots.js";
import { parseMultivariantPlaylist } from "../../../media/hls/parse-multivariant.js";
import { getResolvedSelectedTrackDuration } from "../../../media/utils/track-selection.js";
import { calculatePresentationDuration } from "../../behaviors/calculate-presentation-duration.js";
import { endOfStream } from "../../behaviors/dom/end-of-stream.js";
import { setupAudioBufferActors, setupVideoBufferActors } from "../../behaviors/dom/setup-buffer-actors.js";
import { setupMediaSource } from "../../behaviors/dom/setup-mediasource.js";
import { syncTextTracks } from "../../behaviors/dom/sync-text-tracks.js";
import { updateMediaSourceDuration } from "../../behaviors/dom/update-mediasource-duration.js";
import { switchVideoQuality } from "../../behaviors/quality-switching.js";
import { resolvePresentation } from "../../behaviors/resolve-presentation.js";
import { resolveAudioTrack, resolveTextTrack, resolveVideoTrack } from "../../behaviors/resolve-track.js";
import { selectAudioTrack, selectTextTrack } from "../../behaviors/select-tracks.js";
import { syncPreload } from "../../behaviors/sync-preload.js";
//#region src/playback/engines/hls/engine.ts
/**
* Generic `shareSignals` instantiated against the HLS engine's full state
* and context — captures composition signal refs into the consumer's
* `onSignalsReady` callback at setup time.
*/
const shareSignals = makeShareSignals();
/**
* Create an HLS playback engine.
*
* Composes SPF behaviors into a reactive pipeline for HLS playback over MSE:
* manifest resolution, track selection, ABR, segment loading, and
* end-of-stream coordination.
*
* @example
* ```ts
* let signals: SimpleHlsEngineSignals;
* const engine = createSimpleHlsEngine({
*   initialBandwidth: 2_000_000,
*   preferredAudioLanguage: 'en',
*   onSignalsReady: (refs) => {
*     signals = refs;
*   },
* });
*
* signals.context.mediaElement.set(videoEl);
* signals.state.presentation.set({ url: 'https://example.com/stream.m3u8' });
*
* videoEl.play();
*
* await engine.destroy();
* ```
*/
function createSimpleHlsEngine(config = {}) {
	const finalConfig = {
		...config,
		resolveTextTrackSegment: config.resolveTextTrackSegment ?? resolveVttSegment,
		resolveDuration: config.resolveDuration ?? getResolvedSelectedTrackDuration,
		parsePresentation: config.parsePresentation ?? parseMultivariantPlaylist,
		addSubtitlesTracksToMedia: config.addSubtitlesTracksToMedia ?? addSubtitlesTracksToMedia,
		getShowingSubtitlesTrackFromMedia: config.getShowingSubtitlesTrackFromMedia ?? getShowingSubtitlesTrackFromMedia,
		removeAllSubtitlesTracksFromMedia: config.removeAllSubtitlesTracksFromMedia ?? removeAllSubtitlesTracksFromMedia
	};
	return createComposition([
		syncPreload,
		trackLoadTriggers,
		resolvePresentation,
		selectAudioTrack,
		selectTextTrack,
		resolveVideoTrack,
		resolveAudioTrack,
		resolveTextTrack,
		calculatePresentationDuration,
		setupMediaSource,
		updateMediaSourceDuration,
		setupVideoBufferActors,
		setupAudioBufferActors,
		trackCurrentTime,
		switchVideoQuality,
		loadVideoSegments,
		loadAudioSegments,
		endOfStream,
		syncTextTracks,
		setupTextTrackActors,
		loadTextTrackSegments,
		shareSignals
	], {
		config: finalConfig,
		initialState: { bandwidthState: {
			fastEstimate: 0,
			fastTotalWeight: 0,
			slowEstimate: 0,
			slowTotalWeight: 0,
			bytesSampled: 0
		} }
	});
}
//#endregion
export { createSimpleHlsEngine };
