// @videojs/html/dist/default/ui/menu/menu-radio-group-element.js
import { MediaElement } from "../media-element.js";
import { menuContext, menuRadioGroupContext } from "./context.js";
import { applyElementProps, applyStateDataAttrs } from "@videojs/core/dom";
import { ContextConsumer, ContextProvider } from "@videojs/element/context";
//#region src/ui/menu/menu-radio-group-element.ts
var MenuRadioGroupElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.value = "";
		this.label = void 0;
		this.#menuCtx = new ContextConsumer(this, {
			context: menuContext,
			subscribe: true
		});
		this.#provider = new ContextProvider(this, { context: menuRadioGroupContext });
	}
	static {
		this.tagName = "media-menu-radio-group";
	}
	static {
		this.properties = {
			value: { type: String },
			label: { type: String }
		};
	}
	#menuCtx;
	#provider;
	update(_changed) {
		super.update(_changed);
		applyElementProps(this, {
			role: "group",
			"aria-label": this.label
		});
		const ctx = this.#menuCtx.value;
		if (ctx) applyStateDataAttrs(this, ctx.state, ctx.stateAttrMap);
		this.#provider.setValue({
			value: this.value,
			onValueChange: (next) => {
				this.value = next;
				this.dispatchEvent(new CustomEvent("value-change", {
					detail: { value: next },
					bubbles: true
				}));
			}
		});
	}
};
//#endregion
export { MenuRadioGroupElement };
