// @videojs/core/dist/default/dom/media/mux/mux-data.js
import { Hls } from "../hls/index.js";
import { getPlayerVersion } from "./env.js";
import Mux from "mux-embed";
//#region src/dom/media/mux/mux-data.ts
const muxDataMediaDefaultProps = {
	MuxDataSdk: void 0,
	beaconCollectionDomain: void 0,
	disableCookies: false,
	envKey: void 0,
	playerSoftwareName: void 0,
	playerSoftwareVersion: void 0,
	playerInitTime: void 0,
	metadata: void 0
};
const MUX_VIDEO_DOMAIN = "mux.com";
const MuxDataMediaMixin = (BaseClass) => {
	class MuxDataMedia extends BaseClass {
		#MuxDataSdk = Mux;
		#MuxDataSdkInitializedBefore = false;
		#beaconCollectionDomain;
		#disableCookies = false;
		#metadata;
		#envKey;
		#playerSoftwareName = this.constructor.PLAYER_SOFTWARE_NAME;
		#playerSoftwareVersion = getPlayerVersion();
		#playerInitTime = this.#generatePlayerInitTime();
		get MuxDataSdk() {
			return this.#MuxDataSdk;
		}
		set MuxDataSdk(value) {
			this.#MuxDataSdk = value;
		}
		get beaconCollectionDomain() {
			return this.#beaconCollectionDomain;
		}
		set beaconCollectionDomain(value) {
			this.#beaconCollectionDomain = value;
		}
		get disableCookies() {
			return this.#disableCookies;
		}
		set disableCookies(value) {
			this.#disableCookies = value;
		}
		get envKey() {
			return this.#envKey;
		}
		set envKey(value) {
			this.#envKey = value;
		}
		get playerSoftwareName() {
			return this.#playerSoftwareName;
		}
		set playerSoftwareName(value) {
			this.#playerSoftwareName = value;
		}
		get playerSoftwareVersion() {
			return this.#playerSoftwareVersion;
		}
		set playerSoftwareVersion(value) {
			this.#playerSoftwareVersion = value;
		}
		get playerInitTime() {
			return this.#playerInitTime;
		}
		set playerInitTime(value) {
			this.#playerInitTime = value;
		}
		get metadata() {
			return this.#metadata;
		}
		set metadata(value) {
			this.#metadata = value;
		}
		attach(target) {
			super.attach(target);
			if (this.#MuxDataSdkInitializedBefore) this.#initializeMuxDataSdk();
		}
		detach() {
			if (this.target?.mux) {
				this.target.mux.destroy();
				delete this.target.mux;
			}
			super.detach();
		}
		load() {
			super.load();
			this.#initializeMuxDataSdk();
		}
		#initializeMuxDataSdk() {
			const target = this.target;
			if (!this.MuxDataSdk || !target || target.mux && !target.mux.deleted) return;
			this.#MuxDataSdkInitializedBefore = true;
			const { debug, beaconCollectionDomain, disableCookies, engine: hlsjs, envKey: env_key, playerSoftwareName: player_software_name, playerSoftwareVersion: player_software_version, playerInitTime: player_init_time, metadata = {} } = this;
			const { view_session_id = this.MuxDataSdk?.utils.generateUUID() } = metadata;
			const video_id = toVideoId(this);
			metadata.view_session_id = view_session_id;
			if (video_id) metadata.video_id = video_id;
			this.MuxDataSdk?.monitor(target, {
				debug,
				...beaconCollectionDomain ? { beaconCollectionDomain } : {},
				...disableCookies ? { disableCookies } : {},
				...hlsjs ? { hlsjs } : {},
				Hls,
				data: {
					...env_key ? { env_key } : {},
					...player_software_name ? { player_software_name } : {},
					...player_software_name ? { player_software: player_software_name } : {},
					...player_software_version ? { player_software_version } : {},
					...player_init_time ? { player_init_time } : {},
					...metadata
				}
			});
		}
		#generatePlayerInitTime() {
			if (!this.MuxDataSdk) return void 0;
			return this.MuxDataSdk.utils.now();
		}
	}
	return MuxDataMedia;
};
function toVideoId(props) {
	if (props.metadata?.video_id) return props.metadata.video_id;
	if (!isMuxVideoSrc(props)) return props.src;
	return toPlaybackIdFromSrc(props.src) ?? props.src;
}
function toPlaybackIdFromSrc(src) {
	if (!src || !src.startsWith("https://stream.")) return void 0;
	const [playbackId] = new URL(src).pathname.slice(1).split(/\.m3u8|\//);
	return playbackId || void 0;
}
function isMuxVideoSrc({ src }) {
	if (typeof src !== "string") return false;
	const base = window?.location.href;
	return new URL(src, base).hostname.toLocaleLowerCase().includes(MUX_VIDEO_DOMAIN);
}
//#endregion
export { MuxDataMediaMixin, muxDataMediaDefaultProps };
