// @videojs/html/dist/default/ui/media-button-element.js
import { MediaElement } from "./media-element.js";
import { AriaKeyShortcutsController } from "./hotkey/aria-key-shortcuts-controller.js";
import { applyElementProps, applyStateDataAttrs, createButton } from "@videojs/core/dom";
//#region src/ui/media-button-element.ts
/** Abstract base for HTML custom elements that render a media-control button. */
var MediaButtonElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.disabled = false;
		this.label = "";
		this.hotkeyAction = void 0;
		this.#disconnect = null;
		this.#hotkeyRegistry = null;
	}
	static {
		this.properties = {
			label: { type: String },
			disabled: { type: Boolean }
		};
	}
	get $state() {
		return this.core.state;
	}
	#disconnect;
	#hotkeyRegistry;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		if (this.hotkeyAction && !this.#hotkeyRegistry) this.#hotkeyRegistry = new AriaKeyShortcutsController(this, this.hotkeyAction);
		this.#disconnect = new AbortController();
		const buttonProps = createButton({
			onActivate: () => this.activate(this.mediaState.value),
			isDisabled: () => this.disabled || !this.mediaState.value
		});
		applyElementProps(this, buttonProps, { signal: this.#disconnect.signal });
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#disconnect?.abort();
		this.#disconnect = null;
	}
	/** Returns the button's current label derived from media state. */
	getLabel() {
		return this.core.state.current.label || void 0;
	}
	willUpdate(changed) {
		super.willUpdate(changed);
		this.core.setProps?.(this);
	}
	update(changed) {
		super.update(changed);
		const media = this.mediaState.value;
		if (!media) return;
		this.core.setMedia(media);
		const state = this.core.getState();
		applyElementProps(this, {
			...this.core.getAttrs?.(state),
			"aria-keyshortcuts": this.#hotkeyRegistry?.value
		});
		applyStateDataAttrs(this, state, this.stateAttrMap);
	}
};
//#endregion
export { MediaButtonElement };
