// @videojs/html/cdn/base-CaXY3kMA.js
//#region ../icons/dist/element/base.js
var MediaIconElement = class MediaIconElement extends HTMLElement {
	static #families = /* @__PURE__ */ new Map();
	static #loaders = /* @__PURE__ */ new Map();
	static #loading = /* @__PURE__ */ new Map();
	static #instances = /* @__PURE__ */ new Set();
	static register(family, icons) {
		const map = MediaIconElement.#families.get(family) ?? /* @__PURE__ */ new Map();
		for (const [name, svg] of Object.entries(icons)) map.set(name, svg);
		MediaIconElement.#families.set(family, map);
		MediaIconElement.#renderFamily(family);
	}
	static registerLoader(family, load) {
		MediaIconElement.#loaders.set(family, load);
	}
	static load(family) {
		if (MediaIconElement.#families.has(family)) return Promise.resolve();
		const pending = MediaIconElement.#loading.get(family);
		if (pending) return pending;
		const loader = MediaIconElement.#loaders.get(family);
		if (!loader) return Promise.resolve();
		const loading = Promise.resolve().then(() => loader()).then((icons) => {
			if (icons) MediaIconElement.register(family, icons);
		}).finally(() => MediaIconElement.#loading.delete(family));
		MediaIconElement.#loading.set(family, loading);
		return loading;
	}
	static #renderFamily(family) {
		for (const icon of MediaIconElement.#instances) if (icon.#family === family) icon.#render();
	}
	static get observedAttributes() {
		return ["name", "family"];
	}
	attributeChangedCallback() {
		this.#render();
	}
	connectedCallback() {
		MediaIconElement.#instances.add(this);
		this.#render();
	}
	disconnectedCallback() {
		MediaIconElement.#instances.delete(this);
	}
	get #family() {
		return this.getAttribute("family") || "default";
	}
	#render() {
		const name = this.getAttribute("name");
		if (!name) return;
		const family = this.#family;
		const svg = MediaIconElement.#families.get(family)?.get(name);
		if (!svg) {
			if (MediaIconElement.#families.has(family)) return;
			MediaIconElement.load(family).then(() => {
				if (!this.isConnected || this.getAttribute("name") !== name || this.#family !== family || !MediaIconElement.#families.has(family)) return;
				this.#render();
			}, () => {});
			return;
		}
		this.innerHTML = svg;
	}
};

//#endregion
export { MediaIconElement as t };
