// @videojs/html/cdn/live-button-element-BMneBAzX.js
import { C as selectBuffer, N as defaults, T as selectLive, b as applyElementProps, j as selectTime, v as applyStateDataAttrs, x as createButton, y as logMissingFeature } from "./compounds-DxdDY3jr.js";
import { E as createState, j as MediaElement, n as PlayerController } from "./create-player-_WED6HtK.js";
import { l as playerContext, t as isFunction } from "./predicate-Dqd8wbW6.js";

//#region ../core/dist/dev/core/ui/live-button/live-button-core.js
/**
* Fallback offset (in seconds) from the end of the seekable window used to
* decide "at live edge" when `liveEdgeStart` is unavailable.
*/
const LIVE_EDGE_OFFSET = 10;
/**
* Grace window (in seconds) before `liveEdgeStart` that still counts as
* "at the live edge". Absorbs the small gap between the player's initial
* playback position (e.g. hls.js `liveSyncDuration`) and the manifest's
* `HOLD-BACK`, so autoplay reliably reports live.
*/
const LIVE_EDGE_TOLERANCE = 5;
/**
* Core state machine for a "Live" button. Indicates whether the player is
* playing at the live edge and seeks to the Seekable Live Edge when activated.
*
* @see https://github.com/video-dev/media-ui-extensions/blob/main/proposals/0007-live-edge.md
*/
var LiveButtonCore = class LiveButtonCore {
	/**
	* Default visible text. Auto-inserted by `media-live-button` and `<LiveButton>`
	* when no children are provided. Override globally for i18n:
	*
	* ```ts
	* LiveButtonCore.defaultText = 'En Vivo';
	* ```
	*/
	static defaultText = "Live";
	static defaultProps = {
		label: "",
		disabled: false
	};
	state = createState({
		live: false,
		liveEdge: false,
		label: ""
	});
	#props = { ...LiveButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, LiveButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		if (state.liveEdge) return "Playing live";
		return "Seek to live edge";
	}
	getAttrs(state) {
		const inactive = this.#props.disabled || state.liveEdge;
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": inactive ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		const live = isLiveMedia(media);
		const liveEdge = live && this.#isAtLiveEdge(media);
		this.state.patch({
			live,
			liveEdge
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	/** Seek to the Seekable Live Edge. No-op when not live or already at edge. */
	async seekToLive(media) {
		if (this.#props.disabled) return;
		if (!isLiveMedia(media)) return;
		if (this.#isAtLiveEdge(media)) return;
		const target = liveEdgeTarget(media);
		if (target == null) return;
		await media.seek(target);
	}
	#isAtLiveEdge(media) {
		const { currentTime, liveEdgeStart } = media;
		if (Number.isFinite(liveEdgeStart)) return currentTime >= liveEdgeStart - LIVE_EDGE_TOLERANCE;
		const target = liveEdgeTarget(media);
		if (target == null) return false;
		return currentTime >= target - LIVE_EDGE_OFFSET;
	}
};
function isLiveMedia(media) {
	return !Number.isNaN(media.targetLiveWindow);
}
function liveEdgeTarget(media) {
	const { seekable } = media;
	if (seekable.length === 0) return null;
	const end = seekable[seekable.length - 1][1];
	return Number.isFinite(end) ? end : null;
}

//#endregion
//#region ../core/dist/dev/core/ui/live-button/live-button-data-attrs.js
const LiveButtonDataAttrs = {
	live: "data-live",
	liveEdge: "data-live-edge"
};

//#endregion
//#region src/ui/live-button/live-button-element.ts
/**
* `<media-live-button>` — selects from `live`, `time`, and `buffer` features
* and composes them into the `LiveButtonMediaState` consumed by
* `LiveButtonCore`.
*
* Doesn't extend `MediaButtonElement` because that base couples a button to
* a single feature selector; the LiveButton needs three.
*/
var LiveButtonElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.disabled = false;
		this.label = "";
		this.core = new LiveButtonCore();
		this.live = new PlayerController(this, playerContext, selectLive);
		this.time = new PlayerController(this, playerContext, selectTime);
		this.buffer = new PlayerController(this, playerContext, selectBuffer);
		this.#disconnect = null;
	}
	static {
		this.tagName = "media-live-button";
	}
	static {
		this.properties = {
			label: { type: String },
			disabled: { type: Boolean }
		};
	}
	get $state() {
		return this.core.state;
	}
	#disconnect;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		if (!this.textContent?.trim()) this.textContent = LiveButtonCore.defaultText;
		this.#disconnect = new AbortController();
		const buttonProps = createButton({
			onActivate: () => {
				const media = this.#getMedia();
				if (media) this.core.seekToLive(media);
			},
			isDisabled: () => this.disabled || !this.#getMedia()
		});
		applyElementProps(this, buttonProps, { signal: this.#disconnect.signal });
		if (!this.#getMedia()) logMissingFeature(this.localName, this.live.displayName ?? "live");
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#disconnect?.abort();
		this.#disconnect = null;
	}
	/** Returns the button's current label derived from media state. */
	getLabel() {
		return this.core.state.current.label || void 0;
	}
	willUpdate(changed) {
		super.willUpdate(changed);
		this.core.setProps(this);
	}
	update(changed) {
		super.update(changed);
		const media = this.#getMedia();
		if (!media) return;
		this.core.setMedia(media);
		const state = this.core.getState();
		applyElementProps(this, this.core.getAttrs(state));
		applyStateDataAttrs(this, state, LiveButtonDataAttrs);
	}
	/**
	* Compose the LiveButton media state from the three feature slices.
	* Returns `null` when any are missing so the button stays disabled until
	* all three features are registered on the player.
	*/
	#getMedia() {
		const live = this.live.value;
		const time = this.time.value;
		const buffer = this.buffer.value;
		if (!live || !time || !buffer) return null;
		return {
			currentTime: time.currentTime,
			seek: time.seek,
			seekable: buffer.seekable,
			liveEdgeStart: live.liveEdgeStart,
			targetLiveWindow: live.targetLiveWindow
		};
	}
};

//#endregion
export { LiveButtonElement as t };
