// @videojs/html/cdn/media-attach-mixin-C1KQy-y8.js
import { c as mediaContext, d as ContextRequestEvent } from "./predicate-Dqd8wbW6.js";

//#region ../utils/dist/dom/attributes.js
/**
* Convert a NamedNodeMap to a plain object.
*/
function namedNodeMapToObject(namedNodeMap) {
	const obj = {};
	for (const attr of namedNodeMap) obj[attr.name] = attr.value;
	return obj;
}
/**
* Helper function to serialize attributes into a string.
*/
function serializeAttributes(attrs) {
	let html = "";
	for (const key in attrs) {
		const value = attrs[key];
		if (value === "") html += ` ${key}`;
		else html += ` ${key}="${value}"`;
	}
	return html;
}

//#endregion
//#region src/store/media-attach-mixin.ts
/**
* Create a mixin that consumes `mediaContext` and registers the
* element as the media with the provider.
*
* Uses the raw context-request protocol so it works with any
* `HTMLElement` subclass — no `ReactiveControllerHost` required.
*
* @param context - The media context to consume.
*/
function createMediaAttachMixin(context) {
	return (BaseClass) => {
		class MediaAttachElement extends BaseClass {
			#setMedia = null;
			#unsubscribe = null;
			getMediaTarget() {
				return this;
			}
			connectedCallback() {
				super.connectedCallback?.();
				this.dispatchEvent(new ContextRequestEvent(context, this, (value, unsubscribe) => {
					if (unsubscribe) this.#unsubscribe = unsubscribe;
					this.#setMedia = value?.setMedia ?? null;
					if (this.isConnected) this.#setMedia?.(this.getMediaTarget());
				}, true));
			}
			disconnectedCallback() {
				super.disconnectedCallback?.();
				this.#setMedia?.(null);
				this.#unsubscribe?.();
				this.#unsubscribe = null;
				this.#setMedia = null;
			}
		}
		return MediaAttachElement;
	};
}
const MediaAttachMixin = createMediaAttachMixin(mediaContext);

//#endregion
export { namedNodeMapToObject as n, serializeAttributes as r, MediaAttachMixin as t };
