// @videojs/html/cdn/native-hls-D4f2-E9D.js
import { t as MediaStreamTypes } from "./types-B_4Es-UH.js";
import { t as HTMLVideoElementHost } from "./video-host-BU9t26fe.js";

//#region ../core/dist/dev/core/media/media-error.js
var MediaError = class MediaError extends Error {
	static MEDIA_ERR_ABORTED = 1;
	static MEDIA_ERR_NETWORK = 2;
	static MEDIA_ERR_DECODE = 3;
	static MEDIA_ERR_SRC_NOT_SUPPORTED = 4;
	static MEDIA_ERR_ENCRYPTED = 5;
	static MEDIA_ERR_CUSTOM = 100;
	static defaultMessages = {
		1: "You aborted the media playback",
		2: "A network error caused the media download to fail.",
		3: "A media error caused playback to be aborted. The media could be corrupt or your browser does not support this format.",
		4: "An unsupported error occurred. The server or network failed, or your browser does not support this format.",
		5: "The media is encrypted and there are no keys to decrypt it."
	};
	name;
	code;
	context;
	fatal;
	data;
	constructor(message, code = MediaError.MEDIA_ERR_CUSTOM, fatal, context) {
		super(message);
		this.name = "MediaError";
		this.code = code;
		this.context = context;
		this.fatal = fatal ?? (code >= MediaError.MEDIA_ERR_NETWORK && code <= MediaError.MEDIA_ERR_ENCRYPTED);
		if (!this.message) this.message = MediaError.defaultMessages[this.code] ?? "";
	}
};

//#endregion
//#region ../core/dist/dev/dom/media/native-hls/errors.js
function NativeHlsMediaErrorsMixin(BaseClass) {
	class NativeHlsMediaErrors extends BaseClass {
		#disconnect = null;
		#error = null;
		get error() {
			return this.#error;
		}
		attach(target) {
			super.attach?.(target);
			this.#init(target);
		}
		detach() {
			this.#destroy();
			super.detach?.();
		}
		destroy() {
			this.#destroy();
			super.destroy?.();
		}
		#destroy() {
			this.#disconnect?.abort();
			this.#disconnect = null;
			this.#error = null;
		}
		#init(target) {
			this.#destroy();
			this.#disconnect = new AbortController();
			const signal = this.#disconnect.signal;
			target.addEventListener("error", (event) => {
				event.stopImmediatePropagation();
				const native = target.error;
				if (!native) return;
				const error = new MediaError(native.message, native.code, true);
				this.#error = error;
				this.dispatchEvent(new ErrorEvent("error", {
					error,
					message: error.message
				}));
			}, {
				signal,
				capture: true
			});
			target.addEventListener("emptied", () => {
				this.#error = null;
			}, { signal });
		}
	}
	return NativeHlsMediaErrors;
}

//#endregion
//#region ../core/dist/dev/dom/media/native-hls/m3u8-utils.js
/**
* Returns `true` when `src` looks like an HLS playlist URL. Permissive: a
* path or query string containing `.m3u8` is enough.
*/
function looksLikeM3u8(src) {
	return src.toLowerCase().includes(".m3u8");
}
/**
* Returns `true` when the playlist text is a multivariant (master) playlist.
*
* The presence of `#EXT-X-STREAM-INF` is conclusive — media playlists only
* contain `#EXTINF` segment tags.
*/
function isMultivariantPlaylist(playlist) {
	return playlist.includes("#EXT-X-STREAM-INF");
}
/**
* Resolves the first media playlist URL referenced by a multivariant
* playlist, relative to `baseUrl`. Returns `null` when none is found or the
* URL cannot be parsed.
*/
function resolveFirstMediaPlaylistUrl(multivariant, baseUrl) {
	const lines = multivariant.split(/\r?\n/);
	const start = lines.findIndex((l) => l.startsWith("#EXT-X-STREAM-INF"));
	if (start === -1) return null;
	const uri = lines.slice(start + 1).map((l) => l.trim()).find((l) => l && !l.startsWith("#"));
	if (!uri) return null;
	try {
		return new URL(uri, baseUrl).toString();
	} catch {
		return null;
	}
}
/**
* Parses the subset of media-playlist tags needed to derive live edge state:
* `#EXT-X-PLAYLIST-TYPE`, `#EXT-X-ENDLIST`, `#EXT-X-TARGETDURATION`,
* `#EXT-X-PART-INF`.
*
* See spec:
* - VOD or `#EXT-X-ENDLIST` present → on-demand, `targetLiveWindow = NaN`.
* - `EVENT` playlist → DVR, `targetLiveWindow = Infinity`.
* - Otherwise → standard live sliding window, `targetLiveWindow = 0`.
*
* The edge offset is `PART-TARGET * 2` for low-latency live and
* `TARGETDURATION * 3` otherwise.
*/
function parseStreamInfo(playlist) {
	const lines = playlist.split(/\r?\n/);
	let playlistType;
	let hasEndList = false;
	let targetDuration;
	let partTarget;
	for (const raw of lines) {
		const line = raw.trim();
		if (line.startsWith("#EXT-X-PLAYLIST-TYPE:")) playlistType = line.slice(21).trim().toUpperCase();
		else if (line === "#EXT-X-ENDLIST") hasEndList = true;
		else if (line.startsWith("#EXT-X-TARGETDURATION:")) {
			const value = Number(line.slice(22));
			if (Number.isFinite(value)) targetDuration = value;
		} else if (line.startsWith("#EXT-X-PART-INF")) {
			const match = /PART-TARGET\s*=\s*([0-9.]+)/i.exec(line);
			if (match) {
				const value = Number(match[1]);
				if (Number.isFinite(value)) partTarget = value;
			}
		}
	}
	if (playlistType === "VOD" || hasEndList) return {
		targetLiveWindow: NaN,
		liveEdgeStartOffset: void 0
	};
	return {
		targetLiveWindow: playlistType === "EVENT" ? Number.POSITIVE_INFINITY : 0,
		liveEdgeStartOffset: partTarget !== void 0 ? partTarget * 2 : targetDuration !== void 0 ? targetDuration * 3 : void 0
	};
}
async function fetchPlaylist(url, init) {
	const response = await fetch(url, init);
	if (!response.ok) throw new Error(`Failed to fetch playlist (${response.status}): ${url}`);
	return {
		text: await response.text(),
		url: response.url || url
	};
}
/**
* Fetches the HLS playlist at `src`, following the first variant if it's a
* multivariant playlist, and parses it into a {@link StreamInfo}.
*
* @throws when the fetch fails or no media playlist URL can be resolved.
*/
async function getStreamInfoFromSrc(src, signal) {
	const init = signal ? { signal } : {};
	const { text, url } = await fetchPlaylist(src, init);
	if (!isMultivariantPlaylist(text)) return parseStreamInfo(text);
	const mediaUrl = resolveFirstMediaPlaylistUrl(text, url);
	if (!mediaUrl) throw new Error("No media playlist URL found in multivariant playlist");
	return parseStreamInfo((await fetchPlaylist(mediaUrl, init)).text);
}

//#endregion
//#region ../core/dist/dev/dom/media/native-hls/live.js
function NativeHlsMediaLiveMixin(BaseClass) {
	class NativeHlsMediaLive extends BaseClass {
		#targetLiveWindow = NaN;
		#liveEdgeStartOffset;
		#disconnect = null;
		#currentSrc = "";
		get targetLiveWindow() {
			return this.#targetLiveWindow;
		}
		get liveEdgeStart() {
			if (this.#liveEdgeStartOffset === void 0) return NaN;
			const target = this.target;
			if (!target) return NaN;
			const { seekable, buffered } = target;
			const ranges = seekable.length ? seekable : buffered;
			if (!ranges.length) return NaN;
			return ranges.end(ranges.length - 1) - this.#liveEdgeStartOffset;
		}
		attach(target) {
			super.attach?.(target);
			this.#init(target);
		}
		detach() {
			this.#destroy();
			super.detach?.();
		}
		destroy() {
			this.#destroy();
			super.destroy?.();
		}
		#destroy() {
			this.#disconnect?.abort();
			this.#disconnect = null;
			this.#currentSrc = "";
			this.#liveEdgeStartOffset = void 0;
			this.#setTargetLiveWindow(NaN);
		}
		#init(target) {
			this.#destroy();
			this.#disconnect = new AbortController();
			const { signal } = this.#disconnect;
			target.addEventListener("loadstart", () => this.#refresh(target), { signal });
			target.addEventListener("emptied", () => {
				this.#currentSrc = "";
				this.#liveEdgeStartOffset = void 0;
				this.#setTargetLiveWindow(NaN);
			}, { signal });
			if (target.currentSrc || target.src) this.#refresh(target);
		}
		async #refresh(target) {
			const src = target.currentSrc || target.src;
			if (!src || !looksLikeM3u8(src) || src === this.#currentSrc) return;
			this.#currentSrc = src;
			this.#liveEdgeStartOffset = void 0;
			this.#setTargetLiveWindow(NaN);
			const signal = this.#disconnect?.signal;
			try {
				const info = await getStreamInfoFromSrc(src, signal);
				if (signal?.aborted) return;
				if ((target.currentSrc || target.src) !== src) return;
				this.#liveEdgeStartOffset = info.liveEdgeStartOffset;
				this.#setTargetLiveWindow(info.targetLiveWindow);
			} catch {}
		}
		#setTargetLiveWindow(value) {
			if (Object.is(this.#targetLiveWindow, value)) return;
			this.#targetLiveWindow = value;
			this.dispatchEvent(new Event("targetlivewindowchange"));
		}
	}
	return NativeHlsMediaLive;
}

//#endregion
//#region ../core/dist/dev/dom/media/native-hls/stream-type.js
function NativeHlsMediaStreamTypeMixin(BaseClass) {
	class NativeHlsMediaStreamType extends BaseClass {
		#streamType = MediaStreamTypes.UNKNOWN;
		#isUserStreamType = false;
		#disconnect = null;
		get streamType() {
			return this.#streamType;
		}
		set streamType(value) {
			if (value === MediaStreamTypes.UNKNOWN) {
				this.#isUserStreamType = false;
				this.#setDetected(this.#detect());
				return;
			}
			this.#isUserStreamType = true;
			this.#update(value);
		}
		attach(target) {
			super.attach?.(target);
			this.#init(target);
		}
		detach() {
			this.#destroy();
			this.#setDetected(MediaStreamTypes.UNKNOWN);
			super.detach?.();
		}
		destroy() {
			this.#destroy();
			super.destroy?.();
		}
		#destroy() {
			this.#disconnect?.abort();
			this.#disconnect = null;
		}
		#init(target) {
			this.#destroy();
			this.#disconnect = new AbortController();
			const { signal } = this.#disconnect;
			const detect = () => this.#setDetected(this.#detect(target));
			target.addEventListener("durationchange", detect, { signal });
			target.addEventListener("loadedmetadata", detect, { signal });
			target.addEventListener("emptied", () => this.#setDetected(MediaStreamTypes.UNKNOWN), { signal });
			detect();
		}
		#detect(target = this.target) {
			if (!target) return MediaStreamTypes.UNKNOWN;
			const { duration } = target;
			if (duration === Infinity) return MediaStreamTypes.LIVE;
			if (Number.isFinite(duration) && duration > 0) return MediaStreamTypes.ON_DEMAND;
			return MediaStreamTypes.UNKNOWN;
		}
		#setDetected(value) {
			if (this.#isUserStreamType) return;
			this.#update(value);
		}
		#update(value) {
			if (this.#streamType === value) return;
			this.#streamType = value;
			this.dispatchEvent(new Event("streamtypechange"));
		}
	}
	return NativeHlsMediaStreamType;
}

//#endregion
//#region ../core/dist/dev/dom/media/native-hls/index.js
const nativeHlsMediaDefaultProps = {
	src: "",
	preload: "metadata",
	streamType: MediaStreamTypes.UNKNOWN
};
var NativeHlsMediaBase = class extends HTMLVideoElementHost {
	#src = nativeHlsMediaDefaultProps.src;
	#preload = nativeHlsMediaDefaultProps.preload;
	get engine() {
		return null;
	}
	get src() {
		return this.#src;
	}
	set src(src) {
		this.#src = src;
		if (this.target) this.target.src = src;
	}
	get preload() {
		return this.#preload;
	}
	set preload(value) {
		this.#preload = value;
		if (this.target) this.target.preload = value;
	}
	attach(target) {
		super.attach(target);
		if (this.preload !== target.preload) target.preload = this.preload;
		if (this.src) target.src = this.src;
	}
	detach() {
		super.detach();
	}
	destroy() {
		this.detach();
	}
};
var NativeHlsMedia = class extends NativeHlsMediaLiveMixin(NativeHlsMediaStreamTypeMixin(NativeHlsMediaErrorsMixin(NativeHlsMediaBase))) {};

//#endregion
export { MediaError as n, NativeHlsMedia as t };
