// @videojs/html/cdn/playback-rate-button-element-DN78puhg.js
import { N as defaults, O as selectPlaybackRate, _ as MediaButtonElement } from "./compounds-DxdDY3jr.js";
import { E as createState, n as PlayerController } from "./create-player-_WED6HtK.js";
import { l as playerContext, t as isFunction } from "./predicate-Dqd8wbW6.js";

//#region ../core/dist/dev/core/ui/playback-rate-button/playback-rate-button-core.js
var PlaybackRateButtonCore = class PlaybackRateButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	state = createState({
		rate: 1,
		label: ""
	});
	#props = { ...PlaybackRateButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, PlaybackRateButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return `Playback rate ${state.rate}`;
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		this.state.patch({ rate: media.playbackRate });
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	cycle(media) {
		if (this.#props.disabled) return;
		const { playbackRates, playbackRate } = media;
		if (playbackRates.length === 0) return;
		const idx = playbackRates.indexOf(playbackRate);
		const next = idx === -1 ? playbackRates.find((r) => r > playbackRate) ?? playbackRates[0] : playbackRates[(idx + 1) % playbackRates.length];
		media.setPlaybackRate(next);
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/playback-rate-button/playback-rate-button-data-attrs.js
const PlaybackRateButtonDataAttrs = { rate: "data-rate" };

//#endregion
//#region src/ui/playback-rate-button/playback-rate-button-element.ts
var PlaybackRateButtonElement = class extends MediaButtonElement {
	constructor(..._args) {
		super(..._args);
		this.core = new PlaybackRateButtonCore();
		this.stateAttrMap = PlaybackRateButtonDataAttrs;
		this.mediaState = new PlayerController(this, playerContext, selectPlaybackRate);
	}
	static {
		this.tagName = "media-playback-rate-button";
	}
	activate(state) {
		this.core.cycle(state);
	}
};

//#endregion
export { PlaybackRateButtonElement as t };
