// @videojs/html/cdn/poster-element-IdY3MYJj.js
import { A as selectTextTrack, D as selectPlayback, E as selectPiP, N as defaults, _ as MediaButtonElement, k as selectRemotePlayback, v as applyStateDataAttrs, w as selectFullscreen, y as logMissingFeature } from "./compounds-DxdDY3jr.js";
import { E as createState, j as MediaElement, n as PlayerController } from "./create-player-_WED6HtK.js";
import { l as playerContext, t as isFunction } from "./predicate-Dqd8wbW6.js";

//#region ../core/dist/dev/core/ui/buffering-indicator/buffering-indicator-core.js
var BufferingIndicatorCore = class BufferingIndicatorCore {
	static defaultProps = { delay: 500 };
	state = createState({ visible: false });
	#props = { ...BufferingIndicatorCore.defaultProps };
	#timer = null;
	setProps(props) {
		this.#props = defaults(props, BufferingIndicatorCore.defaultProps);
	}
	destroy() {
		this.#clearTimer();
	}
	update(media) {
		const buffering = media.waiting && !media.paused;
		if (buffering && !this.state.current.visible && !this.#timer) this.#timer = setTimeout(() => {
			this.#timer = null;
			this.state.patch({ visible: true });
		}, this.#props.delay);
		else if (!buffering) {
			this.#clearTimer();
			this.state.patch({ visible: false });
		}
	}
	#clearTimer() {
		if (this.#timer !== null) {
			clearTimeout(this.#timer);
			this.#timer = null;
		}
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/buffering-indicator/buffering-indicator-data-attrs.js
const BufferingIndicatorDataAttrs = { visible: "data-visible" };

//#endregion
//#region ../core/dist/dev/core/ui/captions-button/captions-button-core.js
var CaptionsButtonCore = class CaptionsButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	state = createState({
		subtitlesShowing: false,
		availability: "unavailable",
		label: ""
	});
	#props = { ...CaptionsButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, CaptionsButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return state.subtitlesShowing ? "Disable captions" : "Enable captions";
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		const availability = media.textTrackList.some((t) => t.kind === "captions" || t.kind === "subtitles") ? "available" : "unavailable";
		this.state.patch({
			subtitlesShowing: media.subtitlesShowing,
			availability
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	toggle(media) {
		if (this.#props.disabled) return;
		media.toggleSubtitles();
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/captions-button/captions-button-data-attrs.js
const CaptionsButtonDataAttrs = {
	subtitlesShowing: "data-active",
	availability: "data-availability"
};

//#endregion
//#region ../core/dist/dev/core/ui/cast-button/cast-button-core.js
var CastButtonCore = class CastButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	state = createState({
		castState: "disconnected",
		availability: "unsupported",
		label: ""
	});
	#props = { ...CastButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, CastButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		if (state.castState === "connected") return "Stop casting";
		if (state.castState === "connecting") return "Connecting";
		return "Start casting";
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		const castSupported = !!globalThis.chrome;
		this.state.patch({
			castState: media.remotePlaybackState,
			availability: castSupported ? media.remotePlaybackAvailability : "unsupported"
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	async toggle(media) {
		if (this.#props.disabled) return;
		if (media.remotePlaybackAvailability !== "available") return;
		try {
			await media.toggleRemotePlayback();
		} catch {}
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/cast-button/cast-button-data-attrs.js
const CastButtonDataAttrs = {
	castState: "data-cast-state",
	availability: "data-availability"
};

//#endregion
//#region ../core/dist/dev/core/ui/fullscreen-button/fullscreen-button-core.js
var FullscreenButtonCore = class FullscreenButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	state = createState({
		fullscreen: false,
		availability: "available",
		label: ""
	});
	#props = { ...FullscreenButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, FullscreenButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return state.fullscreen ? "Exit fullscreen" : "Enter fullscreen";
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		this.state.patch({
			fullscreen: media.fullscreen,
			availability: media.fullscreenAvailability
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	async toggle(media) {
		if (this.#props.disabled) return;
		if (media.fullscreenAvailability !== "available") return;
		try {
			if (media.fullscreen) await media.exitFullscreen();
			else await media.requestFullscreen();
		} catch {}
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/fullscreen-button/fullscreen-button-data-attrs.js
const FullscreenButtonDataAttrs = {
	fullscreen: "data-fullscreen",
	availability: "data-availability"
};

//#endregion
//#region ../core/dist/dev/core/ui/pip-button/pip-button-core.js
var PiPButtonCore = class PiPButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	state = createState({
		pip: false,
		availability: "available",
		label: ""
	});
	#props = { ...PiPButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, PiPButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return state.pip ? "Exit picture-in-picture" : "Enter picture-in-picture";
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		this.state.patch({
			pip: media.pip,
			availability: media.pipAvailability
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	async toggle(media) {
		if (this.#props.disabled) return;
		if (media.pipAvailability !== "available") return;
		try {
			if (media.pip) await media.exitPictureInPicture();
			else await media.requestPictureInPicture();
		} catch {}
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/pip-button/pip-button-data-attrs.js
const PiPButtonDataAttrs = {
	pip: "data-pip",
	availability: "data-availability"
};

//#endregion
//#region ../core/dist/dev/core/ui/poster/poster-core.js
var PosterCore = class {
	#media = null;
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		return { visible: !this.#media.started };
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/poster/poster-data-attrs.js
const PosterDataAttrs = { visible: "data-visible" };

//#endregion
//#region src/ui/buffering-indicator/buffering-indicator-element.ts
var BufferingIndicatorElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.delay = BufferingIndicatorCore.defaultProps.delay;
		this.#core = new BufferingIndicatorCore();
		this.#state = new PlayerController(this, playerContext, selectPlayback);
		this.#disconnect = null;
	}
	static {
		this.tagName = "media-buffering-indicator";
	}
	static {
		this.properties = { delay: { type: Number } };
	}
	#core;
	#state;
	#disconnect;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		this.#disconnect = new AbortController();
		this.#core.state.subscribe(() => this.requestUpdate(), { signal: this.#disconnect.signal });
		if (!this.#state.value) logMissingFeature(this.localName, this.#state.displayName);
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#disconnect?.abort();
		this.#disconnect = null;
	}
	willUpdate(changed) {
		super.willUpdate(changed);
		this.#core.setProps(this);
	}
	update(changed) {
		super.update(changed);
		const media = this.#state.value;
		if (!media) return;
		this.#core.update(media);
		applyStateDataAttrs(this, this.#core.state.current, BufferingIndicatorDataAttrs);
	}
};

//#endregion
//#region src/ui/captions-button/captions-button-element.ts
var CaptionsButtonElement = class extends MediaButtonElement {
	constructor(..._args) {
		super(..._args);
		this.core = new CaptionsButtonCore();
		this.stateAttrMap = CaptionsButtonDataAttrs;
		this.mediaState = new PlayerController(this, playerContext, selectTextTrack);
		this.hotkeyAction = "toggleSubtitles";
	}
	static {
		this.tagName = "media-captions-button";
	}
	activate(state) {
		this.core.toggle(state);
	}
};

//#endregion
//#region src/ui/cast-button/cast-button-element.ts
var CastButtonElement = class extends MediaButtonElement {
	constructor(..._args) {
		super(..._args);
		this.core = new CastButtonCore();
		this.stateAttrMap = CastButtonDataAttrs;
		this.mediaState = new PlayerController(this, playerContext, selectRemotePlayback);
	}
	static {
		this.tagName = "media-cast-button";
	}
	activate(state) {
		this.core.toggle(state);
	}
};

//#endregion
//#region src/ui/fullscreen-button/fullscreen-button-element.ts
var FullscreenButtonElement = class extends MediaButtonElement {
	constructor(..._args) {
		super(..._args);
		this.core = new FullscreenButtonCore();
		this.stateAttrMap = FullscreenButtonDataAttrs;
		this.mediaState = new PlayerController(this, playerContext, selectFullscreen);
		this.hotkeyAction = "toggleFullscreen";
	}
	static {
		this.tagName = "media-fullscreen-button";
	}
	activate(state) {
		this.core.toggle(state);
	}
};

//#endregion
//#region src/ui/pip-button/pip-button-element.ts
var PiPButtonElement = class extends MediaButtonElement {
	constructor(..._args) {
		super(..._args);
		this.core = new PiPButtonCore();
		this.stateAttrMap = PiPButtonDataAttrs;
		this.mediaState = new PlayerController(this, playerContext, selectPiP);
		this.hotkeyAction = "togglePictureInPicture";
	}
	static {
		this.tagName = "media-pip-button";
	}
	activate(state) {
		this.core.toggle(state);
	}
};

//#endregion
//#region src/ui/media-ui-element.ts
/** Abstract base for HTML custom elements that display media state with data attributes. */
var MediaUIElement = class extends MediaElement {
	connectedCallback() {
		super.connectedCallback();
		if (!this.mediaState.value && this.mediaState.displayName) logMissingFeature(this.localName, this.mediaState.displayName);
	}
	update(changed) {
		super.update(changed);
		const media = this.mediaState.value;
		if (!media) return;
		this.core.setMedia(media);
		const state = this.core.getState();
		applyStateDataAttrs(this, state, this.stateAttrMap);
	}
};

//#endregion
//#region src/ui/poster/poster-element.ts
var PosterElement = class extends MediaUIElement {
	constructor(..._args) {
		super(..._args);
		this.core = new PosterCore();
		this.stateAttrMap = PosterDataAttrs;
		this.mediaState = new PlayerController(this, playerContext, selectPlayback);
	}
	static {
		this.tagName = "media-poster";
	}
};

//#endregion
export { CaptionsButtonElement as a, CastButtonElement as i, PiPButtonElement as n, BufferingIndicatorElement as o, FullscreenButtonElement as r, PosterElement as t };
