// @videojs/html/cdn/seek-button-element-BkC8xrVR.js
import { N as defaults, O as selectPlaybackRate, _ as MediaButtonElement, b as applyElementProps, d as MenuRadioItemElement, f as MenuRadioGroupElement, j as selectTime, m as MenuElement, p as MenuItemIndicatorElement, v as applyStateDataAttrs, y as logMissingFeature } from "./compounds-DxdDY3jr.js";
import { E as createState, j as MediaElement, n as PlayerController } from "./create-player-_WED6HtK.js";
import { l as playerContext, o as isUndefined, t as isFunction } from "./predicate-Dqd8wbW6.js";

//#region ../core/dist/dev/core/ui/playback-rate-menu/playback-rate-menu-core.js
function formatPlaybackRate(rate) {
	return `${rate}×`;
}
var PlaybackRateMenuCore = class PlaybackRateMenuCore {
	static defaultProps = {
		label: "",
		formatRate: formatPlaybackRate,
		disabled: false
	};
	state = createState({
		rate: 1,
		rates: [],
		disabled: false,
		label: ""
	});
	#props = { ...PlaybackRateMenuCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, PlaybackRateMenuCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return `Playback rate ${state.rate}`;
	}
	getRateLabel(rate) {
		return this.#props.formatRate(rate);
	}
	getRateValue(rate) {
		return String(rate);
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": state.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		this.state.patch({
			rate: media.playbackRate,
			rates: media.playbackRates,
			disabled: this.#props.disabled || media.playbackRates.length === 0
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	select(media, rate) {
		if (this.#props.disabled) return;
		if (!media.playbackRates.includes(rate)) return;
		media.setPlaybackRate(rate);
	}
	selectValue(media, value) {
		const rate = media.playbackRates.find((candidate) => this.getRateValue(candidate) === value);
		if (isUndefined(rate)) return;
		this.select(media, rate);
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/playback-rate-menu/playback-rate-menu-data-attrs.js
const PlaybackRateMenuDataAttrs = {
	rate: "data-rate",
	disabled: "data-disabled"
};

//#endregion
//#region ../core/dist/dev/core/ui/seek-button/seek-button-core.js
var SeekButtonCore = class SeekButtonCore {
	static defaultProps = {
		seconds: 30,
		label: "",
		disabled: false
	};
	state = createState({
		seeking: false,
		direction: "forward",
		label: ""
	});
	#props = { ...SeekButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, SeekButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		const abs = Math.abs(this.#props.seconds);
		return state.direction === "backward" ? `Seek backward ${abs} seconds` : `Seek forward ${abs} seconds`;
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		const direction = this.#props.seconds < 0 ? "backward" : "forward";
		this.state.patch({
			seeking: media.seeking,
			direction
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	async seek(media) {
		if (this.#props.disabled) return;
		await media.seek(media.currentTime + this.#props.seconds);
	}
};

//#endregion
//#region ../core/dist/dev/core/ui/seek-button/seek-button-data-attrs.js
const SeekButtonDataAttrs = {
	seeking: "data-seeking",
	direction: "data-direction"
};

//#endregion
//#region src/ui/playback-rate-menu/playback-rate-menu-element.ts
var PlaybackRateMenuElement = class extends MenuElement {
	constructor(..._args) {
		super(..._args);
		this.label = "";
		this.disabled = false;
		this.align = "center";
		this.formatRate = PlaybackRateMenuCore.defaultProps.formatRate;
		this.#core = new PlaybackRateMenuCore();
		this.#mediaState = new PlayerController(this, playerContext, selectPlaybackRate);
	}
	static {
		this.tagName = "media-playback-rate-menu";
	}
	static {
		this.properties = {
			...MenuElement.properties,
			label: { type: String },
			disabled: { type: Boolean }
		};
	}
	#core;
	#mediaState;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		if (!this.#mediaState.value && this.#mediaState.displayName) logMissingFeature(this.localName, this.#mediaState.displayName);
	}
	update(changed) {
		super.update(changed);
		const media = this.#mediaState.value;
		if (!media) return;
		this.#core.setProps(this);
		this.#core.setMedia(media);
		const state = this.#core.getState();
		applyElementProps(this, this.#core.getAttrs(state));
		applyStateDataAttrs(this, state, PlaybackRateMenuDataAttrs);
	}
};

//#endregion
//#region src/ui/playback-rate-menu/playback-rate-menu-trigger-element.ts
var PlaybackRateMenuTriggerElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.label = "";
		this.disabled = false;
		this.commandfor = void 0;
		this.formatRate = PlaybackRateMenuCore.defaultProps.formatRate;
		this.#core = new PlaybackRateMenuCore();
		this.#mediaState = new PlayerController(this, playerContext, selectPlaybackRate);
		this.#disconnect = null;
		this.#handleClick = (event) => {
			if (this.#mediaState.value && !this.#core.state.current.disabled) return;
			event.preventDefault();
			event.stopImmediatePropagation();
		};
		this.#handleKeyDown = (event) => {
			if (event.target !== event.currentTarget) return;
			if (!this.#mediaState.value || this.#core.state.current.disabled) {
				if (event.key !== "Tab") event.preventDefault();
				return;
			}
			if (event.key === "Enter" || event.key === " ") {
				event.preventDefault();
				this.click();
			}
		};
	}
	static {
		this.tagName = "media-playback-rate-menu-trigger";
	}
	static {
		this.properties = {
			label: { type: String },
			disabled: { type: Boolean },
			commandfor: { type: String }
		};
	}
	#core;
	#mediaState;
	#disconnect;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		this.#disconnect = new AbortController();
		applyElementProps(this, {
			onClick: this.#handleClick,
			onKeyDown: this.#handleKeyDown
		}, { signal: this.#disconnect.signal });
		if (!this.#mediaState.value && this.#mediaState.displayName) logMissingFeature(this.localName, this.#mediaState.displayName);
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#disconnect?.abort();
		this.#disconnect = null;
	}
	/** Returns the trigger's current label derived from media state. */
	getLabel() {
		return this.#core.state.current.label || void 0;
	}
	update(changed) {
		super.update(changed);
		const media = this.#mediaState.value;
		if (!media) return;
		this.#core.setProps(this);
		this.#core.setMedia(media);
		const state = this.#core.getState();
		applyElementProps(this, {
			role: "button",
			tabIndex: 0,
			...this.#core.getAttrs(state)
		});
		applyStateDataAttrs(this, state, PlaybackRateMenuDataAttrs);
	}
	#handleClick;
	#handleKeyDown;
};

//#endregion
//#region src/ui/playback-rate-menu/playback-rate-options-element.ts
var PlaybackRateOptionsElement = class extends MenuRadioGroupElement {
	constructor(..._args) {
		super(..._args);
		this.disabled = false;
		this.formatRate = PlaybackRateMenuCore.defaultProps.formatRate;
		this.#core = new PlaybackRateMenuCore();
		this.#mediaState = new PlayerController(this, playerContext, selectPlaybackRate);
		this.#ratesKey = "";
		this.#disconnect = null;
		this.#handleValueChange = (event) => {
			if (event.target !== this) return;
			const media = this.#mediaState.value;
			if (!media) return;
			const { value } = event.detail;
			this.#core.selectValue(media, value);
		};
	}
	static {
		this.tagName = "media-playback-rate-options";
	}
	static {
		this.properties = {
			...MenuRadioGroupElement.properties,
			disabled: { type: Boolean }
		};
	}
	#core;
	#mediaState;
	#ratesKey;
	#disconnect;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		this.#disconnect = new AbortController();
		this.addEventListener("value-change", this.#handleValueChange, { signal: this.#disconnect.signal });
		if (!this.#mediaState.value && this.#mediaState.displayName) logMissingFeature(this.localName, this.#mediaState.displayName);
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#disconnect?.abort();
		this.#disconnect = null;
	}
	update(changed) {
		const media = this.#mediaState.value;
		let state = null;
		if (media) {
			this.#core.setProps({
				formatRate: this.formatRate,
				disabled: this.disabled
			});
			this.#core.setMedia(media);
			state = this.#core.getState();
			this.value = this.#core.getRateValue(state.rate);
			this.label = this.label || "Playback rate";
			this.#syncContent(state);
		}
		super.update(changed);
		if (state) applyStateDataAttrs(this, state, PlaybackRateMenuDataAttrs);
	}
	#syncContent(state) {
		const template = this.#getTemplate();
		const templateKey = template?.innerHTML ?? "";
		const ratesKey = `${state.rates.join("|")}::${templateKey}`;
		if (ratesKey !== this.#ratesKey) {
			this.#ratesKey = ratesKey;
			for (const child of [...this.children]) {
				if (child instanceof HTMLTemplateElement) continue;
				child.remove();
			}
			this.append(...state.rates.map((rate) => this.#createItem(rate, template)));
		}
		for (const item of this.querySelectorAll(MenuRadioItemElement.tagName)) {
			const checked = item.value === this.value;
			item.disabled = state.disabled;
			for (const indicator of item.querySelectorAll(MenuItemIndicatorElement.tagName)) indicator.checked = checked;
		}
	}
	#createItem(rate, template) {
		const item = this.#createItemFromTemplate(template);
		const value = this.#core.getRateValue(rate);
		item.value = value;
		item.setAttribute("data-rate", value);
		this.#setLabel(item, this.#core.getRateLabel(rate));
		return item;
	}
	#createItemFromTemplate(template) {
		if (!template) return document.createElement(MenuRadioItemElement.tagName);
		const root = template.content.cloneNode(true).firstElementChild;
		if (!root || root.localName !== MenuRadioItemElement.tagName || root.nextElementSibling) return document.createElement(MenuRadioItemElement.tagName);
		return root;
	}
	#setLabel(item, label) {
		const labelPart = item.querySelector("[data-part~=\"label\"]");
		if (labelPart) labelPart.textContent = label;
		else item.textContent = label;
	}
	#getTemplate() {
		for (const child of this.children) if (child instanceof HTMLTemplateElement) return child;
		return null;
	}
	#handleValueChange;
};

//#endregion
//#region src/ui/seek-button/seek-button-element.ts
var SeekButtonElement = class extends MediaButtonElement {
	constructor(..._args) {
		super(..._args);
		this.seconds = SeekButtonCore.defaultProps.seconds;
		this.core = new SeekButtonCore();
		this.stateAttrMap = SeekButtonDataAttrs;
		this.mediaState = new PlayerController(this, playerContext, selectTime);
	}
	static {
		this.tagName = "media-seek-button";
	}
	static {
		this.properties = {
			...MediaButtonElement.properties,
			seconds: { type: Number }
		};
	}
	activate(state) {
		this.core.seek(state);
	}
};

//#endregion
export { PlaybackRateMenuElement as i, PlaybackRateOptionsElement as n, PlaybackRateMenuTriggerElement as r, SeekButtonElement as t };
