// @videojs/html/cdn/video-host-BU9t26fe.js
import { n as namedNodeMapToObject, r as serializeAttributes } from "./media-attach-mixin-C1KQy-y8.js";
import { n as kebabCase, t as pick } from "./pick-D-wCJ3iA.js";
import { t as isFunction } from "./predicate-Dqd8wbW6.js";

//#region ../utils/dist/object/omit.js
/**
* Creates a new object without the specified keys.
*
* @example
* const obj = { a: 1, b: 2, c: 3 };
* omit(obj, ['b']); // { a: 1, c: 3 }
*/
function omit(obj, keys) {
	const result = {};
	for (const key in obj) if (!keys.includes(key)) result[key] = obj[key];
	return result;
}

//#endregion
//#region ../core/dist/dev/dom/media/custom-media-element/index.js
/** CSS custom property names for video elements. */
const VideoCSSVars = {
	borderRadius: "--media-video-border-radius",
	objectFit: "--media-object-fit",
	objectPosition: "--media-object-position",
	captionTrackDuration: "--media-caption-track-duration",
	captionTrackDelay: "--media-caption-track-delay",
	captionTrackY: "--media-caption-track-y"
};
/** Helper function to generate the HTML template for video elements. */
function getVideoTemplateHTML(attrs) {
	return `
    <style>
      :host {
        display: contents;
      }

      video {
        display: block;
        width: 100%;
        height: 100%;
        border-radius: var(${VideoCSSVars.borderRadius});
        object-fit: var(${VideoCSSVars.objectFit}, contain);
        object-position: var(${VideoCSSVars.objectPosition}, center);
      }

      video::-webkit-media-text-track-container {
        transition: translate var(${VideoCSSVars.captionTrackDuration}, 0) ease-out;
        transition-delay: var(${VideoCSSVars.captionTrackDelay}, 0);
        translate: 0 var(${VideoCSSVars.captionTrackY}, 0);
        scale: 0.98;
        z-index: 1;
        font-family: inherit;
      }
    </style>
    <slot name="media">
      <video${serializeAttributes(attrs)}></video>
    </slot>
    <slot></slot>
  `;
}
/** Helper function to generate the HTML template for other elements. */
function getCommonTemplateHTML(tag) {
	return (attrs) => {
		return `
      <style>
        :host {
          display: inline-flex;
          line-height: 0;
          flex-direction: column;
          justify-content: end;
        }

        ${tag} {
          width: 100%;
        }
      </style>
      <slot name="media">
        <${tag}${serializeAttributes(attrs)}></${tag}>
      </slot>
      <slot></slot>
    `;
	};
}
const excludedProperties = [
	"attach",
	"detach",
	"destroy"
];
function CustomMediaElement(tag, MediaHost) {
	const mediaHostAttrToProp = /* @__PURE__ */ new Map();
	let isDefined = false;
	class CustomMedia extends (globalThis.HTMLElement ?? class {}) {
		static getTemplateHTML = tag.endsWith("video") ? getVideoTemplateHTML : getCommonTemplateHTML(tag);
		static shadowRootOptions = { mode: "open" };
		static properties = {
			autoPictureInPicture: { type: Boolean },
			autoplay: { type: Boolean },
			controls: { type: Boolean },
			controlsList: { type: String },
			crossOrigin: { type: String },
			defaultMuted: {
				type: Boolean,
				attribute: "muted"
			},
			disablePictureInPicture: { type: Boolean },
			disableRemotePlayback: { type: Boolean },
			loading: { type: String },
			loop: { type: Boolean },
			playsInline: { type: Boolean },
			poster: { type: String },
			preload: { type: String },
			src: { type: String }
		};
		static get observedAttributes() {
			CustomMedia.#define(this);
			return [...getAttrsFromProps(this.properties)];
		}
		static #define(ctor) {
			if (isDefined) return;
			isDefined = true;
			for (let proto = MediaHost.prototype; proto && proto !== Object.prototype; proto = Object.getPrototypeOf(proto)) for (const prop of Object.getOwnPropertyNames(proto)) {
				if (prop in CustomMedia.prototype || excludedProperties.includes(prop)) continue;
				const descriptor = Object.getOwnPropertyDescriptor(proto, prop);
				if (!descriptor) continue;
				const config = {
					enumerable: true,
					configurable: true
				};
				if (typeof descriptor.value === "function") config.value = function(...args) {
					return this.#mediaHost[prop](...args);
				};
				else if (descriptor.get) {
					config.get = function() {
						return this.#mediaHost[prop];
					};
					if (descriptor.set) {
						const attr = kebabCase(prop);
						if (ctor.observedAttributes.includes(attr)) {
							mediaHostAttrToProp.set(attr, prop);
							config.set = function(val) {
								if (val === true || val === false || val == null) this.toggleAttribute(attr, Boolean(val));
								else this.setAttribute(attr, String(val));
							};
						} else config.set = function(val) {
							this.#mediaHost[prop] = val;
						};
					}
				}
				Object.defineProperty(CustomMedia.prototype, prop, config);
			}
			const properties = ctor.properties;
			for (const [prop, { type, attribute }] of Object.entries(properties)) {
				if (prop in CustomMedia.prototype) continue;
				const attr = attribute ?? prop.toLowerCase();
				Object.defineProperty(CustomMedia.prototype, prop, {
					get: function() {
						return type === Boolean ? this.hasAttribute(attr) : this.getAttribute(attr);
					},
					set: function(val) {
						if (type === Boolean) this.toggleAttribute(attr, Boolean(val));
						else this.setAttribute(attr, val);
					},
					enumerable: true,
					configurable: true
				});
			}
		}
		#mediaHost;
		#bridgedEventTypes = /* @__PURE__ */ new Set();
		#childMap = /* @__PURE__ */ new Map();
		#childObserver;
		constructor() {
			super();
			if (!this.shadowRoot) {
				const ctor = this.constructor;
				this.attachShadow(ctor.shadowRootOptions);
				const allowedKeys = getAttrsFromProps(ctor.properties);
				const disallowedKeys = [...mediaHostAttrToProp.keys()];
				const attrs = omit(pick(namedNodeMapToObject(this.attributes), allowedKeys), disallowedKeys);
				if (tag && !attrs.part) attrs.part = tag;
				this.shadowRoot.innerHTML = ctor.getTemplateHTML(attrs);
			}
			this.#mediaHost = new MediaHost();
			this.#attachToTarget();
			this.#childObserver = new MutationObserver(this.#syncMediaChildAttribute.bind(this));
			this.shadowRoot.addEventListener("slotchange", () => {
				this.#attachToTarget();
				this.#syncMediaChildren();
			});
			this.#syncMediaChildren();
		}
		#attachToTarget() {
			const target = this.target;
			if (target === this.#mediaHost.target) return;
			if (this.#mediaHost.target) this.#mediaHost.detach();
			this.#mediaHost.attach(target);
		}
		get target() {
			return this.querySelector(":scope > [slot=media]") ?? this.querySelector(tag) ?? this.shadowRoot?.querySelector(tag) ?? null;
		}
		disconnectedCallback() {
			if (!this.hasAttribute("keep-alive")) this.#mediaHost.destroy();
		}
		addEventListener(type, listener, options) {
			super.addEventListener(type, listener, options);
			if (!this.#bridgedEventTypes.has(type)) {
				this.#bridgedEventTypes.add(type);
				this.#mediaHost.addEventListener(type, this.#bridgeEvent);
			}
		}
		removeEventListener(type, listener, options) {
			super.removeEventListener(type, listener, options);
		}
		#bridgeEvent = (event) => {
			if (!event.composed) this.dispatchEvent(new event.constructor(event.type, event));
		};
		attributeChangedCallback(attrName, oldValue, newValue) {
			const prop = mediaHostAttrToProp.get(attrName);
			if (prop) {
				if (oldValue !== newValue) {
					const valueType = typeof this.#mediaHost[prop];
					this.#mediaHost[prop] = valueType === "boolean" ? newValue !== null : valueType === "number" ? Number(newValue) : newValue ?? "";
				}
				return;
			}
			if (!CustomMedia.observedAttributes.includes(attrName) && this.constructor.observedAttributes.includes(attrName)) return;
			if (newValue === null) this.target?.removeAttribute(attrName);
			else if (this.target?.getAttribute(attrName) !== newValue) this.target?.setAttribute(attrName, newValue);
		}
		#syncMediaChildren() {
			const defaultSlot = this.shadowRoot?.querySelector("slot:not([name])");
			const mediaChildren = new Set(defaultSlot?.assignedElements({ flatten: true }).filter((el) => el.localName === "track" || el.localName === "source"));
			for (const [el, clone] of this.#childMap) if (!mediaChildren.has(el)) {
				clone.remove();
				this.#childMap.delete(el);
			}
			for (const el of mediaChildren) {
				let clone = this.#childMap.get(el);
				if (!clone) {
					clone = el.cloneNode();
					this.#childMap.set(el, clone);
					this.#childObserver?.observe(el, { attributes: true });
				}
				this.target?.append(clone);
				this.#enableDefaultTrack(clone);
			}
		}
		#syncMediaChildAttribute(mutations) {
			for (const mutation of mutations) if (mutation.type === "attributes") {
				const { target, attributeName } = mutation;
				const clone = this.#childMap.get(target);
				if (clone && attributeName) {
					clone.setAttribute(attributeName, target.getAttribute(attributeName) ?? "");
					this.#enableDefaultTrack(clone);
				}
			}
		}
		#enableDefaultTrack(trackEl) {
			if (trackEl && trackEl.localName === "track" && trackEl.default && (trackEl.kind === "chapters" || trackEl.kind === "metadata") && trackEl.track.mode === "disabled") trackEl.track.mode = "hidden";
		}
	}
	return CustomMedia;
}
function getAttrsFromProps(props) {
	return Object.keys(props).map((prop) => props[prop]?.attribute ?? prop.toLowerCase());
}

//#endregion
//#region ../core/dist/dev/dom/media/media-host.js
const EMPTY_TIME_RANGES = Object.freeze({
	length: 0,
	start() {
		return 0;
	},
	end() {
		return 0;
	}
});
var HTMLMediaElementHost = class extends EventTarget {
	#target = null;
	#types = /* @__PURE__ */ new Set();
	get target() {
		return this.#target;
	}
	attach(target) {
		if (!target || this.#target === target) return;
		this.#target = target;
		for (const type of this.#types) target.addEventListener(type, this.#forwardEvent);
	}
	detach() {
		if (!this.#target) return;
		for (const type of this.#types) this.#target.removeEventListener(type, this.#forwardEvent);
		this.#target = null;
	}
	querySelectorAll(selectors) {
		return this.target?.querySelectorAll(selectors) ?? [];
	}
	querySelector(selectors) {
		return this.target?.querySelector(selectors) ?? null;
	}
	addEventListener(type, listener, options) {
		if (!this.#types.has(type)) {
			this.#types.add(type);
			this.target?.addEventListener(type, this.#forwardEvent);
		}
		super.addEventListener(type, listener, options);
	}
	removeEventListener(type, listener, options) {
		super.removeEventListener(type, listener, options);
	}
	#forwardEvent = (event) => {
		this.dispatchEvent(new event.constructor(event.type, event));
	};
	get title() {
		return this.target?.title ?? "";
	}
	set title(value) {
		if (this.target) this.target.title = value;
	}
	get paused() {
		return this.target?.paused ?? true;
	}
	get ended() {
		return this.target?.ended ?? false;
	}
	get loop() {
		return this.target?.loop ?? false;
	}
	set loop(value) {
		if (this.target) this.target.loop = value;
	}
	play() {
		return this.target?.play() ?? Promise.reject();
	}
	pause() {
		this.target?.pause();
	}
	get currentTime() {
		return this.target?.currentTime ?? 0;
	}
	set currentTime(value) {
		if (this.target) this.target.currentTime = value;
	}
	get duration() {
		return this.target?.duration ?? NaN;
	}
	get seeking() {
		return this.target?.seeking ?? false;
	}
	get src() {
		return this.target?.src ?? "";
	}
	set src(value) {
		if (this.target) this.target.src = value;
	}
	get currentSrc() {
		return this.target?.currentSrc ?? "";
	}
	get readyState() {
		return this.target?.readyState ?? 0;
	}
	load() {
		this.target?.load();
	}
	get volume() {
		return this.target?.volume ?? 1;
	}
	set volume(value) {
		if (this.target) this.target.volume = value;
	}
	get muted() {
		return this.target?.muted ?? false;
	}
	set muted(value) {
		if (this.target) this.target.muted = value;
	}
	get playbackRate() {
		return this.target?.playbackRate ?? 1;
	}
	set playbackRate(value) {
		if (this.target) this.target.playbackRate = value;
	}
	get buffered() {
		return this.target?.buffered ?? EMPTY_TIME_RANGES;
	}
	get seekable() {
		return this.target?.seekable ?? EMPTY_TIME_RANGES;
	}
	get error() {
		return this.target?.error ?? null;
	}
	get textTracks() {
		return this.target?.textTracks ?? [];
	}
	get remote() {
		return this.target?.remote;
	}
	get disableRemotePlayback() {
		return this.target?.disableRemotePlayback ?? false;
	}
	set disableRemotePlayback(value) {
		if (this.target) this.target.disableRemotePlayback = value;
	}
};

//#endregion
//#region ../core/dist/dev/dom/media/video-host.js
var HTMLVideoElementHost = class extends HTMLMediaElementHost {
	get poster() {
		return this.target?.poster ?? "";
	}
	set poster(value) {
		if (this.target) this.target.poster = value;
	}
	get webkitPresentationMode() {
		return this.target?.webkitPresentationMode;
	}
	get webkitSetPresentationMode() {
		const target = this.target;
		const fn = target?.webkitSetPresentationMode;
		return isFunction(fn) ? fn.bind(target) : void 0;
	}
	get isPictureInPicture() {
		return !!this.target && globalThis.document?.pictureInPictureElement === this.target || this.webkitPresentationMode === "picture-in-picture";
	}
	get isFullscreen() {
		if (!this.target) return false;
		if (this.webkitPresentationMode === "fullscreen") return true;
		const doc = globalThis.document;
		return doc?.fullscreenElement === this.target || doc?.webkitFullscreenElement === this.target;
	}
	async requestPictureInPicture() {
		if (!this.target) return Promise.reject();
		return this.target.requestPictureInPicture();
	}
	async exitPictureInPicture() {
		if (!this.target) return Promise.reject();
		return globalThis.document?.exitPictureInPicture();
	}
	requestFullscreen() {
		if (!this.target) return Promise.reject();
		return this.target.requestFullscreen();
	}
	exitFullscreen() {
		if (!this.target) return Promise.reject();
		return globalThis.document?.exitFullscreen();
	}
};

//#endregion
export { CustomMediaElement as n, HTMLVideoElementHost as t };
