// @videojs/html/dist/default/ui/error-dialog/error-dialog-element.js
import { playerContext } from "../../player/context.js";
import { MediaElement } from "../media-element.js";
import { PlayerController } from "../../player/player-controller.js";
import { alertDialogContext } from "../alert-dialog/context.js";
import { applyElementProps, applyStateDataAttrs, createAlertDialog, createTransition, selectError } from "@videojs/core/dom";
import { ContextProvider } from "@videojs/element/context";
import { SnapshotController } from "@videojs/store/html";
import { AlertDialogDataAttrs, ErrorDialogCore } from "@videojs/core";
//#region src/ui/error-dialog/error-dialog-element.ts
const FALLBACK_MESSAGE = "An error occurred. Please try again.";
let idCounter = 0;
var ErrorDialogElement = class extends MediaElement {
	static {
		this.tagName = "media-error-dialog";
	}
	#core = new ErrorDialogCore();
	#provider = new ContextProvider(this, { context: alertDialogContext });
	#titleId = `vjs-error-dialog-title-${idCounter++}`;
	#descriptionId = `vjs-error-dialog-desc-${idCounter++}`;
	#errorState = new PlayerController(this, playerContext, selectError);
	#dialog = null;
	#snapshot = null;
	#lastErrorMessage = null;
	constructor() {
		super();
		this.#core.setTitleId(this.#titleId);
		this.#core.setDescriptionId(this.#descriptionId);
	}
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		this.#dialog = createAlertDialog({
			transition: createTransition(),
			onOpenChange: (nextOpen) => {
				if (!nextOpen) this.#errorState.value?.dismissError();
			}
		});
		this.#dialog.setElement(this);
		if (this.#snapshot) this.#snapshot.track(this.#dialog.input);
		else this.#snapshot = new SnapshotController(this, this.#dialog.input);
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#dialog?.destroy();
		this.#dialog = null;
	}
	willUpdate(_changed) {
		super.willUpdate(_changed);
		if (!this.#dialog) return;
		const errorState = this.#errorState.value;
		const hasError = Boolean(errorState?.error);
		const { active: isOpen } = this.#dialog.input.current;
		if (errorState?.error) {
			const message = errorState.error.message?.trim();
			this.#lastErrorMessage = message || null;
		}
		const desc = this.querySelector("media-alert-dialog-description");
		if (desc) desc.textContent = this.#lastErrorMessage ?? FALLBACK_MESSAGE;
		if (hasError && !isOpen) this.#dialog.open();
		else if (!hasError && isOpen) this.#dialog.close();
	}
	update(_changed) {
		super.update(_changed);
		if (!this.#dialog) return;
		const input = this.#dialog.input.current;
		this.#core.setInput(input);
		const state = this.#core.getState();
		applyElementProps(this, this.#core.getAttrs(state));
		applyStateDataAttrs(this, state, AlertDialogDataAttrs);
		this.#provider.setValue({
			state,
			stateAttrMap: AlertDialogDataAttrs,
			close: () => this.#dialog?.close()
		});
	}
};
//#endregion
export { ErrorDialogElement };
