// @videojs/core/dist/dev/dom/hotkey/coordinator.js
import { toAriaKeyShortcut } from "./aria.js";
import { matchesHotkeyEvent, parseHotkeyPattern } from "./hotkey.js";
import { isEditableTarget, isInteractiveActivation, listen } from "@videojs/utils/dom";
//#region src/dom/hotkey/coordinator.ts
var HotkeyCoordinator = class {
	#target;
	#bindings = [];
	#nextId = 0;
	#disconnect = null;
	#docDisconnect = null;
	/** Action name → bound keys. Controls query this to set `aria-keyshortcuts`. */
	#ariaRegistry = /* @__PURE__ */ new Map();
	#subscribers = /* @__PURE__ */ new Set();
	#destroyed = false;
	constructor(target) {
		this.#target = target;
	}
	subscribe(callback) {
		this.#subscribers.add(callback);
		return () => this.#subscribers.delete(callback);
	}
	add(options) {
		const parsed = parseHotkeyPattern(options.keys);
		const binding = {
			parsed,
			options,
			id: this.#nextId++
		};
		this.#bindings.push(binding);
		this.#sortBindings();
		if (options.action) this.#addToAriaRegistry(options.action, parsed);
		if (options.target === "document") this.#connectDocument();
		else this.#connect();
		let removed = false;
		return () => {
			if (removed) return;
			removed = true;
			const idx = this.#bindings.indexOf(binding);
			if (idx !== -1) this.#bindings.splice(idx, 1);
			if (options.action) this.#removeFromAriaRegistry(options.action, parsed);
			this.#maybeDisconnect();
		};
	}
	getAriaKeys(action) {
		const bindings = this.#ariaRegistry.get(action);
		if (!bindings?.length) return void 0;
		return toAriaKeyShortcut(bindings);
	}
	destroy() {
		if (this.#destroyed) return;
		this.#destroyed = true;
		this.#disconnect?.abort();
		this.#disconnect = null;
		this.#docDisconnect?.abort();
		this.#docDisconnect = null;
		this.#bindings = [];
		this.#ariaRegistry.clear();
	}
	#sortBindings() {
		this.#bindings.sort((a, b) => {
			const specDiff = b.parsed[0].modifiers.size - a.parsed[0].modifiers.size;
			if (specDiff !== 0) return specDiff;
			return a.id - b.id;
		});
	}
	#connect() {
		if (this.#disconnect) return;
		this.#disconnect = new AbortController();
		listen(this.#target, "keydown", this.#handleEvent, { signal: this.#disconnect.signal });
	}
	#connectDocument() {
		if (this.#docDisconnect) return;
		this.#docDisconnect = new AbortController();
		listen(document, "keydown", this.#handleEvent, { signal: this.#docDisconnect.signal });
	}
	#maybeDisconnect() {
		const hasPlayer = this.#bindings.some((b) => b.options.target !== "document");
		const hasDoc = this.#bindings.some((b) => b.options.target === "document");
		if (!hasPlayer) {
			this.#disconnect?.abort();
			this.#disconnect = null;
		}
		if (!hasDoc) {
			this.#docDisconnect?.abort();
			this.#docDisconnect = null;
		}
	}
	#handleEvent = (event) => {
		if (event.key === "Unidentified") return;
		if (isInteractiveActivation(event)) return;
		if (event.defaultPrevented) return;
		const editable = isEditableTarget(event);
		for (const binding of this.#bindings) {
			const { options, parsed } = binding;
			if (options.disabled) continue;
			if (event.repeat && options.repeatable === false) continue;
			if (options.target === "document" !== (event.currentTarget === document)) continue;
			for (const p of parsed) {
				if (!matchesHotkeyEvent(p, event)) continue;
				if (editable && p.modifiers.size === 0) continue;
				if (this.#subscribers.size > 0) {
					const activateEvent = {
						source: "hotkey",
						action: options.action,
						value: options.value,
						event
					};
					for (const cb of this.#subscribers) try {
						cb(activateEvent);
					} catch (error) {
						console.warn("[vjs-hotkey] subscribe callback threw:", error);
					}
				}
				event.preventDefault();
				options.onActivate(event, p.originalKey);
				return;
			}
		}
	};
	#addToAriaRegistry(action, bindings) {
		let existing = this.#ariaRegistry.get(action);
		if (!existing) {
			existing = [];
			this.#ariaRegistry.set(action, existing);
		}
		existing.push(...bindings);
	}
	#removeFromAriaRegistry(action, bindings) {
		const existing = this.#ariaRegistry.get(action);
		if (!existing) return;
		const filtered = existing.filter((b) => !bindings.includes(b));
		if (filtered.length === 0) this.#ariaRegistry.delete(action);
		else this.#ariaRegistry.set(action, filtered);
	}
};
//#endregion
export { HotkeyCoordinator };
