// @videojs/core/dist/default/dom/ui/menu/create-menu.js
import { MenuItemDataAttrs } from "../../../core/ui/menu/menu-item-data-attrs.js";
import { createPopover } from "../popover/popover.js";
import { createState } from "@videojs/store";
//#region src/dom/ui/menu/create-menu.ts
function isMenuNavigationKey(event) {
	const { key } = event;
	return key === "ArrowDown" || key === "ArrowUp" || key === "ArrowLeft" || key === "ArrowRight" || key === "Home" || key === "End" || key === "Enter" || key === " " || key === "Escape" || key.length === 1 && !event.ctrlKey && !event.altKey && !event.metaKey;
}
function getRootPositionOptions(side, align) {
	if (!side || !align) return null;
	return {
		side,
		align
	};
}
function completeMenuItemSelection(menu, parentMenu = null) {
	if (parentMenu) parentMenu.pop();
	else menu.close();
}
function createMenu(options) {
	const items = [];
	let highlightedItem = null;
	let triggerElement = null;
	let contentElement = null;
	let typeaheadBuffer = "";
	let typeaheadTimer = null;
	let openRafId = 0;
	let lastCloseReason = null;
	const navigationState = createState({
		stack: [],
		direction: "forward"
	});
	function push(menuId, triggerId) {
		const stack = navigationState.current.stack;
		if (stack[stack.length - 1]?.menuId === menuId) return;
		navigationState.patch({
			stack: [...stack, {
				menuId,
				triggerId
			}],
			direction: "forward"
		});
	}
	function pop() {
		const stack = navigationState.current.stack;
		if (stack.length === 0) return;
		navigationState.patch({
			stack: stack.slice(0, -1),
			direction: "back"
		});
	}
	function highlight(element, highlightOptions) {
		if (highlightedItem === element) return;
		if (highlightedItem) {
			highlightedItem.tabIndex = -1;
			highlightedItem.removeAttribute(MenuItemDataAttrs.highlighted);
		}
		highlightedItem = element;
		if (element) {
			element.tabIndex = 0;
			element.setAttribute(MenuItemDataAttrs.highlighted, "");
			if (highlightOptions?.focus !== false) if (highlightOptions?.preventScroll) element.focus({ preventScroll: true });
			else element.focus();
		}
		options.onHighlightChange?.(element);
	}
	function clearHighlight() {
		if (highlightedItem) {
			highlightedItem.tabIndex = -1;
			highlightedItem.removeAttribute(MenuItemDataAttrs.highlighted);
			highlightedItem = null;
			options.onHighlightChange?.(null);
		}
	}
	function highlightFirstItem(options) {
		highlight(items[0] ?? null, options);
	}
	function getInitialHighlightItem() {
		return items.find((item) => item.matches("[aria-checked=\"true\"], [aria-selected=\"true\"]")) ?? items[0] ?? null;
	}
	function clearTypeahead() {
		if (typeaheadTimer !== null) {
			clearTimeout(typeaheadTimer);
			typeaheadTimer = null;
		}
		typeaheadBuffer = "";
	}
	function scheduleInitialHighlight() {
		cancelAnimationFrame(openRafId);
		openRafId = requestAnimationFrame(() => {
			openRafId = 0;
			if (!popover.input.current.active || popover.input.current.status === "ending" || highlightedItem) return;
			highlight(getInitialHighlightItem());
		});
	}
	function handleTypeahead(char) {
		typeaheadBuffer = typeaheadBuffer.length === 1 && typeaheadBuffer.toLowerCase() === char.toLowerCase() ? char : typeaheadBuffer + char;
		if (typeaheadTimer !== null) clearTimeout(typeaheadTimer);
		typeaheadTimer = setTimeout(clearTypeahead, 500);
		const searchStart = (highlightedItem ? items.indexOf(highlightedItem) : -1) + 1;
		const candidates = [...items.slice(searchStart), ...items.slice(0, searchStart)];
		const needle = typeaheadBuffer.toLowerCase();
		const match = candidates.find((candidate) => {
			return (candidate.textContent?.trim().toLowerCase() ?? "").startsWith(needle);
		});
		if (match) highlight(match);
	}
	const popover = createPopover({
		transition: options.transition,
		onOpenChange(open, details) {
			lastCloseReason = open ? null : details.reason;
			options.onOpenChange(open, details);
			if (open) scheduleInitialHighlight();
			else {
				clearHighlight();
				clearTypeahead();
				navigationState.patch({
					stack: [],
					direction: "forward"
				});
			}
		},
		onOpenChangeComplete(open) {
			options.onOpenChangeComplete?.(open);
			if (!open && lastCloseReason !== "imperative-action" && lastCloseReason !== "group-open") triggerElement?.focus();
		},
		closeOnEscape: options.closeOnEscape,
		closeOnOutsideClick: options.closeOnOutsideClick,
		...options.group ? { group: options.group } : {}
	});
	const contentProps = {
		onFocusOut: popover.popupProps.onFocusOut,
		onKeyDown(event) {
			const { key } = event;
			if (key !== "Escape" && isMenuNavigationKey(event) && !event.defaultPrevented) event.preventDefault();
			if (items.length === 0) return;
			switch (key) {
				case "ArrowDown":
					event.preventDefault();
					highlight(items[((highlightedItem ? items.indexOf(highlightedItem) : -1) + 1) % items.length] ?? null);
					break;
				case "ArrowUp": {
					event.preventDefault();
					const currentIndex = highlightedItem ? items.indexOf(highlightedItem) : 0;
					highlight(items[(currentIndex <= 0 ? items.length : currentIndex) - 1] ?? null);
					break;
				}
				case "Home":
					event.preventDefault();
					highlight(items[0] ?? null);
					break;
				case "End":
					event.preventDefault();
					highlight(items[items.length - 1] ?? null);
					break;
				case "Enter":
				case " ":
					event.preventDefault();
					highlightedItem?.click();
					break;
				default: if (key.length === 1 && !event.ctrlKey && !event.altKey && !event.metaKey) handleTypeahead(key);
			}
		}
	};
	function handleTriggerKeyDown(event) {
		const input = popover.input.current;
		if (!input.active || input.status === "ending") return;
		if (event.key === "Escape") return;
		if (!isMenuNavigationKey(event)) return;
		contentProps.onKeyDown(event);
		event.stopPropagation();
	}
	function setTriggerElement(element) {
		triggerElement = element;
		popover.setTriggerElement(element);
	}
	function setContentElement(element) {
		contentElement = element;
		popover.setPopupElement(element);
	}
	function compareItems(a, b) {
		if (a === b) return 0;
		const position = a.compareDocumentPosition(b);
		if (position & Node.DOCUMENT_POSITION_FOLLOWING) return -1;
		if (position & Node.DOCUMENT_POSITION_PRECEDING) return 1;
		return 0;
	}
	function registerItem(element) {
		element.tabIndex = -1;
		element.setAttribute(MenuItemDataAttrs.item, "");
		items.push(element);
		items.sort(compareItems);
		if (popover.input.current.active && popover.input.current.status !== "ending" && !highlightedItem) scheduleInitialHighlight();
		return () => {
			const index = items.indexOf(element);
			if (index !== -1) items.splice(index, 1);
			if (highlightedItem === element) clearHighlight();
		};
	}
	function destroy() {
		cancelAnimationFrame(openRafId);
		openRafId = 0;
		clearTypeahead();
		popover.destroy();
	}
	return {
		input: popover.input,
		navigationInput: navigationState,
		triggerProps: {
			onClick: popover.triggerProps.onClick,
			onKeyDown: handleTriggerKeyDown
		},
		contentProps,
		get triggerElement() {
			return triggerElement;
		},
		get contentElement() {
			return contentElement;
		},
		setTriggerElement,
		setContentElement,
		registerItem,
		highlight,
		highlightFirstItem,
		push,
		pop,
		open: popover.open,
		close: popover.close,
		destroy
	};
}
//#endregion
export { completeMenuItemSelection, createMenu, getRootPositionOptions, isMenuNavigationKey };
