// @videojs/core/dist/dev/dom/media/native-hls/index.js
import { MediaStreamTypes } from "../../../core/media/types.js";
import { HTMLVideoElementHost } from "../video-host.js";
import { NativeHlsMediaErrorsMixin } from "./errors.js";
import { NativeHlsMediaLiveMixin } from "./live.js";
import { NativeHlsMediaStreamTypeMixin } from "./stream-type.js";
//#region src/dom/media/native-hls/index.ts
const StreamTypes = MediaStreamTypes;
const nativeHlsMediaDefaultProps = {
	src: "",
	preload: "metadata",
	streamType: MediaStreamTypes.UNKNOWN
};
var NativeHlsMediaBase = class extends HTMLVideoElementHost {
	#src = nativeHlsMediaDefaultProps.src;
	#preload = nativeHlsMediaDefaultProps.preload;
	get engine() {
		return null;
	}
	get src() {
		return this.#src;
	}
	set src(src) {
		this.#src = src;
		if (this.target) this.target.src = src;
	}
	get preload() {
		return this.#preload;
	}
	set preload(value) {
		this.#preload = value;
		if (this.target) this.target.preload = value;
	}
	attach(target) {
		super.attach(target);
		if (this.preload !== target.preload) target.preload = this.preload;
		if (this.src) target.src = this.src;
	}
	detach() {
		super.detach();
	}
	destroy() {
		this.detach();
	}
};
var NativeHlsMedia = class extends NativeHlsMediaLiveMixin(NativeHlsMediaStreamTypeMixin(NativeHlsMediaErrorsMixin(NativeHlsMediaBase))) {};
//#endregion
export { NativeHlsMedia, StreamTypes, nativeHlsMediaDefaultProps };
