// @videojs/core/dist/default/dom/media/hls/hlsjs.js
import { HTMLVideoElementHost } from "../video-host.js";
import { HlsJsMediaErrorsMixin } from "./errors.js";
import { HlsJsMediaLiveMixin } from "./live.js";
import { HlsJsMediaMetadataTracksMixin } from "./metadata-tracks.js";
import { HlsJsMediaPreloadMixin } from "./preload.js";
import { HlsJsMediaStreamTypeMixin } from "./stream-type.js";
import { HlsJsMediaTextTracksMixin } from "./text-tracks.js";
import Hls from "hls.js";
//#region src/dom/media/hls/hlsjs.ts
const defaultHlsConfig = {
	backBufferLength: 30,
	renderTextTracksNatively: false,
	liveDurationInfinity: true,
	capLevelToPlayerSize: true,
	capLevelOnFPSDrop: true,
	autoStartLoad: false
};
var HlsJsMediaBase = class extends HTMLVideoElementHost {
	#engine = null;
	constructor(params) {
		super();
		this.#engine = new Hls({
			...defaultHlsConfig,
			...params.config
		});
	}
	get engine() {
		return this.#engine;
	}
	get src() {
		return this.#engine?.url ?? "";
	}
	set src(src) {
		this.#engine?.loadSource(src);
	}
	attach(target) {
		super.attach(target);
		this.#engine?.attachMedia(target);
	}
	detach() {
		this.#engine?.detachMedia();
		super.detach();
	}
	destroy() {
		this.detach();
		this.#engine?.destroy();
		this.#engine = null;
	}
};
var HlsJsMedia = class extends HlsJsMediaPreloadMixin(HlsJsMediaLiveMixin(HlsJsMediaStreamTypeMixin(HlsJsMediaMetadataTracksMixin(HlsJsMediaTextTracksMixin(HlsJsMediaErrorsMixin(HlsJsMediaBase)))))) {};
//#endregion
export { HlsJsMedia };
