// @videojs/core/dist/default/core/ui/time-slider/time-slider-core.js
import { SliderCore } from "../slider/slider-core.js";
import { defaults } from "@videojs/utils/object";
import { formatTimeAsPhrase } from "@videojs/utils/time";
//#region src/core/ui/time-slider/time-slider-core.ts
/** Time-domain slider: maps media time/buffer state to slider state. */
var TimeSliderCore = class TimeSliderCore extends SliderCore {
	static defaultProps = {
		...SliderCore.defaultProps,
		label: "Seek",
		changeThrottle: 100
	};
	#props = { ...TimeSliderCore.defaultProps };
	#media = null;
	constructor(props) {
		super();
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, TimeSliderCore.defaultProps);
		super.setProps({
			...props,
			min: 0
		});
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const { duration, currentTime, seeking, buffered } = this.#media;
		super.setProps({
			...this.#props,
			min: 0,
			max: duration
		});
		const base = super.getSliderState(currentTime);
		const bufferedEnd = buffered.length > 0 ? buffered[buffered.length - 1][1] : 0;
		const bufferPercent = duration > 0 ? bufferedEnd / duration * 100 : 0;
		return {
			...base,
			currentTime,
			duration,
			seeking,
			bufferPercent
		};
	}
	getLabel(state) {
		return super.getLabel(state) || "Seek";
	}
	getAttrs(state) {
		const base = super.getAttrs(state);
		const announceValue = state.dragging ? this.rawValueFromPercent(state.pointerPercent) : state.value;
		const currentPhrase = formatTimeAsPhrase(announceValue);
		const durationPhrase = formatTimeAsPhrase(state.duration);
		const valuetext = durationPhrase ? `${currentPhrase} of ${durationPhrase}` : currentPhrase;
		return {
			...base,
			"aria-valuenow": announceValue,
			"aria-valuetext": valuetext
		};
	}
};
//#endregion
export { TimeSliderCore };
