// @videojs/spf/dist/default/playback/actors/dom/segment-loader.js
import { peek } from "../../../core/signals/primitives.js";
import { effect } from "../../../core/signals/effect.js";
import { SerialRunner, Task } from "../../../core/tasks/task.js";
import { createMachineActor } from "../../../core/actors/create-machine-actor.js";
import { DEFAULT_FORWARD_BUFFER_CONFIG, calculateForwardFlushPoint, getSegmentsToLoad } from "../../../media/buffer/forward-buffer.js";
import { SEGMENT_TIME_EPSILON } from "../../../media/types/index.js";
import { DEFAULT_BACK_BUFFER_CONFIG, calculateBackBufferFlushPoint } from "../../../media/buffer/back-buffer.js";
//#region src/playback/actors/dom/segment-loader.ts
/**
* Resolves when the SourceBufferActor snapshot reaches 'idle'.
* Rejects if the signal is aborted or the actor is destroyed.
*
* Used to sequence SourceBufferActor operations without awaiting send()
* directly — send() is fire-and-forget; callers observe completion via
* state transition.
*/
function waitForIdle(snapshot, signal) {
	return new Promise((resolve, reject) => {
		if (snapshot.get().value === "idle") {
			resolve();
			return;
		}
		if (snapshot.get().value === "destroyed") {
			reject(new DOMException("Aborted", "AbortError"));
			return;
		}
		if (signal.aborted) {
			reject(signal.reason);
			return;
		}
		let stop;
		const cleanup = (fn) => {
			stop?.();
			signal.removeEventListener("abort", onAbort);
			fn();
		};
		const onAbort = () => cleanup(() => reject(signal.reason));
		stop = effect(() => {
			const value = snapshot.get().value;
			if (value === "idle") cleanup(resolve);
			else if (value === "destroyed") cleanup(() => reject(new DOMException("Aborted", "AbortError")));
		});
		signal.addEventListener("abort", onAbort, { once: true });
	});
}
/**
* Wraps a LoadTask descriptor into a Task that fetches (if needed) and
* forwards to SourceBufferActor. Updates in-flight context around async
* operations so the loading handler can make accurate continue/preempt
* decisions at any point.
*/
function makeLoadTask(op, { getContext, setContext, fetchBytes, sourceBufferActor }) {
	return new Task(async (taskSignal) => {
		if (taskSignal.aborted) return;
		if (op.type === "remove") {
			sourceBufferActor.send(op);
			await waitForIdle(sourceBufferActor.snapshot, taskSignal);
			return;
		}
		if (op.type === "append-init") {
			setContext({
				...getContext(),
				inFlightInitTrackId: op.meta.trackId
			});
			try {
				const data = await fetchBytes(op, {
					signal: taskSignal,
					minChunkSize: Infinity
				});
				if (!taskSignal.aborted) {
					sourceBufferActor.send({
						type: "append-init",
						data,
						meta: op.meta
					});
					await waitForIdle(sourceBufferActor.snapshot, taskSignal);
				}
			} finally {
				setContext({
					...getContext(),
					inFlightInitTrackId: null
				});
			}
			return;
		}
		setContext({
			...getContext(),
			inFlightSegmentId: op.meta.id
		});
		try {
			const stream = await fetchBytes(op, { signal: taskSignal });
			if (!taskSignal.aborted) {
				sourceBufferActor.send({
					type: "append-segment",
					data: stream,
					meta: op.meta
				});
				await waitForIdle(sourceBufferActor.snapshot, taskSignal);
			}
		} finally {
			setContext({
				...getContext(),
				inFlightSegmentId: null
			});
		}
	});
}
/**
* Creates a SegmentLoaderActor for one track type (video or audio).
*
* Receives load assignments via `send()` and owns all execution: planning,
* removes, fetches, and appends. Coordinates with the SourceBufferActor for
* all physical SourceBuffer operations.
*
* Planning (Cases 1–3) happens in the `load` handler on every incoming
* message, producing an ordered LoadTask list. The runner drains that list
* sequentially via SerialRunner. When a new message arrives mid-run, the
* handler replans and either continues the in-flight operation (abortPending
* + schedule new remainder) or preempts it (abortAll + cancel SourceBuffer
* if needed + schedule new plan).
*
* @param sourceBufferActor - Shared SourceBufferActor reference (not owned)
* @param fetchBytes - Tracked fetch closure (owns throughput sampling for segments).
*   Accepts an optional `minChunkSize` in options; init segments pass `Infinity`
*   so the entire body accumulates as one chunk before appending.
*/
function createSegmentLoaderActor(sourceBufferActor, fetchBytes, config = {}) {
	const forwardBufferConfig = {
		...DEFAULT_FORWARD_BUFFER_CONFIG,
		...config.forwardBuffer
	};
	const backBufferConfig = {
		...DEFAULT_BACK_BUFFER_CONFIG,
		...config.backBuffer
	};
	const getBufferedSegments = (allSegments) => {
		const bufferedIds = new Set(peek(sourceBufferActor.snapshot).context.segments.filter((s) => !s.partial).map((s) => s.id));
		return allSegments.filter((s) => bufferedIds.has(s.id));
	};
	/**
	* Translate a load message into an ordered LoadTask list based on committed
	* actor state. In-flight awareness is handled separately in the load handler.
	*
	* @todo Rename alongside LoadTask (e.g. planOps).
	*
	* Case 1 — Removes: forward and back buffer flush points, segment-aligned.
	*   No flush on track switch: appending new content overwrites existing buffer
	*   ranges, and the actor's time-aligned deduplication keeps the segment model
	*   accurate as new segments arrive.
	*
	* Case 2 — Init: schedule if not yet committed for this track.
	*
	* Case 3 — Segments: all segments in the load window not yet committed.
	*/
	const planTasks = (message) => {
		const { track, range } = message;
		const actorCtx = peek(sourceBufferActor.snapshot).context;
		const bufferedSegments = getBufferedSegments(track.segments);
		const currentTime = range?.start ?? 0;
		const tasks = [];
		if (range) {
			const forwardFlushStart = calculateForwardFlushPoint(bufferedSegments, currentTime, forwardBufferConfig);
			if (forwardFlushStart < Infinity) tasks.push({
				type: "remove",
				start: forwardFlushStart,
				end: Infinity
			});
			const backFlushEnd = calculateBackBufferFlushPoint(bufferedSegments, currentTime, backBufferConfig);
			if (backFlushEnd > 0) tasks.push({
				type: "remove",
				start: 0,
				end: backFlushEnd
			});
		}
		if (actorCtx.initTrackId !== track.id) tasks.push({
			type: "append-init",
			meta: { trackId: track.id },
			url: track.initialization.url,
			...track.initialization.byteRange !== void 0 && { byteRange: track.initialization.byteRange }
		});
		if (range) {
			const segmentsToLoad = getSegmentsToLoad(track.segments, bufferedSegments, currentTime, forwardBufferConfig).filter((seg) => {
				const existing = actorCtx.segments.find((s) => Math.abs(s.startTime - seg.startTime) < SEGMENT_TIME_EPSILON);
				if (existing?.partial) return true;
				if (!existing?.trackBandwidth || !track.bandwidth) return true;
				return track.bandwidth > existing.trackBandwidth;
			});
			for (const segment of segmentsToLoad) tasks.push({
				type: "append-segment",
				meta: {
					id: segment.id,
					startTime: segment.startTime,
					duration: segment.duration,
					trackId: track.id,
					trackBandwidth: track.bandwidth
				},
				url: segment.url,
				...segment.byteRange !== void 0 && { byteRange: segment.byteRange }
			});
		}
		return tasks;
	};
	const scheduleAll = (tasks, { getContext, setContext, runner }) => {
		tasks.forEach((op) => {
			runner.schedule(makeLoadTask(op, {
				getContext,
				setContext,
				fetchBytes,
				sourceBufferActor
			})).then(void 0, (e) => {
				if (e instanceof Error && e.name === "AbortError") return;
				console.error("Unexpected error in segment loader:", e);
				runner.abortPending();
			});
		});
	};
	return createMachineActor({
		runner: () => new SerialRunner(),
		initial: "idle",
		context: {
			inFlightInitTrackId: null,
			inFlightSegmentId: null
		},
		states: {
			idle: { on: { load: (msg, ctx) => {
				const allTasks = planTasks(msg);
				if (allTasks.length === 0) return;
				ctx.transition("loading");
				scheduleAll(allTasks, ctx);
			} } },
			loading: {
				onSettled: "idle",
				on: { load: (msg, ctx) => {
					const { context, runner } = ctx;
					const allTasks = planTasks(msg);
					if (context.inFlightSegmentId !== null && allTasks.some((t) => t.type === "append-segment" && t.meta.id === context.inFlightSegmentId) || context.inFlightInitTrackId !== null && allTasks.some((t) => t.type === "append-init" && t.meta.trackId === context.inFlightInitTrackId)) {
						runner.abortPending();
						scheduleAll(allTasks.filter((t) => !(t.type === "append-segment" && t.meta.id === context.inFlightSegmentId) && !(t.type === "append-init" && t.meta.trackId === context.inFlightInitTrackId)), ctx);
					} else {
						runner.abortAll();
						if (context.inFlightSegmentId !== null || context.inFlightInitTrackId !== null && allTasks.some((t) => t.type === "append-init" && t.meta.trackId !== context.inFlightInitTrackId)) sourceBufferActor.send({ type: "cancel" });
						scheduleAll(allTasks, ctx);
					}
				} }
			}
		}
	});
}
//#endregion
export { createSegmentLoaderActor };
