// vidstack/dev/global/vidstack-player.js
import { defineCustomElement, isString, kebabToCamelCase } from "../chunks/vidstack-BwClSYSC.js";
import { isHTMLAudioElement, isHTMLIFrameElement, isHTMLVideoElement } from "../chunks/vidstack-BWkbKDUg.js";
import { isHTMLElement } from "../chunks/vidstack-CToJFZVs.js";
import "../define/vidstack-player.js";
//#region src/global/layouts/default.ts
var VidstackPlayerLayout = class {
	props;
	constructor(props) {
		this.props = props;
	}
	name = "vidstack";
	async load() {
		await import("../define/vidstack-player-default-layout.js");
		await import("../define/vidstack-player-ui.js");
	}
	create() {
		const layouts = [document.createElement("media-audio-layout"), document.createElement("media-video-layout")];
		if (this.props) for (const [prop, value] of Object.entries(this.props)) for (const el of layouts) el[prop] = value;
		return layouts;
	}
};
//#endregion
//#region src/global/layouts/plyr.ts
var PlyrLayout = class {
	props;
	constructor(props) {
		this.props = props;
	}
	name = "plyr";
	async load() {
		await import("../define/plyr-layout.js");
	}
	create() {
		const layout = document.createElement("media-plyr-layout");
		if (this.props) for (const [prop, value] of Object.entries(this.props)) layout[prop] = value;
		return [layout];
	}
};
//#endregion
//#region src/global/player.ts
const LAYOUT_LOADED = Symbol();
var VidstackPlayer = class {
	static async create({ target, layout, tracks, ...props }) {
		if (isString(target)) target = document.querySelector(target);
		if (!isHTMLElement(target)) throw Error(`[vidstack] target must be of type \`HTMLElement\`, found \`${typeof target}\``);
		let player = document.createElement("media-player"), provider = document.createElement("media-provider"), layouts, isTargetContainer = !isHTMLAudioElement(target) && !isHTMLVideoElement(target) && !isHTMLIFrameElement(target);
		player.setAttribute("keep-alive", "");
		if (props.poster && layout?.name !== "plyr") {
			if (!customElements.get("media-poster")) {
				const { MediaPosterElement } = await import("../chunks/vidstack-CTpPHQzF.js").then((n) => n.poster_element_exports);
				defineCustomElement(MediaPosterElement);
			}
			const poster = document.createElement("media-poster");
			if (layout?.name === "vidstack") poster.classList.add("vds-poster");
			provider.append(poster);
		}
		if (layout) {
			target.removeAttribute("controls");
			if (!layout[LAYOUT_LOADED]) {
				await layout.load();
				layout[LAYOUT_LOADED] = true;
			}
			layouts = await layout.create();
		}
		const title = target.getAttribute("title");
		if (title) player.setAttribute("title", title);
		const width = target.getAttribute("width"), height = target.getAttribute("height");
		if (width || height) {
			if (width) player.style.width = width;
			if (height) player.style.height = height;
			player.style.aspectRatio = "unset";
		}
		for (const attr of target.attributes) {
			const name = attr.name.replace("data-", ""), propName = kebabToCamelCase(name);
			if (propName in player) player.setAttribute(name, attr.value);
			else if (layouts?.length) {
				for (const layout of layouts) if (propName in layout) layout.setAttribute(name, attr.value);
			}
		}
		for (const [prop, value] of Object.entries(props)) player[prop] = value;
		if (tracks) for (const track of tracks) player.textTracks.add(track);
		player.append(provider);
		if (layouts) for (const layout of layouts) player.append(layout);
		if (isTargetContainer) target.append(player);
		else {
			for (const child of [...target.children]) provider.append(child);
			target.replaceWith(player);
		}
		return player;
	}
};
//#endregion
export { PlyrLayout, VidstackPlayer, VidstackPlayerLayout };
