// @videojs/core/dist/dev/core/ui/playback-rate-menu/playback-rate-menu-core.js
import { createState } from "@videojs/store";
import { defaults } from "@videojs/utils/object";
import { isFunction, isUndefined } from "@videojs/utils/predicate";
//#region src/core/ui/playback-rate-menu/playback-rate-menu-core.ts
function formatPlaybackRate(rate) {
	return `${rate}×`;
}
var PlaybackRateMenuCore = class PlaybackRateMenuCore {
	static defaultProps = {
		label: "",
		formatRate: formatPlaybackRate,
		disabled: false
	};
	state = createState({
		rate: 1,
		rates: [],
		disabled: false,
		label: ""
	});
	#props = { ...PlaybackRateMenuCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, PlaybackRateMenuCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		return `Playback rate ${state.rate}`;
	}
	getRateLabel(rate) {
		return this.#props.formatRate(rate);
	}
	getRateValue(rate) {
		return String(rate);
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": state.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		this.state.patch({
			rate: media.playbackRate,
			rates: media.playbackRates,
			disabled: this.#props.disabled || media.playbackRates.length === 0
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	select(media, rate) {
		if (this.#props.disabled) return;
		if (!media.playbackRates.includes(rate)) return;
		media.setPlaybackRate(rate);
	}
	selectValue(media, value) {
		const rate = media.playbackRates.find((candidate) => this.getRateValue(candidate) === value);
		if (isUndefined(rate)) return;
		this.select(media, rate);
	}
};
//#endregion
export { PlaybackRateMenuCore };
