// @videojs/html/dist/dev/ui/status-indicator/status-indicator-element.js
import { InputIndicatorElement } from "../input-indicators/input-indicator-element.js";
import { LiveIndicator } from "../input-indicators/live-indicator.js";
import { createTransition } from "@videojs/core/dom";
import { StatusIndicatorCore, StatusIndicatorDataAttrs, getStatusIndicatorDisplayValue } from "@videojs/core";
//#region src/ui/status-indicator/status-indicator-element.ts
var StatusIndicatorElement = class extends InputIndicatorElement {
	static {
		this.tagName = "media-status-indicator";
	}
	static {
		this.properties = {
			actions: { type: String },
			closeDelay: {
				type: Number,
				attribute: "close-delay"
			}
		};
	}
	#core = new StatusIndicatorCore();
	#transition = createTransition();
	#liveIndicator = new LiveIndicator({
		host: this,
		dataAttrs: StatusIndicatorDataAttrs,
		render: renderStatusIndicator
	});
	get core() {
		return this.#core;
	}
	get transition() {
		return this.#transition;
	}
	get liveIndicator() {
		return this.#liveIndicator;
	}
	syncCoreProps() {
		this.#core.setProps({
			actions: parseActions(this.actions),
			closeDelay: this.closeDelay
		});
	}
};
function parseActions(actions) {
	return actions?.split(/[\s,]+/).filter(Boolean);
}
function renderStatusIndicator(element, state) {
	const value = element.querySelector("media-status-indicator-value");
	if (!value) return;
	value.textContent = getStatusIndicatorDisplayValue(state);
}
//#endregion
export { StatusIndicatorElement };
