// vidstack/dev/chunks/vidstack-BdKB12NY2.js
import { listenEvent } from "./vidstack-BwClSYSC.js";
//#region src/providers/google-cast/utils.ts
function getCastFrameworkURL() {
	return "https://www.gstatic.com/cv/js/sender/v1/cast_sender.js?loadCastFramework=1";
}
/**
* Whether the Google Cast framework has loaded.
*/
function hasLoadedCastFramework() {
	return !!window.cast?.framework;
}
/**
* Whether Google Cast is available on this platform.
*/
function isCastAvailable() {
	return !!window.chrome?.cast?.isAvailable;
}
/**
* Whether the cast sender is connected.
*/
function isCastConnected() {
	return getCastContext().getCastState() === cast.framework.CastState.CONNECTED;
}
/**
* @see {@link https://developers.google.com/cast/docs/reference/web_sender/cast.framework.CastContext}
*/
function getCastContext() {
	return window.cast.framework.CastContext.getInstance();
}
/**
* @see {@link https://developers.google.com/cast/docs/reference/web_sender/cast.framework.CastSession}
*/
function getCastSession() {
	return getCastContext().getCurrentSession();
}
/**
* @see {@link https://developers.google.com/cast/docs/reference/web_sender/chrome.cast.media.Media}
*/
function getCastSessionMedia() {
	return getCastSession()?.getSessionObj().media[0];
}
function hasActiveCastSession(src) {
	return getCastSessionMedia()?.media.contentId === src?.src;
}
/**
* @see {@link https://developers.google.com/cast/docs/reference/web_sender/cast.framework.CastOptions}
*/
function getDefaultCastOptions() {
	return {
		language: "en-US",
		autoJoinPolicy: chrome.cast.AutoJoinPolicy.ORIGIN_SCOPED,
		receiverApplicationId: chrome.cast.media.DEFAULT_MEDIA_RECEIVER_APP_ID,
		resumeSavedSession: true,
		androidReceiverCompatible: true
	};
}
function getCastErrorMessage(code) {
	const defaultMessage = `Google Cast Error Code: ${code}`;
	switch (code) {
		case chrome.cast.ErrorCode.API_NOT_INITIALIZED: return "The API is not initialized.";
		case chrome.cast.ErrorCode.CANCEL: return "The operation was canceled by the user";
		case chrome.cast.ErrorCode.CHANNEL_ERROR: return "A channel to the receiver is not available.";
		case chrome.cast.ErrorCode.EXTENSION_MISSING: return "The Cast extension is not available.";
		case chrome.cast.ErrorCode.INVALID_PARAMETER: return "The parameters to the operation were not valid.";
		case chrome.cast.ErrorCode.RECEIVER_UNAVAILABLE: return "No receiver was compatible with the session request.";
		case chrome.cast.ErrorCode.SESSION_ERROR: return "A session could not be created, or a session was invalid.";
		case chrome.cast.ErrorCode.TIMEOUT: return "The operation timed out.";
		default: return defaultMessage;
	}
	return defaultMessage;
}
function listenCastContextEvent(type, handler) {
	return listenEvent(getCastContext(), type, handler);
}
//#endregion
export { getCastContext, getCastErrorMessage, getCastFrameworkURL, getCastSession, getCastSessionMedia, getDefaultCastOptions, hasActiveCastSession, hasLoadedCastFramework, isCastAvailable, isCastConnected, listenCastContextEvent };
