// @videojs/spf/dist/dev/media/abr/quality-selection.js
//#region src/media/abr/quality-selection.ts
/**
* Default quality selection configuration.
* Values match Shaka Player upgrade threshold (0.85 = 15% headroom).
*/
const DEFAULT_QUALITY_CONFIG = {
	safetyMargin: .85,
	upgradeMargin: 1.15
};
/**
* Select the track to apply now, given current bandwidth, a current
* selection (optional), and tuning. Returns:
*
* - The bandwidth-fitting optimal when no `currentTrack` is supplied.
* - The optimal when it's a downgrade vs. `currentTrack` (downgrades
*   apply immediately — no hysteresis).
* - The optimal when it clears `currentTrack.bandwidth * upgradeMargin`
*   (upgrade clears hysteresis).
* - `currentTrack` itself when an upgrade doesn't clear the margin
*   (stay put — caller checks identity to no-op).
*
* "Optimal" is the highest-bandwidth track where the available bandwidth
* meets the safety requirement (`currentBandwidth >= track.bandwidth / safetyMargin`).
* Falls back to the lowest-bandwidth track when nothing fits the safety
* margin (preserves a definitive pick under under-bandwidth conditions).
*
* @example
* const tracks = [low, mid, high];
* selectQuality(tracks, 5_000_000, { currentTrack: low });
* // Returns `high` if 5 Mbps clears safety AND high.bandwidth >= low.bandwidth * upgradeMargin.
* // Returns `low` (no-op signal) otherwise.
*/
function selectQuality(tracks, currentBandwidth, opts = {}) {
	if (tracks.length === 0) return;
	const safetyMargin = opts.safetyMargin ?? DEFAULT_QUALITY_CONFIG.safetyMargin;
	const upgradeMargin = opts.upgradeMargin ?? DEFAULT_QUALITY_CONFIG.upgradeMargin;
	const { currentTrack } = opts;
	const sortedTracks = tracks.slice().sort((a, b) => a.bandwidth - b.bandwidth);
	let chosen;
	for (const track of sortedTracks) if (currentBandwidth >= track.bandwidth / safetyMargin) {
		if (!chosen || track.bandwidth > chosen.bandwidth || track.bandwidth === chosen.bandwidth && hasHigherResolution(track, chosen)) chosen = track;
	}
	const optimal = chosen ?? sortedTracks[0];
	if (!optimal) return void 0;
	if (!currentTrack) return optimal;
	if (optimal.bandwidth < currentTrack.bandwidth) return optimal;
	if (optimal.bandwidth >= currentTrack.bandwidth * upgradeMargin) return optimal;
	return currentTrack;
}
/**
* Select the lowest-bandwidth track from the candidate set.
*
* Used as a safety-net fallback by callers that need a definitive pick when
* a primary selection algorithm declines to choose. Pairs naturally with
* `selectQuality` — when ABR has no good answer (e.g., all candidates
* exceed available bandwidth and the algorithm doesn't fall back
* internally), the lowest bitrate is the safest default.
*
* @param tracks - Candidate tracks (can be unsorted)
* @returns The lowest-bandwidth track, or `undefined` if `tracks` is empty
*
* @example
* const fallback = selectLowestQuality(tracks);
*/
function selectLowestQuality(tracks) {
	if (tracks.length === 0) return void 0;
	return tracks.reduce((min, t) => t.bandwidth < min.bandwidth ? t : min);
}
/**
* Check if track A has higher resolution than track B.
* Compares by total pixel count (width × height).
*
* @param trackA - First track to compare
* @param trackB - Second track to compare
* @returns True if trackA has more pixels than trackB
*/
function hasHigherResolution(trackA, trackB) {
	return (trackA.width ?? 0) * (trackA.height ?? 0) > (trackB.width ?? 0) * (trackB.height ?? 0);
}
//#endregion
export { DEFAULT_QUALITY_CONFIG, selectLowestQuality, selectQuality };
