// vidstack/server/chunks/vidstack-C0MM2kpJ.js
import { EventsController, isDOMNode, isFunction, isKeyboardClick, isTouchEvent, setAttribute, setStyle } from "./vidstack-CgxsH2Dv.js";
import { autoUpdate, computePosition, flip, shift } from "@floating-ui/dom";
//#region src/utils/number.ts
/**
* Round a number to the given number of `decimalPlaces`.
*/
function round(num, decimalPlaces = 2) {
	return Number(num.toFixed(decimalPlaces));
}
/**
* Get the number of decimal places in the given `num`.
*
* @example `1 -> 0`
* @example `1.0 -> 0`
* @example `1.1 -> 1`
* @example `1.12 -> 2`
*/
function getNumberOfDecimalPlaces(num) {
	return String(num).split(".")[1]?.length ?? 0;
}
/**
* Clamp a given `value` between a minimum and maximum value.
*/
function clampNumber(min, value, max) {
	return Math.max(min, Math.min(max, value));
}
//#endregion
//#region src/utils/dom.ts
function isEventInside(el, event) {
	const target = event.composedPath()[0];
	return isDOMNode(target) && el.contains(target);
}
const rafJobs = /* @__PURE__ */ new Set();
function scheduleRafJob(job) {
	rafJobs.add(job);
	return () => rafJobs.delete(job);
}
function setAttributeIfEmpty(target, name, value) {
	if (!target.hasAttribute(name)) target.setAttribute(name, value);
}
function setARIALabel(target, $label) {
	if (target.hasAttribute("aria-label") || target.hasAttribute("data-no-label")) return;
	if (!isFunction($label)) {
		setAttribute(target, "aria-label", $label);
		return;
	}
	function updateAriaDescription() {
		setAttribute(target, "aria-label", $label());
	}
	updateAriaDescription();
}
function isElementVisible(el) {
	const style = getComputedStyle(el);
	return style.display !== "none" && parseInt(style.opacity) > 0;
}
function checkVisibility(el) {
	return !!el && ("checkVisibility" in el ? el.checkVisibility({
		checkOpacity: true,
		checkVisibilityCSS: true
	}) : isElementVisible(el));
}
function observeVisibility(el, callback) {
	return scheduleRafJob(() => callback(checkVisibility(el)));
}
function isElementParent(owner, node, test) {
	while (node) if (node === owner) return true;
	else if (test?.(node)) break;
	else node = node.parentElement;
	return false;
}
function onPress(target, handler) {
	return new EventsController(target).add("pointerup", (event) => {
		if (event.button === 0 && !event.defaultPrevented) handler(event);
	}).add("keydown", (event) => {
		if (isKeyboardClick(event)) handler(event);
	});
}
function isTouchPinchEvent(event) {
	return isTouchEvent(event) && (event.touches.length > 1 || event.changedTouches.length > 1);
}
function requestScopedAnimationFrame(callback) {
	return callback();
}
function autoPlacement(el, trigger, placement, { offsetVarName, xOffset, yOffset, ...options }) {
	if (!el) return;
	const floatingPlacement = placement.replace(" ", "-").replace("-center", "");
	setStyle(el, "visibility", !trigger ? "hidden" : null);
	if (!trigger) return;
	let isTop = placement.includes("top");
	const negateX = (x) => placement.includes("left") ? `calc(-1 * ${x})` : x, negateY = (y) => isTop ? `calc(-1 * ${y})` : y;
	return autoUpdate(trigger, el, () => {
		computePosition(trigger, el, {
			placement: floatingPlacement,
			middleware: [
				...options.middleware ?? [],
				flip({
					fallbackAxisSideDirection: "start",
					crossAxis: false
				}),
				shift()
			],
			...options
		}).then(({ x, y, middlewareData }) => {
			const hasFlipped = !!middlewareData.flip?.index;
			isTop = placement.includes(hasFlipped ? "bottom" : "top");
			el.setAttribute("data-placement", hasFlipped ? placement.startsWith("top") ? placement.replace("top", "bottom") : placement.replace("bottom", "top") : placement);
			Object.assign(el.style, {
				top: `calc(${y + "px"} + ${negateY(yOffset ? yOffset + "px" : `var(--${offsetVarName}-y-offset, 0px)`)})`,
				left: `calc(${x + "px"} + ${negateX(xOffset ? xOffset + "px" : `var(--${offsetVarName}-x-offset, 0px)`)})`
			});
		});
	});
}
function hasAnimation(el) {
	return getComputedStyle(el).animationName !== "none";
}
function isHTMLElement(el) {
	return el instanceof HTMLElement;
}
//#endregion
//#region src/providers/type-check.ts
/** @see {@link https://www.vidstack.io/docs/player/providers/audio} */
function isAudioProvider(provider) {
	return provider?.$$PROVIDER_TYPE === "AUDIO";
}
/** @see {@link https://www.vidstack.io/docs/player/providers/video} */
function isVideoProvider(provider) {
	return provider?.$$PROVIDER_TYPE === "VIDEO";
}
/** @see {@link https://www.vidstack.io/docs/player/providers/hls} */
function isHLSProvider(provider) {
	return provider?.$$PROVIDER_TYPE === "HLS";
}
function isDASHProvider(provider) {
	return provider?.$$PROVIDER_TYPE === "DASH";
}
/** @see {@link https://www.vidstack.io/docs/player/providers/youtube} */
function isYouTubeProvider(provider) {
	return provider?.$$PROVIDER_TYPE === "YOUTUBE";
}
/** @see {@link https://www.vidstack.io/docs/player/providers/vimeo} */
function isVimeoProvider(provider) {
	return provider?.$$PROVIDER_TYPE === "VIMEO";
}
/** @see {@link https://www.vidstack.io/docs/player/providers/google-cast} */
function isGoogleCastProvider(provider) {
	return provider?.$$PROVIDER_TYPE === "GOOGLE_CAST";
}
/** @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLAudioElement} */
function isHTMLAudioElement(element) {
	return false;
}
/** @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement} */
function isHTMLVideoElement(element) {
	return false;
}
/** @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement} */
function isHTMLMediaElement(element) {
	return isHTMLAudioElement(element) || isHTMLVideoElement(element);
}
/** @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLIFrameElement} */
function isHTMLIFrameElement(element) {
	return false;
}
//#endregion
export { autoPlacement, clampNumber, getNumberOfDecimalPlaces, hasAnimation, isAudioProvider, isDASHProvider, isElementParent, isEventInside, isGoogleCastProvider, isHLSProvider, isHTMLAudioElement, isHTMLElement, isHTMLIFrameElement, isHTMLMediaElement, isHTMLVideoElement, isTouchPinchEvent, isVideoProvider, isVimeoProvider, isYouTubeProvider, observeVisibility, onPress, requestScopedAnimationFrame, round, setARIALabel, setAttributeIfEmpty };
