// @videojs/core/dist/default/core/ui/cast-button/cast-button-core.js
import { createState } from "@videojs/store";
import { defaults } from "@videojs/utils/object";
import { isFunction } from "@videojs/utils/predicate";
//#region src/core/ui/cast-button/cast-button-core.ts
var CastButtonCore = class CastButtonCore {
	static defaultProps = {
		label: "",
		disabled: false
	};
	state = createState({
		castState: "disconnected",
		availability: "unsupported",
		label: ""
	});
	#props = { ...CastButtonCore.defaultProps };
	#media = null;
	constructor(props) {
		if (props) this.setProps(props);
	}
	setProps(props) {
		this.#props = defaults(props, CastButtonCore.defaultProps);
	}
	getLabel(state) {
		const { label } = this.#props;
		if (isFunction(label)) {
			const customLabel = label(state);
			if (customLabel) return customLabel;
		} else if (label) return label;
		if (state.castState === "connected") return "Stop casting";
		if (state.castState === "connecting") return "Connecting";
		return "Start casting";
	}
	getAttrs(state) {
		return {
			"aria-label": this.getLabel(state),
			"aria-disabled": this.#props.disabled ? "true" : void 0
		};
	}
	setMedia(media) {
		this.#media = media;
	}
	getState() {
		const media = this.#media;
		const castSupported = !!globalThis.chrome;
		this.state.patch({
			castState: media.remotePlaybackState,
			availability: castSupported ? media.remotePlaybackAvailability : "unsupported"
		});
		this.state.patch({ label: this.getLabel(this.state.current) });
		return this.state.current;
	}
	async toggle(media) {
		if (this.#props.disabled) return;
		if (media.remotePlaybackAvailability !== "available") return;
		try {
			await media.toggleRemotePlayback();
		} catch {}
	}
};
//#endregion
export { CastButtonCore };
