// @videojs/html/cdn/media/simple-hls-video.dev.js
import { t as MediaAttachMixin } from "../media-attach-mixin-C1KQy-y8.js";
import { t as listen } from "../listen-CJT7C4kM.js";
import { f as safeDefine } from "../predicate-Dqd8wbW6.js";
import { t as anyAbortSignal } from "../abort-DQxwCs92.js";
import { n as CustomMediaElement, t as HTMLVideoElementHost } from "../video-host-BU9t26fe.js";

//#region ../utils/dist/string/generate-id.js
/**
* Generate a unique-ish string ID via timestamp + random.
*
* @returns String in `timestamp-random` format (e.g. `"1738423156789-542891"`).
*
* @example
* const id = generateId(); // "1738423156789-542891"
*/
function generateId() {
	return `${Date.now()}-${Math.floor(Math.random() * 1e6)}`;
}

//#endregion
//#region ../../node_modules/.pnpm/signal-polyfill@0.2.2/node_modules/signal-polyfill/dist/index.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {
	enumerable: true,
	configurable: true,
	writable: true,
	value
}) : obj[key] = value;
var __publicField = (obj, key, value) => {
	__defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
	return value;
};
var __accessCheck = (obj, member, msg) => {
	if (!member.has(obj)) throw TypeError("Cannot " + msg);
};
var __privateIn = (member, obj) => {
	if (Object(obj) !== obj) throw TypeError("Cannot use the \"in\" operator on this value");
	return member.has(obj);
};
var __privateAdd = (obj, member, value) => {
	if (member.has(obj)) throw TypeError("Cannot add the same private member more than once");
	member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
};
var __privateMethod = (obj, member, method) => {
	__accessCheck(obj, member, "access private method");
	return method;
};
/**
* @license
* Copyright Google LLC All Rights Reserved.
*
* Use of this source code is governed by an MIT-style license that can be
* found in the LICENSE file at https://angular.io/license
*/
function defaultEquals(a, b) {
	return Object.is(a, b);
}
/**
* @license
* Copyright Google LLC All Rights Reserved.
*
* Use of this source code is governed by an MIT-style license that can be
* found in the LICENSE file at https://angular.io/license
*/
let activeConsumer = null;
let inNotificationPhase = false;
let epoch = 1;
const SIGNAL = /* @__PURE__ */ Symbol("SIGNAL");
function setActiveConsumer(consumer) {
	const prev = activeConsumer;
	activeConsumer = consumer;
	return prev;
}
function getActiveConsumer() {
	return activeConsumer;
}
function isInNotificationPhase() {
	return inNotificationPhase;
}
const REACTIVE_NODE = {
	version: 0,
	lastCleanEpoch: 0,
	dirty: false,
	producerNode: void 0,
	producerLastReadVersion: void 0,
	producerIndexOfThis: void 0,
	nextProducerIndex: 0,
	liveConsumerNode: void 0,
	liveConsumerIndexOfThis: void 0,
	consumerAllowSignalWrites: false,
	consumerIsAlwaysLive: false,
	producerMustRecompute: () => false,
	producerRecomputeValue: () => {},
	consumerMarkedDirty: () => {},
	consumerOnSignalRead: () => {}
};
function producerAccessed(node) {
	if (inNotificationPhase) throw new Error(typeof ngDevMode !== "undefined" && ngDevMode ? `Assertion error: signal read during notification phase` : "");
	if (activeConsumer === null) return;
	activeConsumer.consumerOnSignalRead(node);
	const idx = activeConsumer.nextProducerIndex++;
	assertConsumerNode(activeConsumer);
	if (idx < activeConsumer.producerNode.length && activeConsumer.producerNode[idx] !== node) {
		if (consumerIsLive(activeConsumer)) {
			const staleProducer = activeConsumer.producerNode[idx];
			producerRemoveLiveConsumerAtIndex(staleProducer, activeConsumer.producerIndexOfThis[idx]);
		}
	}
	if (activeConsumer.producerNode[idx] !== node) {
		activeConsumer.producerNode[idx] = node;
		activeConsumer.producerIndexOfThis[idx] = consumerIsLive(activeConsumer) ? producerAddLiveConsumer(node, activeConsumer, idx) : 0;
	}
	activeConsumer.producerLastReadVersion[idx] = node.version;
}
function producerIncrementEpoch() {
	epoch++;
}
function producerUpdateValueVersion(node) {
	if (!node.dirty && node.lastCleanEpoch === epoch) return;
	if (!node.producerMustRecompute(node) && !consumerPollProducersForChange(node)) {
		node.dirty = false;
		node.lastCleanEpoch = epoch;
		return;
	}
	node.producerRecomputeValue(node);
	node.dirty = false;
	node.lastCleanEpoch = epoch;
}
function producerNotifyConsumers(node) {
	if (node.liveConsumerNode === void 0) return;
	const prev = inNotificationPhase;
	inNotificationPhase = true;
	try {
		for (const consumer of node.liveConsumerNode) if (!consumer.dirty) consumerMarkDirty(consumer);
	} finally {
		inNotificationPhase = prev;
	}
}
function producerUpdatesAllowed() {
	return (activeConsumer == null ? void 0 : activeConsumer.consumerAllowSignalWrites) !== false;
}
function consumerMarkDirty(node) {
	var _a;
	node.dirty = true;
	producerNotifyConsumers(node);
	(_a = node.consumerMarkedDirty) == null || _a.call(node.wrapper ?? node);
}
function consumerBeforeComputation(node) {
	node && (node.nextProducerIndex = 0);
	return setActiveConsumer(node);
}
function consumerAfterComputation(node, prevConsumer) {
	setActiveConsumer(prevConsumer);
	if (!node || node.producerNode === void 0 || node.producerIndexOfThis === void 0 || node.producerLastReadVersion === void 0) return;
	if (consumerIsLive(node)) for (let i = node.nextProducerIndex; i < node.producerNode.length; i++) producerRemoveLiveConsumerAtIndex(node.producerNode[i], node.producerIndexOfThis[i]);
	while (node.producerNode.length > node.nextProducerIndex) {
		node.producerNode.pop();
		node.producerLastReadVersion.pop();
		node.producerIndexOfThis.pop();
	}
}
function consumerPollProducersForChange(node) {
	assertConsumerNode(node);
	for (let i = 0; i < node.producerNode.length; i++) {
		const producer = node.producerNode[i];
		const seenVersion = node.producerLastReadVersion[i];
		if (seenVersion !== producer.version) return true;
		producerUpdateValueVersion(producer);
		if (seenVersion !== producer.version) return true;
	}
	return false;
}
function producerAddLiveConsumer(node, consumer, indexOfThis) {
	var _a;
	assertProducerNode(node);
	assertConsumerNode(node);
	if (node.liveConsumerNode.length === 0) {
		(_a = node.watched) == null || _a.call(node.wrapper);
		for (let i = 0; i < node.producerNode.length; i++) node.producerIndexOfThis[i] = producerAddLiveConsumer(node.producerNode[i], node, i);
	}
	node.liveConsumerIndexOfThis.push(indexOfThis);
	return node.liveConsumerNode.push(consumer) - 1;
}
function producerRemoveLiveConsumerAtIndex(node, idx) {
	var _a;
	assertProducerNode(node);
	assertConsumerNode(node);
	if (typeof ngDevMode !== "undefined" && ngDevMode && idx >= node.liveConsumerNode.length) throw new Error(`Assertion error: active consumer index ${idx} is out of bounds of ${node.liveConsumerNode.length} consumers)`);
	if (node.liveConsumerNode.length === 1) {
		(_a = node.unwatched) == null || _a.call(node.wrapper);
		for (let i = 0; i < node.producerNode.length; i++) producerRemoveLiveConsumerAtIndex(node.producerNode[i], node.producerIndexOfThis[i]);
	}
	const lastIdx = node.liveConsumerNode.length - 1;
	node.liveConsumerNode[idx] = node.liveConsumerNode[lastIdx];
	node.liveConsumerIndexOfThis[idx] = node.liveConsumerIndexOfThis[lastIdx];
	node.liveConsumerNode.length--;
	node.liveConsumerIndexOfThis.length--;
	if (idx < node.liveConsumerNode.length) {
		const idxProducer = node.liveConsumerIndexOfThis[idx];
		const consumer = node.liveConsumerNode[idx];
		assertConsumerNode(consumer);
		consumer.producerIndexOfThis[idxProducer] = idx;
	}
}
function consumerIsLive(node) {
	var _a;
	return node.consumerIsAlwaysLive || (((_a = node == null ? void 0 : node.liveConsumerNode) == null ? void 0 : _a.length) ?? 0) > 0;
}
function assertConsumerNode(node) {
	node.producerNode ?? (node.producerNode = []);
	node.producerIndexOfThis ?? (node.producerIndexOfThis = []);
	node.producerLastReadVersion ?? (node.producerLastReadVersion = []);
}
function assertProducerNode(node) {
	node.liveConsumerNode ?? (node.liveConsumerNode = []);
	node.liveConsumerIndexOfThis ?? (node.liveConsumerIndexOfThis = []);
}
/**
* @license
* Copyright Google LLC All Rights Reserved.
*
* Use of this source code is governed by an MIT-style license that can be
* found in the LICENSE file at https://angular.io/license
*/
function computedGet(node) {
	producerUpdateValueVersion(node);
	producerAccessed(node);
	if (node.value === ERRORED) throw node.error;
	return node.value;
}
function createComputed(computation) {
	const node = Object.create(COMPUTED_NODE);
	node.computation = computation;
	const computed = () => computedGet(node);
	computed[SIGNAL] = node;
	return computed;
}
const UNSET = /* @__PURE__ */ Symbol("UNSET");
const COMPUTING = /* @__PURE__ */ Symbol("COMPUTING");
const ERRORED = /* @__PURE__ */ Symbol("ERRORED");
const COMPUTED_NODE = {
	...REACTIVE_NODE,
	value: UNSET,
	dirty: true,
	error: null,
	equal: defaultEquals,
	producerMustRecompute(node) {
		return node.value === UNSET || node.value === COMPUTING;
	},
	producerRecomputeValue(node) {
		if (node.value === COMPUTING) throw new Error("Detected cycle in computations.");
		const oldValue = node.value;
		node.value = COMPUTING;
		const prevConsumer = consumerBeforeComputation(node);
		let newValue;
		let wasEqual = false;
		try {
			newValue = node.computation.call(node.wrapper);
			wasEqual = oldValue !== UNSET && oldValue !== ERRORED && node.equal.call(node.wrapper, oldValue, newValue);
		} catch (err) {
			newValue = ERRORED;
			node.error = err;
		} finally {
			consumerAfterComputation(node, prevConsumer);
		}
		if (wasEqual) {
			node.value = oldValue;
			return;
		}
		node.value = newValue;
		node.version++;
	}
};
/**
* @license
* Copyright Google LLC All Rights Reserved.
*
* Use of this source code is governed by an MIT-style license that can be
* found in the LICENSE file at https://angular.io/license
*/
function defaultThrowError() {
	throw new Error();
}
let throwInvalidWriteToSignalErrorFn = defaultThrowError;
function throwInvalidWriteToSignalError() {
	throwInvalidWriteToSignalErrorFn();
}
/**
* @license
* Copyright Google LLC All Rights Reserved.
*
* Use of this source code is governed by an MIT-style license that can be
* found in the LICENSE file at https://angular.io/license
*/
function createSignal(initialValue) {
	const node = Object.create(SIGNAL_NODE);
	node.value = initialValue;
	const getter = () => {
		producerAccessed(node);
		return node.value;
	};
	getter[SIGNAL] = node;
	return getter;
}
function signalGetFn() {
	producerAccessed(this);
	return this.value;
}
function signalSetFn(node, newValue) {
	if (!producerUpdatesAllowed()) throwInvalidWriteToSignalError();
	if (!node.equal.call(node.wrapper, node.value, newValue)) {
		node.value = newValue;
		signalValueChanged(node);
	}
}
const SIGNAL_NODE = {
	...REACTIVE_NODE,
	equal: defaultEquals,
	value: void 0
};
function signalValueChanged(node) {
	node.version++;
	producerIncrementEpoch();
	producerNotifyConsumers(node);
}
/**
* @license
* Copyright 2024 Bloomberg Finance L.P.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const NODE = Symbol("node");
var Signal;
((Signal2) => {
	var _a, _brand, _b, _brand2;
	class State {
		constructor(initialValue, options = {}) {
			__privateAdd(this, _brand);
			__publicField(this, _a);
			const node = createSignal(initialValue)[SIGNAL];
			this[NODE] = node;
			node.wrapper = this;
			if (options) {
				const equals = options.equals;
				if (equals) node.equal = equals;
				node.watched = options[Signal2.subtle.watched];
				node.unwatched = options[Signal2.subtle.unwatched];
			}
		}
		get() {
			if (!(0, Signal2.isState)(this)) throw new TypeError("Wrong receiver type for Signal.State.prototype.get");
			return signalGetFn.call(this[NODE]);
		}
		set(newValue) {
			if (!(0, Signal2.isState)(this)) throw new TypeError("Wrong receiver type for Signal.State.prototype.set");
			if (isInNotificationPhase()) throw new Error("Writes to signals not permitted during Watcher callback");
			const ref = this[NODE];
			signalSetFn(ref, newValue);
		}
	}
	_a = NODE;
	_brand = /* @__PURE__ */ new WeakSet();
	Signal2.isState = (s) => typeof s === "object" && __privateIn(_brand, s);
	Signal2.State = State;
	class Computed {
		constructor(computation, options) {
			__privateAdd(this, _brand2);
			__publicField(this, _b);
			const node = createComputed(computation)[SIGNAL];
			node.consumerAllowSignalWrites = true;
			this[NODE] = node;
			node.wrapper = this;
			if (options) {
				const equals = options.equals;
				if (equals) node.equal = equals;
				node.watched = options[Signal2.subtle.watched];
				node.unwatched = options[Signal2.subtle.unwatched];
			}
		}
		get() {
			if (!(0, Signal2.isComputed)(this)) throw new TypeError("Wrong receiver type for Signal.Computed.prototype.get");
			return computedGet(this[NODE]);
		}
	}
	_b = NODE;
	_brand2 = /* @__PURE__ */ new WeakSet();
	Signal2.isComputed = (c) => typeof c === "object" && __privateIn(_brand2, c);
	Signal2.Computed = Computed;
	((subtle2) => {
		var _a2, _brand3, _assertSignals, assertSignals_fn;
		function untrack(cb) {
			let output;
			let prevActiveConsumer = null;
			try {
				prevActiveConsumer = setActiveConsumer(null);
				output = cb();
			} finally {
				setActiveConsumer(prevActiveConsumer);
			}
			return output;
		}
		subtle2.untrack = untrack;
		function introspectSources(sink) {
			var _a3;
			if (!(0, Signal2.isComputed)(sink) && !(0, Signal2.isWatcher)(sink)) throw new TypeError("Called introspectSources without a Computed or Watcher argument");
			return ((_a3 = sink[NODE].producerNode) == null ? void 0 : _a3.map((n) => n.wrapper)) ?? [];
		}
		subtle2.introspectSources = introspectSources;
		function introspectSinks(signal) {
			var _a3;
			if (!(0, Signal2.isComputed)(signal) && !(0, Signal2.isState)(signal)) throw new TypeError("Called introspectSinks without a Signal argument");
			return ((_a3 = signal[NODE].liveConsumerNode) == null ? void 0 : _a3.map((n) => n.wrapper)) ?? [];
		}
		subtle2.introspectSinks = introspectSinks;
		function hasSinks(signal) {
			if (!(0, Signal2.isComputed)(signal) && !(0, Signal2.isState)(signal)) throw new TypeError("Called hasSinks without a Signal argument");
			const liveConsumerNode = signal[NODE].liveConsumerNode;
			if (!liveConsumerNode) return false;
			return liveConsumerNode.length > 0;
		}
		subtle2.hasSinks = hasSinks;
		function hasSources(signal) {
			if (!(0, Signal2.isComputed)(signal) && !(0, Signal2.isWatcher)(signal)) throw new TypeError("Called hasSources without a Computed or Watcher argument");
			const producerNode = signal[NODE].producerNode;
			if (!producerNode) return false;
			return producerNode.length > 0;
		}
		subtle2.hasSources = hasSources;
		class Watcher {
			constructor(notify) {
				__privateAdd(this, _brand3);
				__privateAdd(this, _assertSignals);
				__publicField(this, _a2);
				let node = Object.create(REACTIVE_NODE);
				node.wrapper = this;
				node.consumerMarkedDirty = notify;
				node.consumerIsAlwaysLive = true;
				node.consumerAllowSignalWrites = false;
				node.producerNode = [];
				this[NODE] = node;
			}
			watch(...signals) {
				if (!(0, Signal2.isWatcher)(this)) throw new TypeError("Called unwatch without Watcher receiver");
				__privateMethod(this, _assertSignals, assertSignals_fn).call(this, signals);
				const node = this[NODE];
				node.dirty = false;
				const prev = setActiveConsumer(node);
				for (const signal of signals) producerAccessed(signal[NODE]);
				setActiveConsumer(prev);
			}
			unwatch(...signals) {
				if (!(0, Signal2.isWatcher)(this)) throw new TypeError("Called unwatch without Watcher receiver");
				__privateMethod(this, _assertSignals, assertSignals_fn).call(this, signals);
				const node = this[NODE];
				assertConsumerNode(node);
				for (let i = node.producerNode.length - 1; i >= 0; i--) if (signals.includes(node.producerNode[i].wrapper)) {
					producerRemoveLiveConsumerAtIndex(node.producerNode[i], node.producerIndexOfThis[i]);
					const lastIdx = node.producerNode.length - 1;
					node.producerNode[i] = node.producerNode[lastIdx];
					node.producerIndexOfThis[i] = node.producerIndexOfThis[lastIdx];
					node.producerNode.length--;
					node.producerIndexOfThis.length--;
					node.nextProducerIndex--;
					if (i < node.producerNode.length) {
						const idxConsumer = node.producerIndexOfThis[i];
						const producer = node.producerNode[i];
						assertProducerNode(producer);
						producer.liveConsumerIndexOfThis[idxConsumer] = i;
					}
				}
			}
			getPending() {
				if (!(0, Signal2.isWatcher)(this)) throw new TypeError("Called getPending without Watcher receiver");
				return this[NODE].producerNode.filter((n) => n.dirty).map((n) => n.wrapper);
			}
		}
		_a2 = NODE;
		_brand3 = /* @__PURE__ */ new WeakSet();
		_assertSignals = /* @__PURE__ */ new WeakSet();
		assertSignals_fn = function(signals) {
			for (const signal of signals) if (!(0, Signal2.isComputed)(signal) && !(0, Signal2.isState)(signal)) throw new TypeError("Called watch/unwatch without a Computed or State argument");
		};
		Signal2.isWatcher = (w) => __privateIn(_brand3, w);
		subtle2.Watcher = Watcher;
		function currentComputed() {
			var _a3;
			return (_a3 = getActiveConsumer()) == null ? void 0 : _a3.wrapper;
		}
		subtle2.currentComputed = currentComputed;
		subtle2.watched = Symbol("watched");
		subtle2.unwatched = Symbol("unwatched");
	})(Signal2.subtle || (Signal2.subtle = {}));
})(Signal || (Signal = {}));

//#endregion
//#region ../spf/dist/dev/core/signals/primitives.js
/** Read a signal value without tracking it as a dependency. */
const untrack = Signal.subtle.untrack;
/**
* Read a signal's current value without tracking it as a dependency. Sugar
* for `untrack(() => signal.get())` to reduce boilerplate at single-read
* sites. Structurally typed to accept any signal-like (Signal, Computed,
* ReadonlySignal).
*
* Accepts an optional `transform` to project the value in the same call;
* the default is the identity function so the single-arg form returns `T`
* unchanged.
*
* @example
* const value = peek(someSignal);
* const id = peek(presentationSignal, (p) => p?.id);
*/
function peek(source, transform = (v) => v) {
	return untrack(() => transform(source.get()));
}
/** Create a writable reactive value. */
function signal(initialValue, options) {
	return new Signal.State(initialValue, options);
}
/** Create a computed reactive value. */
function computed(fn, options) {
	return new Signal.Computed(fn, options);
}
function update(signal, updater) {
	const current = untrack(() => signal.get());
	if (typeof updater === "function") signal.set(updater(current));
	else signal.set({
		...current,
		...updater
	});
}
/**
* Read every signal in a map and return a plain object snapshot. Each read
* tracks in the surrounding Computed/Effect — equivalent to calling `.get()`
* on a single `Signal<S>` over the merged shape.
*
* Convenience for behaviors that pass whole state/context snapshots to pure
* helpers; prefer per-field reads when only a few fields are needed.
*/
function snapshot(map) {
	const out = {};
	for (const key in map) out[key] = map[key].get();
	return out;
}

//#endregion
//#region ../spf/dist/dev/core/composition/create-composition.js
/**
* Create a composition from a set of behaviors.
*
* Composition unions the behaviors' declared `stateKeys` / `contextKeys`
* to know which signals to create. Each signal is seeded from
* `initialState` / `initialContext` when supplied, defaulting to
* `undefined`. Behaviors are responsible for writing their own slots
* once their preconditions are met.
*
* Cross-behavior type conflicts (e.g. two behaviors disagreeing on a
* field's type) surface as a compose-time type error via
* `ValidateComposition`.
*
* @example
* ```ts
* const composition = createComposition([resolvePresentation, switchVideoQuality], {
*   config: { parsePresentation: parseMultivariantPlaylist, initialBandwidth: 2_000_000 },
*   initialState: { bandwidthState: { fastEstimate: 0, ... } },
* });
* ```
*/
/**
* Create a typed signal map for a given set of keys, seeded from an
* optional partial initial value.
*
* Pipeline: `Set` dedupes the iterable (insertion order preserved, so
* first occurrence wins) → `Object.fromEntries` materializes one
* `signal()` per unique key, seeded from `initial[key]` or `undefined`.
*
* Per-key value types live in TypeScript only — at runtime every signal
* is `Signal<unknown>`. The boundary cast at the return narrows the wide
* `Record<PropertyKey, Signal<unknown>>` shape to the caller's expected
* per-key types from `S`.
*
* Used by `createComposition` to derive engine state/context maps from
* the union of behaviors' declared `stateKeys` / `contextKeys`.
*
* @example
* ```ts
* interface State { count?: number; label?: string }
* const state = buildSignalMap<State>(['count', 'label'], { count: 5 });
* state.count.get(); // 5
* state.label.get(); // undefined
* ```
*/
function buildSignalMap(keys, initial) {
	const init = initial;
	const uniqueKeys = new Set(keys);
	return Object.fromEntries([...uniqueKeys].map((key) => [key, signal(init[key])]));
}
function createComposition(behaviors, options) {
	const validBehaviors = behaviors;
	const state = buildSignalMap(validBehaviors.flatMap((b) => b.stateKeys), options?.initialState ?? {});
	const context = buildSignalMap(validBehaviors.flatMap((b) => b.contextKeys), options?.initialContext ?? {});
	const deps = {
		state,
		context,
		config: options?.config ?? {}
	};
	const cleanups = validBehaviors.map((behavior) => behavior.setup(deps));
	return {
		state,
		context,
		async destroy() {
			const results = [];
			for (const cleanup of cleanups) {
				if (cleanup == null) continue;
				if (typeof cleanup === "function") results.push(cleanup());
				else if ("destroy" in cleanup) results.push(cleanup.destroy());
			}
			await Promise.all(results);
			for (const sig of Object.values(state)) sig.set(void 0);
			for (const sig of Object.values(context)) sig.set(void 0);
		}
	};
}
function defineBehavior(behavior) {
	return behavior;
}

//#endregion
//#region ../spf/dist/dev/core/composition/share-signals.js
/**
* Behavior factory that hands the composition's signal refs to a
* consumer-supplied callback (`config.onSignalsReady`) at setup time.
*
* Generic over `S` and `C` — the caller instantiates with their own
* state/context types, and the callback's parameter shape is fully
* type-driven from those. Suitable for both reads and writes (per-slot
* intent can be expressed by typing captured refs as `Signal<T>` or
* `ReadonlySignal<T>` at the call site).
*
* Declares no keys of its own (`stateKeys: []`, `contextKeys: []`); the
* composition's state/context maps come from other behaviors' key
* declarations. This behavior just observes/forwards whatever signals
* the composition built.
*
* Uses a `Behavior<>` literal (not `defineBehavior`) so the empty key
* arrays don't trip the exhaustiveness check — the setup-param state/
* context shapes here describe what the consumer's callback receives,
* not keys this behavior needs created.
*/
function makeShareSignals() {
	return {
		stateKeys: [],
		contextKeys: [],
		setup: ({ state, context, config }) => {
			config.onSignalsReady?.({
				state,
				context
			});
		}
	};
}

//#endregion
//#region ../spf/dist/dev/media/dom/text/resolve-vtt-segment.js
/**
* Parse a VTT segment using browser's native parser.
*
* Creates a dummy video element with a track element to leverage
* the browser's optimized VTT parsing. Returns parsed VTTCue objects.
*/
let dummyVideo = null;
function ensureDummyVideo() {
	if (!dummyVideo) {
		dummyVideo = document.createElement("video");
		dummyVideo.muted = true;
		dummyVideo.preload = "none";
		dummyVideo.style.display = "none";
		dummyVideo.crossOrigin = "anonymous";
	}
	return dummyVideo;
}
function resolveVttSegment(url) {
	const video = ensureDummyVideo();
	const track = document.createElement("track");
	track.kind = "subtitles";
	track.default = true;
	return new Promise((resolve, reject) => {
		const onLoad = () => {
			const cues = [];
			const textTrack = track.track;
			if (textTrack.cues) for (let i = 0; i < textTrack.cues.length; i++) {
				const cue = textTrack.cues[i];
				if (cue) cues.push(cue);
			}
			cleanup();
			resolve(cues);
		};
		const onError = () => {
			cleanup();
			reject(/* @__PURE__ */ new Error(`Failed to load VTT segment: ${url}`));
		};
		const cleanup = () => {
			track.removeEventListener("load", onLoad);
			track.removeEventListener("error", onError);
			video.removeChild(track);
		};
		track.addEventListener("load", onLoad);
		track.addEventListener("error", onError);
		video.appendChild(track);
		track.src = url;
	});
}

//#endregion
//#region ../spf/dist/dev/core/signals/effect.js
const pending = /* @__PURE__ */ new Set();
const watcher = new Signal.subtle.Watcher(() => {
	queueMicrotask(runPending);
});
function runPending() {
	for (const c of watcher.getPending()) pending.add(c);
	watcher.watch();
	for (const c of pending) {
		pending.delete(c);
		c.get();
	}
}
/**
* Run a side effect whenever its signal dependencies change.
*
* Executes immediately (synchronous initial run), then re-runs on the next
* microtask after any dependency changes. If the callback returns a function,
* it is called before each re-run and when the effect is stopped — the same
* cleanup contract as Preact Signals, Maverick Signals, and Svelte 5 $effect.
*
* Returns a cleanup function that stops the effect.
*/
function effect(fn) {
	let cleanup;
	const c = new Signal.Computed(() => {
		if (typeof cleanup === "function") cleanup();
		cleanup = fn();
	});
	watcher.watch(c);
	c.get();
	return () => {
		watcher.unwatch(c);
		if (typeof cleanup === "function") cleanup();
	};
}

//#endregion
//#region ../spf/dist/dev/core/machine.js
/**
* Provisions the shared mechanics for all machine-like primitives: a snapshot
* signal, an untracked state reader, and a transition function.
*
* Internal — consumed by `createMachineActor` and `createMachineReactor`. Not part of the
* public API.
*/
function createMachineCore(initialSnapshot) {
	const snapshotSignal = signal(initialSnapshot);
	const getState = () => untrack(() => snapshotSignal.get().value);
	const transition = (to) => update(snapshotSignal, (current) => ({
		...current,
		value: to
	}));
	return {
		snapshotSignal,
		getState,
		transition
	};
}

//#endregion
//#region ../spf/dist/dev/core/reactors/create-machine-reactor.js
const toArray = (x) => x === void 0 ? [] : Array.isArray(x) ? x : [x];
/**
* Creates a reactive Reactor from a declarative definition.
*
* A Reactor is driven by subscriptions to external signals rather than
* imperative messages. Each state holds an array of effect functions —
* every element becomes one independent `effect()` call gated on that state,
* with its own dependency tracking and cleanup lifecycle.
*
* `'destroying'` and `'destroyed'` are always implicit terminal states.
* `destroy()` transitions through both in sequence: `'destroying'` first (for
* potential async teardown in a future extension), then immediately `'destroyed'`
* for the synchronous base case. Active effect cleanups fire via disposal.
*
* @example
* const reactor = createMachineReactor({
*   initial: 'waiting',
*   monitor: () => srcSignal.get() ? 'active' : 'waiting',
*   states: {
*     active: {
*       // entry: runs once on state entry; fn body is automatically untracked.
*       entry: () => listen(el, 'play', handler),
*       // effects: re-runs whenever tracked signals change.
*       effects: () => { currentTimeSignal.get(); return cleanup; },
*     },
*     waiting: {},
*   }
* });
*/
function createMachineReactor(def) {
	const { snapshotSignal, getState, transition } = createMachineCore({ value: def.initial });
	const effectDisposals = [];
	const wrapResult = (result) => {
		if (!result) return void 0;
		if (typeof result === "function") return result;
		return () => result.abort();
	};
	const untracked = (baseCall) => () => untrack(baseCall);
	const isTerminal = (snapshot) => snapshot.value === "destroying" || snapshot.value === "destroyed";
	const descriptors = [...toArray(def.monitor).map((fn) => ({
		fn: () => {
			const target = fn();
			if (target !== getState()) transition(target);
		},
		shouldSkip: isTerminal
	})), ...Object.entries(def.states).flatMap(([state, stateDef]) => {
		const isNotState = (snapshot) => snapshot.value !== state;
		return [...toArray(stateDef.entry).map((fn) => ({
			fn,
			shouldSkip: isNotState,
			toFnCall: untracked
		})), ...toArray(stateDef.effects).map((fn) => ({
			fn,
			shouldSkip: isNotState
		}))];
	})];
	const toEffect = ({ fn, shouldSkip, toFnCall = (baseCall) => baseCall }) => effect(() => {
		if (shouldSkip(snapshotSignal.get())) return;
		const baseCall = () => fn();
		return wrapResult(toFnCall(baseCall)());
	});
	effectDisposals.push(...descriptors.map(toEffect));
	return {
		get snapshot() {
			return snapshotSignal;
		},
		destroy() {
			const state = getState();
			if (state === "destroying" || state === "destroyed") return;
			transition("destroying");
			transition("destroyed");
			for (const dispose of effectDisposals) dispose();
		}
	};
}

//#endregion
//#region ../spf/dist/dev/media/buffer/forward-buffer.js
/**
* Default forward buffer configuration.
*/
const DEFAULT_FORWARD_BUFFER_CONFIG = { bufferDuration: 30 };
/**
* Get segments that need to be loaded for forward buffer.
*
* Determines which segments to load to maintain target buffer duration.
* Handles discontiguous buffering (gaps after seeks).
*
* Algorithm:
* 1. Calculate target time: currentTime + bufferDuration
* 2. Find all segments in range [currentTime, targetTime)
* 3. Filter out segments already buffered at that time position
* 4. Return segments to load (fills gaps + extends to target)
*
* @param segments - All available segments from playlist
* @param bufferedSegments - Segments already buffered (ordered by startTime)
* @param currentTime - Current playback position in seconds
* @param config - Optional forward buffer configuration
* @returns Array of segments to load (empty if buffer is sufficient)
*
* @example
* // After seek: buffered [0-12, 18-30], playing at 7s
* const toLoad = getSegmentsToLoad(segments, buffered, 7, { bufferDuration: 24 });
* // Returns [seg-12, seg-30] (fills gap, extends to target 31s)
*/
/**
* Calculate the start time from which to flush forward buffer content.
*
* Content that starts at or beyond `currentTime + bufferDuration` is no
* longer needed for the current playback position and should be removed
* from the SourceBuffer. This prevents unbounded accumulation of scattered
* SourceBuffer content after seeks, which can cause QuotaExceededError on
* long-form content.
*
* Returns `Infinity` when nothing needs flushing (no buffered segments
* exist beyond the threshold).
*
* @param bufferedSegments - Segments currently tracked in the buffer model
* @param currentTime - Current playback position in seconds
* @param config - Optional forward buffer configuration
* @returns Start time to flush from (flush range: [flushStart, Infinity)),
*          or Infinity if no flush is needed
*
* @example
* // Playing at 0s, buffered [0,6,12,18,24,30,36], bufferDuration=30
* const flushStart = calculateForwardFlushPoint(segments, 0);
* // Returns 30 — flush [30, Infinity), keep [0, 30)
*/
function calculateForwardFlushPoint(bufferedSegments, currentTime, config = DEFAULT_FORWARD_BUFFER_CONFIG) {
	if (bufferedSegments.length === 0) return Infinity;
	const threshold = currentTime + config.bufferDuration;
	const beyond = bufferedSegments.filter((seg) => seg.startTime >= threshold);
	if (beyond.length === 0) return Infinity;
	return Math.min(...beyond.map((seg) => seg.startTime));
}
/**
* Find the start time of the segment containing `currentTime` (or the last
* segment if `currentTime` is past the end). Returns `undefined` when
* `currentTime` is undefined or when no segment matches.
*
* Used to detect "meaningful currentTime change" — two times that map to the
* same segment start aren't a load-trigger; crossing a segment boundary is.
*/
function segmentStartForTime(currentTime, segments) {
	if (currentTime == null) return void 0;
	return segments?.find(({ startTime, duration }, i, all) => currentTime >= startTime && (currentTime < startTime + duration || i === all.length - 1))?.startTime;
}
function getSegmentsToLoad(segments, bufferedSegments, currentTime, config = DEFAULT_FORWARD_BUFFER_CONFIG) {
	if (segments.length === 0) return [];
	const targetTime = currentTime + config.bufferDuration;
	const bufferedStartTimes = new Set(bufferedSegments.map((seg) => seg.startTime));
	return segments.filter((seg) => {
		const segmentEnd = seg.startTime + seg.duration;
		const isInRange = seg.startTime < targetTime && segmentEnd > currentTime;
		const isNotBuffered = !bufferedStartTimes.has(seg.startTime);
		return isInRange && isNotBuffered;
	});
}

//#endregion
//#region ../spf/dist/dev/media/types/index.js
/**
* Floating-point tolerance for matching segments by `startTime`. Two
* segments are considered the same position when
* `Math.abs(a.startTime - b.startTime) < SEGMENT_TIME_EPSILON`. Used by
* the source-buffer dedup and segment-loader quality-aware filter to
* tolerate sub-millisecond drift in segment timestamps across multiple
* playlists / quality levels.
*/
const SEGMENT_TIME_EPSILON = 1e-4;
function isResolvedTrack(track) {
	return "segments" in track;
}
/**
* Check if a presentation has duration (at least one track resolved).
* Narrows type to include required duration.
*/
function hasPresentationDuration(presentation) {
	return presentation.duration !== void 0;
}
/**
* Narrows a `MaybeResolvedPresentation` to a fully resolved `Presentation`.
*
* A presentation is resolved once `resolvePresentation` has parsed the
* manifest and populated both `id` and `selectionSets`. Both must be
* present — a partial value with only one of them isn't usable, and
* letting it through would have downstream behaviors crash when they
* access `selectionSets`.
*/
function isResolvedPresentation(presentation) {
	return presentation !== void 0 && presentation.id !== void 0 && presentation.selectionSets !== void 0;
}

//#endregion
//#region ../spf/dist/dev/media/utils/tracks.js
/**
* Get the tracks of the given type from a presentation's first switching set.
*
* Returns `[]` when the presentation is unresolved, when no selection set of
* `type` exists, or when its first switching set is empty. Returned tracks may
* be partially resolved (URL only) or fully resolved (with segments) — callers
* narrow as needed.
*
* The "first switching set" assumption matches the rest of the codebase
* (HLS typically has one switching set per type); multi-group / multi-period
* support would generalize this.
*/
function getTracksByType(presentation, type) {
	return presentation.selectionSets?.find(({ type: t }) => t === type)?.switchingSets[0]?.tracks ?? [];
}
/**
* Find a track of the given type and id within a presentation.
*
* Returns the matching track from the first switching set of the matching
* selection set, or `undefined` if either is missing. The returned track may
* be partially resolved (URL only) or fully resolved (with segments) — callers
* narrow as needed.
*/
function findTrack(presentation, type, trackId) {
	return getTracksByType(presentation, type).find(({ id }) => id === trackId);
}
/**
* Find a track by id across all selection sets in a presentation, without
* knowing its type up front. Used when the caller has a track id obtained
* from a downstream consumer (e.g. `SourceBufferActor.initTrackId`) and
* needs to locate the corresponding track in the presentation.
*
* Track ids are unique within a presentation per the HLS spec; the first
* match wins.
*/
function findTrackById(presentation, trackId) {
	for (const selectionSet of presentation.selectionSets ?? []) {
		const track = selectionSet.switchingSets[0]?.tracks.find(({ id }) => id === trackId);
		if (track) return track;
	}
}
/**
* Find a text track of the given id within a presentation and narrow it to
* the fully-resolved `TextTrack` shape (segments populated). Returns
* `undefined` if no track matches the id, the matching track isn't a text
* track, or it hasn't been resolved yet.
*
* The segments-non-empty check stays at the call site — a resolved track
* with zero segments is a valid state, distinct from "ready to load."
*/
function findResolvedTextTrack(presentation, trackId) {
	if (!presentation || !trackId) return void 0;
	const track = findTrack(presentation, "text", trackId);
	if (!track || track.type !== "text" || !isResolvedTrack(track)) return void 0;
	return track;
}
function findResolvedVideoTrack(presentation, trackId) {
	if (!presentation || !trackId) return void 0;
	const track = findTrack(presentation, "video", trackId);
	if (!track || track.type !== "video" || !isResolvedTrack(track)) return void 0;
	return track;
}
function findResolvedAudioTrack(presentation, trackId) {
	if (!presentation || !trackId) return void 0;
	const track = findTrack(presentation, "audio", trackId);
	if (!track || track.type !== "audio" || !isResolvedTrack(track)) return void 0;
	return track;
}
/**
* Whether a track carries a non-empty `codecs` array. Both partially-
* resolved and fully-resolved tracks may carry codecs — they come from
* the multivariant playlist's `EXT-X-STREAM-INF` line, not from the
* per-type media playlist — so this works at either resolution stage.
*
* `TextTrack` doesn't declare a `codecs` field; the `'codecs' in track`
* check narrows it out for the false branch.
*/
function hasCodecs(track) {
	return !!track && "codecs" in track && !!track.codecs?.length;
}
/**
* Updates a track within a presentation (immutably). Generic — works for
* video, audio, or text tracks.
*/
function updateTrackInPresentation(presentation, resolvedTrack) {
	const trackId = resolvedTrack.id;
	return {
		...presentation,
		selectionSets: presentation.selectionSets.map((selectionSet) => ({
			...selectionSet,
			switchingSets: selectionSet.switchingSets.map((switchingSet) => ({
				...switchingSet,
				tracks: switchingSet.tracks.map((track) => track.id === trackId ? resolvedTrack : track)
			}))
		}))
	};
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/track-types.js
/**
* **Per-type config bundles for per-type behavior variants.**
*
* Each `*_TYPE_CONFIG` constant names the per-type slot keys that
* per-type-variant setup-shape helpers consume. Variants reference the
* shared constant at their `defineBehavior` setup body rather than
* constructing the config inline, so the per-type identity (which state
* slot, which context slot, which discriminant) lives in one place per
* type.
*
* Variants spread their composition-time `config` over the defaults so
* engines can layer composition-supplied additions on top (e.g. a
* custom fetch closure, a non-default segment resolver). Defaults
* first, engine config second.
*
* Helpers continue to consume their slice via the existing typed-key
* generics (`<K extends SelectedTrackKey>` etc.); the extra fields on
* the config object (carrying facets other helpers care about) are fine
* under structural typing.
*/
/**
* Type-identity bundle for video-track-typed behaviors.
*
* @see setupVideoBufferActors, loadVideoSegments
*/
const VIDEO_TYPE_CONFIG = {
	type: "video",
	selectedKey: "selectedVideoTrackId",
	actorKey: "videoBufferActor",
	loaderKey: "videoSegmentLoaderActor"
};
/**
* Type-identity bundle for audio-track-typed behaviors.
*
* @see setupAudioBufferActors, loadAudioSegments
*/
const AUDIO_TYPE_CONFIG = {
	type: "audio",
	selectedKey: "selectedAudioTrackId",
	actorKey: "audioBufferActor",
	loaderKey: "audioSegmentLoaderActor"
};
/**
* Type-identity bundle for text-track-typed behaviors. Text tracks have
* no `SourceBufferActor` (MSE doesn't apply), so this bundle omits
* `actorKey`. The `loaderKey` points at the text-track-segment-loader
* actor (a `MessageActor` with continue-vs-preempt semantics, parallel
* to v/a's `SegmentLoaderActor`).
*
* @see loadTextTrackSegments, setupTextTrackActors
*/
const TEXT_TYPE_CONFIG = {
	type: "text",
	selectedKey: "selectedTextTrackId",
	loaderKey: "textTrackSegmentLoaderActor"
};

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/load-segments.js
/**
* **Per-type segment loading dispatch.** Per available track type (video /
* audio / text), reads the per-type segment-loader actor from context and
* sends typed `'load'` messages whenever a meaningful loading condition
* changes (selected track, current time crossing a segment boundary).
*
* Loader-actor lifecycle is owned upstream:
* - Video: `setupVideoBufferActors`
* - Audio: `setupAudioBufferActors`
* - Text: `setupTextTrackActors`
*
* This behavior is pure-consumer: it reads `context[loaderKey]` and
* dispatches typed messages via the variant's per-type `findResolvedTrack`
* resolver.
*
* # Load modes as reactor states
*
* Four states encode the load-gating policy directly:
*
* - `'preconditions-unmet'` — no loader actor in context, or the selected
*   track hasn't resolved.
* - `'dormant'` — `preload === 'none' && !loadActivated`. Nothing fires.
* - `'metadata-only'` — `!loadActivated && preload !== 'auto' && preload !== 'none'`.
*   Fires an init-segment-only `load` message **once on entry**. The
*   variant's loader actor decides what to do — v/a's actor fetches the
*   init segment; text's actor no-ops (no init concept).
* - `'full-range'` — `loadActivated || preload === 'auto'`. Effect re-fires
*   on selected-track change and on segment-boundary crossing.
*
* # Per-type parameterization (inference-driven)
*
* The helper is generic over `Track` — the resolved-track type. Each
* variant supplies its own `findResolvedTrack` resolver via config; TS
* infers `Track` from the resolver's return type. The loader signal's
* value type is constrained to `SegmentLoaderLike<Track>` — anything with
* a `send` method accepting the `Track`-parameterized message. Concrete
* actor types (`SegmentLoaderActor`, `TextTrackSegmentLoaderActor`)
* satisfy this via function-parameter contravariance: an actor whose
* `.send` accepts a wider track type is assignable to a slot expecting a
* narrower track type.
*
* No widening of actor message types is needed; no casts inside the
* helper. Per-variant wiring (right loader paired with right resolver)
* is enforced at the variant call site.
*/
/**
* Specialization helper. Generic over `Track` (inferred from
* `findResolvedTrack`'s return type). The loader value type is
* constrained to `SegmentLoaderLike<Track>` — concrete actor types
* (whose `.send` accepts a wider track union) satisfy this via
* function-parameter contravariance.
*
* Tracks that the helper handles must have a `segments` field — used by
* `segmentBoundarySignal` to compute the load-anchor boundary. Each
* variant's resolver narrows to the right resolved-track shape.
*/
function setupSegmentLoading({ state, context, config }) {
	const { selectedKey, loaderKey, findResolvedTrack } = config;
	const bufferDuration = config.forwardBuffer?.bufferDuration ?? DEFAULT_FORWARD_BUFFER_CONFIG.bufferDuration;
	const selectedTrack = computed(() => findResolvedTrack(state.presentation.get(), state[selectedKey].get()));
	const segmentBoundarySignal = computed(() => {
		const track = selectedTrack.get();
		if (!track) return void 0;
		return segmentStartForTime(state.currentTime.get() ?? 0, track.segments);
	});
	const derivedStateSignal = computed(() => {
		if (!context[loaderKey].get() || !selectedTrack.get()) return "preconditions-unmet";
		if (state.loadActivated.get() || state.preload.get() === "auto") return "full-range";
		if (state.preload.get() === "none") return "dormant";
		return "metadata-only";
	});
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			dormant: {},
			"metadata-only": { entry: () => {
				const track = selectedTrack.get();
				context[loaderKey].get().send({
					type: "load",
					track
				});
			} },
			"full-range": { effects: () => {
				const track = selectedTrack.get();
				segmentBoundarySignal.get();
				const currentTime = peek(state.currentTime) ?? 0;
				peek(context[loaderKey]).send({
					type: "load",
					track,
					range: {
						start: currentTime,
						end: currentTime + bufferDuration
					}
				});
			} }
		}
	});
}
const VIDEO_SEGMENT_LOADING_CONFIG = {
	...VIDEO_TYPE_CONFIG,
	findResolvedTrack: findResolvedVideoTrack
};
const AUDIO_SEGMENT_LOADING_CONFIG = {
	...AUDIO_TYPE_CONFIG,
	findResolvedTrack: findResolvedAudioTrack
};
const TEXT_SEGMENT_LOADING_CONFIG = {
	...TEXT_TYPE_CONFIG,
	findResolvedTrack: findResolvedTextTrack
};
const loadVideoSegments = defineBehavior({
	stateKeys: [
		"presentation",
		"preload",
		"currentTime",
		"loadActivated",
		"selectedVideoTrackId"
	],
	contextKeys: ["videoSegmentLoaderActor"],
	setup: ({ state, context, config = {} }) => setupSegmentLoading({
		state,
		context,
		config: {
			...VIDEO_SEGMENT_LOADING_CONFIG,
			...config
		}
	})
});
const loadAudioSegments = defineBehavior({
	stateKeys: [
		"presentation",
		"preload",
		"currentTime",
		"loadActivated",
		"selectedAudioTrackId"
	],
	contextKeys: ["audioSegmentLoaderActor"],
	setup: ({ state, context, config = {} }) => setupSegmentLoading({
		state,
		context,
		config: {
			...AUDIO_SEGMENT_LOADING_CONFIG,
			...config
		}
	})
});
const loadTextTrackSegments = defineBehavior({
	stateKeys: [
		"presentation",
		"preload",
		"currentTime",
		"loadActivated",
		"selectedTextTrackId"
	],
	contextKeys: ["textTrackSegmentLoaderActor"],
	setup: ({ state, context, config = {} }) => setupSegmentLoading({
		state,
		context,
		config: {
			...TEXT_SEGMENT_LOADING_CONFIG,
			...config
		}
	})
});

//#endregion
//#region ../spf/dist/dev/core/actors/create-transition-actor.js
/**
* Creates a reducer-shaped actor from an initial context and a reducer function.
*
* The reducer receives the current context and a message and returns the next
* context. Returning the same reference (by identity) skips the signal update —
* so early-returning `context` unchanged is both the no-op and the optimization.
*
* Side effects (e.g. DOM mutations) may be performed inside the reducer.
* They run synchronously before the signal is updated.
*
* @example
* const actor = createTransitionActor(
*   { count: 0 },
*   (context, message: { type: 'increment' }) => ({ count: context.count + 1 })
* );
*/
function createTransitionActor(initialContext, reducer) {
	const { snapshotSignal, getState, transition } = createMachineCore({
		value: "active",
		context: initialContext
	});
	const getContext = () => untrack(() => snapshotSignal.get().context);
	const setContext = (context) => update(snapshotSignal, { context });
	return {
		get snapshot() {
			return snapshotSignal;
		},
		send(message) {
			if (getState() === "destroyed") return;
			const context = getContext();
			const newContext = reducer(context, message);
			if (newContext !== context) setContext(newContext);
		},
		destroy() {
			if (getState() === "destroyed") return;
			transition("destroyed");
		}
	};
}

//#endregion
//#region ../spf/dist/dev/playback/actors/dom/text-tracks.js
function isDuplicateCue(cue, existing) {
	return existing.some((r) => r.startTime === cue.startTime && r.endTime === cue.endTime && r.text === cue.text);
}
/** TextTrack actor: wraps all text tracks on a media element, owns cue operations. */
function createTextTracksActor(mediaElement) {
	return createTransitionActor({
		loaded: {},
		segments: {}
	}, (context, message) => {
		if (message.type === "clear") return {
			loaded: {},
			segments: {}
		};
		const { meta, cues } = message;
		const { trackId, id: segmentId, startTime, duration } = meta;
		const textTrack = Array.from(mediaElement.textTracks).find((t) => t.id === trackId);
		if (!textTrack) return context;
		const existingCues = context.loaded[trackId] ?? [];
		const existingSegments = context.segments[trackId] ?? [];
		const prunedCues = cues.filter((cue) => !isDuplicateCue(cue, existingCues));
		const segmentAlreadyLoaded = existingSegments.some((s) => s.id === segmentId);
		if (prunedCues.length === 0 && segmentAlreadyLoaded) return context;
		for (const cue of prunedCues) textTrack.addCue(cue);
		return {
			...context,
			loaded: {
				...context.loaded,
				[trackId]: [...existingCues, ...prunedCues]
			},
			segments: segmentAlreadyLoaded ? context.segments : {
				...context.segments,
				[trackId]: [...existingSegments, {
					id: segmentId,
					startTime,
					duration
				}]
			}
		};
	});
}

//#endregion
//#region ../spf/dist/dev/core/tasks/task.js
/**
* Generic reusable task that wraps an async run function.
*
* Owns its own AbortController so it can always be aborted independently.
* Optionally composes an external AbortSignal so that a parent's cancellation
* propagates into the task's work without requiring the caller to track the
* task separately.
*
* Ordering guarantee: `value` is written before `status` transitions to `'done'`;
* `error` is written before `status` transitions to `'error'`. Any reader
* observing `status === 'done'` is guaranteed `value` is already present.
*/
var Task = class {
	id;
	#runFn;
	#abortController = new AbortController();
	#signal;
	#status = "pending";
	#value = void 0;
	#error = void 0;
	constructor(runFn, config) {
		this.#runFn = runFn;
		const rawId = config?.id;
		this.id = typeof rawId === "function" ? rawId() : rawId ?? generateId();
		this.#signal = config?.signal ? anyAbortSignal([this.#abortController.signal, config.signal]) : this.#abortController.signal;
	}
	get status() {
		return this.#status;
	}
	get value() {
		return this.#value;
	}
	get error() {
		return this.#error;
	}
	async run() {
		this.#status = "running";
		try {
			const result = await this.#runFn(this.#signal);
			this.#value = result;
			this.#status = "done";
			return result;
		} catch (e) {
			this.#error = e;
			this.#status = "error";
			throw e;
		}
	}
	abort() {
		this.#abortController.abort();
	}
};
/**
* Runs tasks concurrently, deduplicated by task id.
*
* If a task with a given id is already in flight, subsequent schedule() calls
* for that id are silently ignored until the first completes. Tasks are stored
* so abortAll() can cancel any in-flight work (e.g. on engine cleanup).
*/
var ConcurrentRunner = class {
	#pending = /* @__PURE__ */ new Map();
	#settled = Promise.resolve();
	#resolveSettled = null;
	#destroyed = false;
	schedule(task) {
		if (this.#destroyed) return Promise.resolve();
		const existing = this.#pending.get(task.id);
		if (existing) return existing.promise;
		if (this.#pending.size === 0) this.#settled = new Promise((resolve) => {
			this.#resolveSettled = resolve;
		});
		const promise = task.run();
		promise.catch(() => {});
		const cleanup = () => {
			this.#pending.delete(task.id);
			if (this.#pending.size === 0) {
				this.#resolveSettled?.();
				this.#resolveSettled = null;
			}
		};
		promise.then(cleanup, cleanup);
		this.#pending.set(task.id, {
			task,
			promise
		});
		return promise;
	}
	/**
	* Registers a callback to fire when all currently in-flight tasks settle.
	* If the runner is already idle, the callback is never called. If abortAll()
	* is called before the batch settles, the callback is superseded and silently
	* dropped — no stale callbacks, no generation token required by the caller.
	*/
	whenSettled(callback) {
		if (this.#pending.size === 0) return;
		const captured = this.#settled;
		captured.then(() => {
			if (this.#settled !== captured) return;
			callback();
		}, () => {});
	}
	abortAll() {
		for (const { task } of this.#pending.values()) task.abort();
		this.#pending.clear();
		this.#resolveSettled?.();
		this.#resolveSettled = null;
		this.#settled = Promise.resolve();
	}
	destroy() {
		this.#destroyed = true;
		this.abortAll();
	}
};
/**
* Runs tasks one at a time in submission order.
*
* Each schedule() call returns a Promise that resolves or rejects with the
* task's result when it is eventually executed. Tasks wait in queue until the
* prior task completes.
*
* Serialization is achieved by chaining each task's run() onto the tail of a
* shared promise chain — no explicit queue or drain loop needed.
*
* abortAll() aborts all pending (not yet started) tasks and the currently
* in-flight task. Pending tasks still run briefly but receive an aborted
* signal and are expected to exit early.
*/
var SerialRunner = class {
	#chain = Promise.resolve();
	#pending = /* @__PURE__ */ new Set();
	#current = null;
	#destroyed = false;
	schedule(task) {
		if (this.#destroyed) return Promise.resolve();
		const t = task;
		this.#pending.add(t);
		const result = this.#chain.then(() => {
			this.#pending.delete(t);
			this.#current = t;
			return task.run();
		}).finally(() => {
			this.#current = null;
		});
		this.#chain = result.then(() => {}, () => {});
		return result;
	}
	/**
	* A promise that resolves when all currently-scheduled tasks have settled.
	* Use the reference as a generation token: capture it after scheduling a
	* batch, then check identity in the resolution callback to detect whether
	* a subsequent abortAll() + new batch has superseded this one.
	*/
	get settled() {
		return this.#chain;
	}
	/**
	* Registers a callback to fire when all currently-pending tasks settle.
	* If the runner is already idle (no pending or running tasks), the callback
	* is never called. If new tasks are scheduled before the current batch
	* settles, the callback is superseded and silently dropped — no stale
	* callbacks, no generation token required by the caller.
	*/
	whenSettled(callback) {
		if (this.#pending.size === 0 && this.#current === null) return;
		const currentChain = this.#chain;
		currentChain.then(() => {
			if (this.#chain !== currentChain) return;
			callback();
		}, () => {});
	}
	/** Aborts and clears queued tasks without touching the in-flight task. */
	abortPending() {
		for (const task of this.#pending) task.abort();
		this.#pending.clear();
	}
	abortAll() {
		this.abortPending();
		this.#current?.abort();
	}
	destroy() {
		this.#destroyed = true;
		this.abortAll();
	}
};

//#endregion
//#region ../spf/dist/dev/core/actors/create-machine-actor.js
/**
* Creates a message-driven actor from a declarative definition.
*
* The actor owns a reactive snapshot signal (state + context), an optional
* runner, and dispatches incoming messages to per-state handlers. `'destroyed'`
* is always the implicit terminal state — `destroy()` transitions there
* unconditionally and all subsequent `send()` calls are no-ops.
*
* When a state declares `onSettled`, the framework calls `runner.whenSettled()`
* after the handler returns. The runner owns the generation-token logic — if
* new tasks are scheduled before the current batch settles, the callback is
* automatically superseded.
*
* @example
* const actor = createMachineActor({
*   runner: () => new SerialRunner(),
*   initial: 'idle',
*   context: {},
*   states: {
*     idle: {
*       on: {
*         load: (msg, { transition, runner }) => {
*           segments.forEach(s => runner.schedule(new Task(...)));
*           transition('loading');
*         }
*       }
*     },
*     loading: {
*       onSettled: 'idle',
*       on: {
*         load: (msg, { runner }) => {
*           runner.abortAll();
*           segments.forEach(s => runner.schedule(new Task(...)));
*         }
*       }
*     }
*   }
* });
*/
function createMachineActor(def) {
	const runner = def.runner?.();
	const { snapshotSignal, getState, transition } = createMachineCore({
		value: def.initial,
		context: def.context
	});
	const getContext = () => untrack(() => snapshotSignal.get().context);
	const setContext = (context) => {
		update(snapshotSignal, { context });
	};
	return {
		get snapshot() {
			return snapshotSignal;
		},
		send(message) {
			const state = getState();
			if (state === "destroyed") return;
			const handler = def.states[state]?.on?.[message.type];
			if (!handler) return;
			handler(message, {
				context: getContext(),
				getContext,
				transition: (to) => transition(to),
				setContext,
				...runner ? { runner } : {}
			});
			const newState = getState();
			if (newState !== "destroyed") {
				const newStateDef = def.states[newState];
				if (newStateDef?.onSettled && runner) {
					const targetState = newStateDef.onSettled;
					runner.whenSettled(() => {
						if (getState() !== newState) return;
						transition(targetState);
					});
				}
			}
		},
		destroy() {
			if (getState() === "destroyed") return;
			runner?.destroy();
			transition("destroyed");
		}
	};
}

//#endregion
//#region ../spf/dist/dev/playback/actors/text-track-segment-loader.js
/**
* Loads text-track segments for a track and delegates cue management
* to a TextTracksActor. Mirrors the v/a `SegmentLoaderActor` shape (FSM
* with `idle` / `loading` and `inFlight*` context for continue-vs-preempt),
* adapted to text:
*
* - No init segment, no flush ops (text cues don't need eviction — they're
*   small and the playhead-relative window is enforced by the runtime).
* - Single in-flight identity (`inFlightSegmentId`) — text has only the
*   media-segment path, no init-segment path.
*
* Planning is done in the load handler on every incoming message:
* `getSegmentsToLoad` filters to the forward window, then the segments
* not already in `TextTracksActor`'s context are scheduled. When a new
* `load` arrives mid-run, the handler replans and either:
*
* - **Continues**: the in-flight segment is still in the new plan →
*   `abortPending` only, schedule the rest of the plan (minus the
*   in-flight item, which covers its slot).
* - **Preempts**: in-flight segment is no longer wanted (track switch,
*   large seek out of window) → `abortAll`, schedule the new plan
*   from scratch.
*
* The cue parser is injected so this factory is host-agnostic. A DOM
* host supplies a VTT parser backed by `<track>`/`TextTrack` APIs; a
* non-DOM host (worker, test fake, alternate runtime) supplies its own.
*/
function createTextTrackSegmentLoaderActor(textTracksActor, resolveSegment, config = {}) {
	const forwardBufferConfig = {
		...DEFAULT_FORWARD_BUFFER_CONFIG,
		...config.forwardBuffer
	};
	/**
	* Translate a load message into an ordered TextLoadTask list based on
	* committed actor state. In-flight awareness is handled separately in
	* the `loading` state's load handler.
	*
	* Metadata mode (no `range`) is a no-op for text — text tracks have
	* no init-segment concept, so there's nothing to load until a range
	* arrives via `'full-range'` dispatch.
	*/
	const planTasks = (message) => {
		const { track, range } = message;
		if (!range) return [];
		const trackId = track.id;
		const bufferedSegments = peek(textTracksActor.snapshot).context.segments[trackId] ?? [];
		return getSegmentsToLoad(track.segments, bufferedSegments, range.start, forwardBufferConfig).map((segment) => ({
			segment,
			trackId
		}));
	};
	/**
	* Wraps a TextLoadTask into a Task that fetches + dispatches `add-cues`.
	* Updates `inFlightSegmentId` around the fetch so the load handler can
	* make accurate continue/preempt decisions.
	*/
	const makeLoadTask = (op, { getContext, setContext }) => {
		return new Task(async (signal) => {
			if (signal.aborted) return;
			setContext({
				...getContext(),
				inFlightTrackId: op.trackId,
				inFlightSegmentId: op.segment.id
			});
			try {
				const cues = await resolveSegment(op.segment.url);
				if (signal.aborted) return;
				textTracksActor.send({
					type: "add-cues",
					meta: {
						trackId: op.trackId,
						id: op.segment.id,
						startTime: op.segment.startTime,
						duration: op.segment.duration
					},
					cues
				});
			} catch (error) {
				console.error("Failed to load text-track segment:", error);
			} finally {
				setContext({
					...getContext(),
					inFlightTrackId: null,
					inFlightSegmentId: null
				});
			}
		});
	};
	const scheduleAll = (tasks, ctx) => {
		for (const op of tasks) ctx.runner.schedule(makeLoadTask(op, ctx)).then(void 0, (e) => {
			if (e instanceof Error && e.name === "AbortError") return;
			console.error("Unexpected error in text-track segment loader:", e);
			ctx.runner.abortPending();
		});
	};
	return createMachineActor({
		runner: () => new SerialRunner(),
		initial: "idle",
		context: {
			inFlightTrackId: null,
			inFlightSegmentId: null
		},
		states: {
			idle: { on: { load: (msg, ctx) => {
				const tasks = planTasks(msg);
				if (tasks.length === 0) return;
				ctx.transition("loading");
				scheduleAll(tasks, ctx);
			} } },
			loading: {
				onSettled: "idle",
				on: { load: (msg, ctx) => {
					const { context, runner } = ctx;
					const tasks = planTasks(msg);
					if (context.inFlightTrackId !== null && context.inFlightSegmentId !== null && tasks.some((t) => t.trackId === context.inFlightTrackId && t.segment.id === context.inFlightSegmentId)) {
						runner.abortPending();
						scheduleAll(tasks.filter((t) => !(t.trackId === context.inFlightTrackId && t.segment.id === context.inFlightSegmentId)), ctx);
					} else {
						runner.abortAll();
						scheduleAll(tasks, ctx);
					}
				} }
			}
		}
	});
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/setup-text-track-actors.js
/**
* **Own the TextTracks actor pair for the current `mediaElement`.** When a
* `mediaElement` is in scope, creates the `TextTracksActor` (bound to the
* element's `textTracks`) and the `TextTrackSegmentLoaderActor` (bound to
* that actor + the injected cue resolver), and publishes both on
* `context`. On element identity change or behavior destroy, destroys both
* actors and clears the slots.
*
* Single-resource synchronous create/destroy driven by one signal — the
* simple-effect form is the right shape (per `behaviors.md` → "Where both
* shapes are legitimate": criterion 4b applies; sole writer of
* `textTracksActor` / `textTrackSegmentLoaderActor`, the effect's cleanup
* return handles destroy + slot clear structurally).
*
* Pairs with the `loadTextTrackSegments` behavior (a per-type variant of
* `setupSegmentLoading` in `load-segments.ts`), which only reads
* `textTrackSegmentLoaderActor`. The cue resolver is injected via
* `config` so this behavior owns the DOM-bound part of the text-track
* pipeline.
*/
function setupTextTrackActorsSetup({ context, config }) {
	return effect(() => {
		const mediaElement = context.mediaElement.get();
		if (!mediaElement) return;
		const textTracksActor = createTextTracksActor(mediaElement);
		const textTrackSegmentLoaderActor = createTextTrackSegmentLoaderActor(textTracksActor, config.resolveTextTrackSegment, { forwardBuffer: config.forwardBuffer });
		context.textTracksActor.set(textTracksActor);
		context.textTrackSegmentLoaderActor.set(textTrackSegmentLoaderActor);
		return () => {
			textTracksActor.destroy();
			textTrackSegmentLoaderActor.destroy();
			context.textTracksActor.set(void 0);
			context.textTrackSegmentLoaderActor.set(void 0);
		};
	});
}
const setupTextTrackActors = defineBehavior({
	stateKeys: [],
	contextKeys: [
		"mediaElement",
		"textTracksActor",
		"textTrackSegmentLoaderActor"
	],
	setup: setupTextTrackActorsSetup
});

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/track-current-time.js
/**
* Mirror `mediaElement.currentTime` into reactive state. Listens for:
* - `timeupdate` — fires during playback (~4 Hz)
* - `seeking` — fires when a seek begins; per spec, `currentTime` is already
*   at the new position when this event dispatches, so buffer management can
*   react immediately rather than waiting for `timeupdate`, which does not
*   fire while paused.
*
* Also syncs immediately when a media element becomes available.
*
* When no media element is attached, writes `config.defaultCurrentTime`
* (default-default `0`, matching the HTMLMediaElement spec) so consumers
* always see a defined position. Read-only mirror; does not push
* `state.currentTime` back to the element.
*/
function trackCurrentTimeSetup({ state, context, config }) {
	const defaultCurrentTime = config?.defaultCurrentTime ?? 0;
	return effect(() => {
		const mediaElement = context.mediaElement.get();
		if (!mediaElement) {
			state.currentTime.set(defaultCurrentTime);
			return;
		}
		const sync = () => state.currentTime.set(mediaElement.currentTime);
		sync();
		const removeTimeupdate = listen(mediaElement, "timeupdate", sync);
		const removeSeeking = listen(mediaElement, "seeking", sync);
		return () => {
			removeTimeupdate();
			removeSeeking();
		};
	});
}
const trackCurrentTime = defineBehavior({
	stateKeys: ["currentTime"],
	contextKeys: ["mediaElement"],
	setup: trackCurrentTimeSetup
});

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/track-load-triggers.js
/**
* Slot-driven FSM. The slot `loadActivated` is the canonical externally-
* observable state; `deriveState` reads it (and the preconditions) to
* derive the local FSM state. The slot is *both* the state and the data —
* no separate internal bookkeeping.
*
* ```
* 'preconditions-unmet' ⟷ 'monitoring' ⟷ 'load-active'
*
* preconditions-unmet → monitoring        element + URL appear
* monitoring          → load-active       slot flips true (listener fires
*                                         or external write)
* load-active         → monitoring        within-state cleanup resets slot
*                                         (URL or element identity change)
* any                 → preconditions-unmet  element or URL → undefined
*
* any state → destroying → destroyed       on destroy()
* ```
*/
function deriveState$5(presentation, mediaElement, loadActivated) {
	if (!mediaElement || !presentation?.url) return "preconditions-unmet";
	if (loadActivated) return "load-active";
	return "monitoring";
}
/**
* Track preload-overriding events per source.
*
* Writes `state.loadActivated = true` the first time a `play` or `seeking`
* event fires on the attached media element for the current source — or
* immediately on entry if the element is already in such a state
* (`!el.paused` or `el.seeking`), mirroring autoplay / native-controls /
* direct-DOM-`play()` scenarios.
*
* Sticky-true *within a source identity*: subsequent play/pause/seek
* cycles don't flip back. Source identity = (mediaElement, presentation
* URL). Either changing — including direct in-place swap with no
* `undefined` intermediate — resets the slot to `false`.
*
* Multi-writer with `hls/adapter.ts:play()` (which writes `true` directly
* on programmatic play) is intentional — different domains. The adapter
* records programmatic intent; this behavior is the DOM-side observer.
* Pre-existing `true` writes are honored because `deriveState` reads the
* slot — a `true` value routes directly to `'load-active'` without
* entering `'monitoring'`.
*
* @example
* const reactor = trackLoadTriggers.setup({ state, context });
* // later:
* reactor.destroy();
*/
function trackLoadTriggersSetup({ state, context }) {
	const derivedStateSignal = computed(() => deriveState$5(state.presentation.get(), context.mediaElement.get(), state.loadActivated.get()));
	const urlSignal = computed(() => state.presentation.get()?.url);
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			monitoring: { effects: () => {
				const el = context.mediaElement.get();
				const setLoadActivated = () => state.loadActivated.set(true);
				if (!el.paused || el.seeking) {
					setLoadActivated();
					return;
				}
				const cleanupPlay = listen(el, "play", setLoadActivated);
				const cleanupSeeking = listen(el, "seeking", setLoadActivated);
				return () => {
					cleanupPlay();
					cleanupSeeking();
				};
			} },
			"load-active": { effects: () => {
				context.mediaElement.get();
				urlSignal.get();
				return () => state.loadActivated.set(false);
			} }
		}
	});
}
const trackLoadTriggers = defineBehavior({
	stateKeys: ["loadActivated", "presentation"],
	contextKeys: ["mediaElement"],
	setup: trackLoadTriggersSetup
});

//#endregion
//#region ../spf/dist/dev/media/dom/text/text-track-slots.js
/**
* SPF-owned `<track>` selector. Each slot created by
* `addSubtitlesTracksToMedia` carries this attribute so reads and removals can
* filter SPF-owned tracks from host-page-owned ones.
*/
const SPF_TRACK_SELECTOR = "track[data-src-track]";
/**
* Allocate text-track slots on `mediaElement` for each model track by creating
* and appending `<track>` children. Marks each element with `data-src-track`
* so it can be distinguished from `<track>` children the host page added
* directly — used by `getShowingSubtitlesTrackFromMedia` and
* `removeAllSubtitlesTracksFromMedia` to scope their reads/removals to
* SPF-owned slots. The spec has no `removeTextTrack` API, so creating
* `<track>` elements is the only mechanism for adding *and* removing entries
* to `mediaElement.textTracks`.
*/
function addSubtitlesTracksToMedia(mediaElement, modelTextTracks) {
	for (const modelTrack of modelTextTracks) {
		const el = document.createElement("track");
		el.id = modelTrack.id;
		el.kind = modelTrack.kind;
		el.label = modelTrack.label;
		el.toggleAttribute("data-src-track", true);
		if (modelTrack.language) el.srclang = modelTrack.language;
		if (modelTrack.default) el.default = true;
		mediaElement.appendChild(el);
	}
}
/**
* Return the SPF-owned subtitle/caption `TextTrack` currently in `'showing'`
* mode, or `undefined` if none. Restricts the search to slots created by
* `addSubtitlesTracksToMedia` (via the `data-src-track` selector) so a showing
* track that the host page added directly is ignored — SPF selection only
* mirrors tracks it owns.
*/
function getShowingSubtitlesTrackFromMedia(mediaElement) {
	const elements = mediaElement.querySelectorAll(SPF_TRACK_SELECTOR);
	for (const el of elements) {
		const track = el.track;
		if (track.mode === "showing" && (track.kind === "subtitles" || track.kind === "captions")) return track;
	}
}
/**
* Remove every SPF-owned `<track>` child from `mediaElement` (those tagged
* with `data-src-track` by `addSubtitlesTracksToMedia`). `<track>` elements
* the host page added directly are left in place.
*/
function removeAllSubtitlesTracksFromMedia(mediaElement) {
	const elements = mediaElement.querySelectorAll(SPF_TRACK_SELECTOR);
	for (const el of elements) el.remove();
}
/**
* Apply a selection to a `TextTrackList` by setting each subtitle/caption
* track's `mode` to `'showing'` if its `id` matches `selectedId` and
* `'disabled'` otherwise. Tracks of other kinds (chapters, metadata,
* descriptions) are left untouched — they may be owned by the host page.
*/
function syncTextTrackModes(textTracks, selectedId) {
	for (let i = 0; i < textTracks.length; i++) {
		const track = textTracks[i];
		if (track.kind !== "subtitles" && track.kind !== "captions") continue;
		track.mode = track.id === selectedId ? "showing" : "disabled";
	}
}

//#endregion
//#region ../spf/dist/dev/media/hls/parse-attributes.js
/**
* Parse HLS attribute list from a tag line.
* Handles both quoted and unquoted values.
*/
function parseAttributeList(line) {
	const attributes = /* @__PURE__ */ new Map();
	for (const match of line.matchAll(/([A-Z0-9-]+)=(?:"([^"]*)"|([^,]*))/g)) {
		const key = match[1];
		const value = match[2] ?? match[3] ?? "";
		if (key) attributes.set(key, value);
	}
	return attributes;
}
/**
* Parse RESOLUTION attribute value (WIDTHxHEIGHT).
*/
function parseResolution(value) {
	const match = /^(\d+)x(\d+)$/.exec(value);
	if (!match) return null;
	return {
		width: Number.parseInt(match[1], 10),
		height: Number.parseInt(match[2], 10)
	};
}
/**
* Parse FRAME-RATE attribute to rational frame rate.
*/
function parseFrameRate(value) {
	const fps = Number.parseFloat(value);
	if (Number.isNaN(fps) || fps <= 0) return void 0;
	if (Math.abs(fps - 23.976) < .01) return {
		frameRateNumerator: 24e3,
		frameRateDenominator: 1001
	};
	if (Math.abs(fps - 29.97) < .01) return {
		frameRateNumerator: 3e4,
		frameRateDenominator: 1001
	};
	if (Math.abs(fps - 59.94) < .01) return {
		frameRateNumerator: 6e4,
		frameRateDenominator: 1001
	};
	if (fps % 1 === 0) return { frameRateNumerator: Math.round(fps) };
	return { frameRateNumerator: Math.round(fps) };
}
/**
* Parse CODECS attribute into separate video and audio codecs.
*/
function parseCodecs(codecs) {
	const parts = codecs.split(",").map((s) => s.trim());
	const result = {};
	for (const codec of parts) if (codec.startsWith("avc1.") || codec.startsWith("hvc1.") || codec.startsWith("hev1.")) result.video = codec;
	else if (codec.startsWith("mp4a.")) result.audio = codec;
	return result;
}
/**
* Parse #EXTINF duration value.
*/
function parseExtInfDuration(value) {
	const durationPart = value.split(",")[0] ?? value;
	const duration = Number.parseFloat(durationPart);
	return Number.isNaN(duration) ? 0 : duration;
}
/**
* Parse BYTERANGE attribute value.
* Format: "length[@offset]"
* If offset is omitted, it continues from the previous byte range end.
*/
function parseByteRange(value, previousEnd) {
	const match = /^(\d+)(?:@(\d+))?$/.exec(value);
	if (!match) return null;
	const length = Number.parseInt(match[1], 10);
	if (Number.isNaN(length)) return null;
	let start;
	if (match[2] !== void 0) {
		start = Number.parseInt(match[2], 10);
		if (Number.isNaN(start)) return null;
	} else if (previousEnd !== void 0) start = previousEnd;
	else return null;
	return {
		start,
		end: start + length - 1
	};
}
/**
* Create AttributeList from raw attribute string.
*/
function createAttributeList(line) {
	const map = parseAttributeList(line);
	return {
		get(key) {
			return map.get(key);
		},
		getInt(key, defaultValue) {
			const value = map.get(key);
			if (value === void 0) return defaultValue;
			const parsed = Number.parseInt(value, 10);
			return Number.isNaN(parsed) ? defaultValue : parsed;
		},
		getFloat(key, defaultValue) {
			const value = map.get(key);
			if (value === void 0) return defaultValue;
			const parsed = Number.parseFloat(value);
			return Number.isNaN(parsed) ? defaultValue : parsed;
		},
		getBool(key) {
			return map.get(key) === "YES";
		},
		getResolution(key) {
			const value = map.get(key);
			if (!value) return void 0;
			return parseResolution(value) ?? void 0;
		},
		getFrameRate(key) {
			const value = map.get(key);
			if (!value) return void 0;
			return parseFrameRate(value);
		}
	};
}
/**
* Match a tag and extract its attributes.
* Returns null if the line doesn't match the tag.
*/
function matchTag(line, tag) {
	const prefix = `#${tag}:`;
	if (!line.startsWith(prefix)) return null;
	return createAttributeList(line.slice(prefix.length));
}

//#endregion
//#region ../spf/dist/dev/media/hls/resolve-url.js
/**
* Resolve a potentially relative URL against a base URL using native URL API.
*/
function resolveUrl(url, baseUrl) {
	return new URL(url, baseUrl).href;
}

//#endregion
//#region ../spf/dist/dev/media/hls/parse-multivariant.js
/**
* Parse HLS multivariant playlist into a Presentation.
*
* Returns Presentation with partially resolved tracks (no segment information).
* Tracks contain metadata from multivariant playlist (bandwidth, resolution, codecs)
* but segment information is added when media playlists are fetched.
*
* @param text - Raw playlist text content
* @param unresolved - Unresolved presentation (contains URL for base URL resolution)
* @returns Presentation with partially resolved tracks (duration is undefined)
*/
function parseMultivariantPlaylist(text, unresolved) {
	const baseUrl = unresolved.url;
	const lines = text.split(/\r?\n/);
	const streams = [];
	const audioRenditions = [];
	const subtitleRenditions = [];
	let pendingStreamInfo = null;
	for (const line of lines) {
		const trimmed = line.trim();
		if (!trimmed || trimmed.startsWith("#") && !trimmed.startsWith("#EXT")) continue;
		if (trimmed === "#EXTM3U" || trimmed.startsWith("#EXT-X-VERSION:") || trimmed.startsWith("#EXT-X-INDEPENDENT-SEGMENTS")) continue;
		const mediaAttrs = matchTag(trimmed, "EXT-X-MEDIA");
		if (mediaAttrs) {
			const type = mediaAttrs.get("TYPE");
			const groupId = mediaAttrs.get("GROUP-ID");
			const name = mediaAttrs.get("NAME");
			if (type === "AUDIO" && groupId && name) {
				const uri = mediaAttrs.get("URI");
				audioRenditions.push({
					groupId,
					name,
					language: mediaAttrs.get("LANGUAGE"),
					uri: uri ? resolveUrl(uri, baseUrl) : void 0,
					default: mediaAttrs.getBool("DEFAULT"),
					autoselect: mediaAttrs.getBool("AUTOSELECT")
				});
			}
			if (type === "SUBTITLES" && groupId && name) {
				const uri = mediaAttrs.get("URI");
				if (uri) subtitleRenditions.push({
					groupId,
					name,
					language: mediaAttrs.get("LANGUAGE"),
					uri: resolveUrl(uri, baseUrl),
					default: mediaAttrs.getBool("DEFAULT"),
					autoselect: mediaAttrs.getBool("AUTOSELECT"),
					forced: mediaAttrs.getBool("FORCED")
				});
			}
			continue;
		}
		const streamInfAttrs = matchTag(trimmed, "EXT-X-STREAM-INF");
		if (streamInfAttrs) {
			pendingStreamInfo = {
				bandwidth: streamInfAttrs.getInt("BANDWIDTH", 0),
				resolution: streamInfAttrs.getResolution("RESOLUTION"),
				codecs: streamInfAttrs.get("CODECS"),
				frameRate: streamInfAttrs.getFrameRate("FRAME-RATE"),
				audioGroupId: streamInfAttrs.get("AUDIO")
			};
			continue;
		}
		if (!trimmed.startsWith("#") && pendingStreamInfo) {
			streams.push({
				...pendingStreamInfo,
				uri: resolveUrl(trimmed, baseUrl)
			});
			pendingStreamInfo = null;
		}
	}
	const videoStreams = [];
	const audioOnlyStreams = [];
	for (const stream of streams) {
		if (!stream.codecs) {
			videoStreams.push(stream);
			continue;
		}
		const parsedCodecs = parseCodecs(stream.codecs);
		if (stream.codecs.split(",").length === 1) if (parsedCodecs.audio && !parsedCodecs.video) audioOnlyStreams.push(stream);
		else videoStreams.push(stream);
		else videoStreams.push(stream);
	}
	const videoTracks = videoStreams.map((stream) => {
		const codecs = stream.codecs ? parseCodecs(stream.codecs) : void 0;
		const track = {
			type: "video",
			id: generateId(),
			url: stream.uri,
			bandwidth: stream.bandwidth,
			mimeType: "video/mp4",
			codecs: []
		};
		if (stream.resolution?.width !== void 0) track.width = stream.resolution.width;
		if (stream.resolution?.height !== void 0) track.height = stream.resolution.height;
		if (codecs?.video) track.codecs = [codecs.video];
		if (stream.frameRate) track.frameRate = stream.frameRate;
		if (stream.audioGroupId) track.audioGroupId = stream.audioGroupId;
		return track;
	});
	const audioOnlyTracks = audioOnlyStreams.map((stream) => {
		const codecs = stream.codecs ? parseCodecs(stream.codecs) : void 0;
		return {
			type: "audio",
			id: generateId(),
			url: stream.uri,
			bandwidth: stream.bandwidth,
			mimeType: "audio/mp4",
			codecs: codecs?.audio ? [codecs.audio] : [],
			groupId: stream.audioGroupId || "default",
			name: "Default",
			sampleRate: 48e3,
			channels: 2
		};
	});
	const audioTracks = [...audioRenditions.map((rendition) => {
		let audioCodecs;
		for (const stream of streams) if (stream.audioGroupId === rendition.groupId && stream.codecs) {
			const codecs = parseCodecs(stream.codecs);
			if (codecs.audio) {
				audioCodecs = [codecs.audio];
				break;
			}
		}
		const track = {
			type: "audio",
			id: generateId(),
			url: rendition.uri ?? "",
			groupId: rendition.groupId,
			name: rendition.name,
			mimeType: "audio/mp4",
			bandwidth: 0,
			sampleRate: 48e3,
			channels: 2,
			codecs: []
		};
		if (rendition.language) track.language = rendition.language;
		if (audioCodecs) track.codecs = audioCodecs;
		if (rendition.default) track.default = rendition.default;
		if (rendition.autoselect) track.autoselect = rendition.autoselect;
		return track;
	}), ...audioOnlyTracks];
	const textTracks = subtitleRenditions.map((rendition) => {
		const track = {
			type: "text",
			id: generateId(),
			url: rendition.uri,
			groupId: rendition.groupId,
			label: rendition.name,
			kind: "subtitles",
			mimeType: "text/vtt",
			bandwidth: 0
		};
		if (rendition.language) track.language = rendition.language;
		if (rendition.default && rendition.autoselect) track.default = true;
		if (rendition.autoselect) track.autoselect = rendition.autoselect;
		if (rendition.forced) track.forced = rendition.forced;
		return track;
	});
	const selectionSets = [];
	if (videoTracks.length > 0) {
		const videoSwitchingSet = {
			id: generateId(),
			type: "video",
			tracks: videoTracks
		};
		const videoSelectionSet = {
			id: generateId(),
			type: "video",
			switchingSets: [videoSwitchingSet]
		};
		selectionSets.push(videoSelectionSet);
	}
	if (audioTracks.length > 0) {
		const audioSwitchingSet = {
			id: generateId(),
			type: "audio",
			tracks: audioTracks
		};
		const audioSelectionSet = {
			id: generateId(),
			type: "audio",
			switchingSets: [audioSwitchingSet]
		};
		selectionSets.push(audioSelectionSet);
	}
	if (textTracks.length > 0) {
		const textSwitchingSet = {
			id: generateId(),
			type: "text",
			tracks: textTracks
		};
		const textSelectionSet = {
			id: generateId(),
			type: "text",
			switchingSets: [textSwitchingSet]
		};
		selectionSets.push(textSelectionSet);
	}
	return {
		id: generateId(),
		url: unresolved.url,
		startTime: 0,
		selectionSets
	};
}

//#endregion
//#region ../spf/dist/dev/media/utils/track-selection.js
/**
* Map track type to selected track ID property key in state.
*/
const SelectedTrackIdKeyByType = {
	video: "selectedVideoTrackId",
	audio: "selectedAudioTrackId",
	text: "selectedTextTrackId"
};
/**
* Get selected track from state by type.
* Returns properly typed track (partially or fully resolved) or undefined.
* Type parameter T is inferred from the type argument.
*
* @example
* const videoTrack = getSelectedTrack(state, 'video');
* if (videoTrack && isResolvedTrack(videoTrack)) {
*   // videoTrack is VideoTrack
* }
*/
function getSelectedTrack(state, type) {
	const { presentation } = state;
	if (!presentation?.selectionSets) return void 0;
	const trackId = state[SelectedTrackIdKeyByType[type]];
	return presentation.selectionSets.find(({ type: selectionSetType }) => selectionSetType === type)?.switchingSets[0]?.tracks.find(({ id }) => id === trackId);
}
/**
* Returns the duration of the first resolved selected track, preferring
* video over audio. A track is "resolved" once its media playlist has been
* parsed (per {@link isResolvedTrack}). Returns `undefined` if neither
* selected track is resolved.
*/
function getResolvedSelectedTrackDuration(state) {
	if (state.selectedVideoTrackId) {
		const video = getSelectedTrack(state, "video");
		if (video && isResolvedTrack(video)) return video.duration;
	}
	if (state.selectedAudioTrackId) {
		const audio = getSelectedTrack(state, "audio");
		if (audio && isResolvedTrack(audio)) return audio.duration;
	}
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/calculate-presentation-duration.js
/**
* **Populate presentation duration via a config-supplied resolver.**
*
* Once a presentation is in place and its `duration` is still undefined,
* calls `config.resolveDuration(state)` whenever a tracked slot changes. The
* resolver decides what counts as "duration is now derivable" for the variant
* in play; the behavior itself stays variant-agnostic:
*
* - **VoD** — derive from the first resolved selected track's `duration`
*   (video preferred, audio fallback). The HLS engine wires
*   `getResolvedSelectedTrackDuration` from `media/utils/track-selection.ts`
*   as the default. Audio-only falls out of the same resolver naturally
*   (no video selected → audio is the first resolved track).
* - **Live** — return `Number.POSITIVE_INFINITY` once the presentation is
*   established as live. This is the MSE-spec value for `mediaSource.duration`
*   under live playback; downstream `updateMediaSourceDuration` propagates it through.
*
* Validation: writes whatever the resolver returns as long as it's a positive
* number, including `Infinity`. `undefined` / `NaN` / `<= 0` are skipped.
*
* Fires at most once per presentation — an already-set `duration` is never
* overwritten, and the next reset arrives structurally when a new
* (unresolved) presentation replaces the current one. The resolver may
* return `undefined` while duration is still indeterminate; subsequent
* tracked-slot changes re-run the effect until the resolver commits a value.
*
* Downstream of `resolveVideoTrack` / `resolveAudioTrack`; upstream of
* `updateMediaSourceDuration` (which writes the value through to `mediaSource.duration`).
*/
function calculatePresentationDurationSetup({ state, config }) {
	return effect(() => {
		const presentation = state.presentation.get();
		if (!presentation || presentation.duration !== void 0) return;
		const duration = config.resolveDuration(untrack(() => snapshot(state)));
		if (duration === void 0 || Number.isNaN(duration) || duration <= 0) return;
		update(state.presentation, (current) => current ? {
			...current,
			duration
		} : current);
	});
}
const calculatePresentationDuration = defineBehavior({
	stateKeys: [
		"presentation",
		"selectedVideoTrackId",
		"selectedAudioTrackId"
	],
	contextKeys: [],
	setup: calculatePresentationDurationSetup
});

//#endregion
//#region ../spf/dist/dev/media/dom/mse/duration.js
/**
* Check if we have the basics to update MediaSource duration:
* a `mediaSource` and a `presentation` with a numeric duration.
*/
function canUpdateDuration(presentation, mediaSource) {
	return !!(mediaSource && presentation && hasPresentationDuration(presentation));
}
/**
* Get the maximum buffered end time across an iterable of SourceBuffers
* (typically `mediaSource.sourceBuffers`). Returns `0` when the collection is
* empty or no buffer has any buffered ranges.
*/
function getMaxBufferedEnd(buffers) {
	let maxEnd = 0;
	for (const buffer of buffers) {
		const { buffered } = buffer;
		if (buffered.length > 0) {
			const end = buffered.end(buffered.length - 1);
			if (end > maxEnd) maxEnd = end;
		}
	}
	return maxEnd;
}
/**
* Check if the preconditions are met to *attempt* a `mediaSource.duration`
* write: a `mediaSource` is in scope and the presentation has a valid
* positive duration (or `Infinity` for live).
*
* Does **not** check `mediaSource.readyState` or `mediaSource.duration` —
* those are DOM properties the caller resolves at write time (e.g., by
* `await`ing `waitForMediaSourceOpen` and re-checking `readyState` after,
* and guarding on the existing `mediaSource.duration` for idempotency).
* Keeping these off the signal-driven predicate lets callers use this
* inside reactor state derivation without smuggling non-reactive DOM
* reads into `computed(...)`.
*
* `Infinity` is allowed — per the MSE spec, `mediaSource.duration = +Infinity`
* is how live playback signals an indefinite duration.
*/
function shouldUpdateDuration(presentation, mediaSource) {
	if (!canUpdateDuration(presentation, mediaSource)) return false;
	const duration = presentation.duration;
	if (Number.isNaN(duration) || duration <= 0) return false;
	return true;
}
/**
* Wait for all currently-updating SourceBuffers in `buffers` to finish, or
* until `signal` aborts — whichever fires first.
*
* The MSE spec forbids setting `MediaSource.duration` while any attached
* SourceBuffer has `updating === true`. This defers until all are idle.
* Listeners are registered with `{ signal }` so an abort tears them down
* up-front rather than leaving them dangling until the next `updateend`.
*/
function waitForSourceBuffersReady(buffers, signal) {
	if (signal.aborted) return Promise.resolve();
	const updating = [];
	for (const buf of buffers) if (buf.updating) updating.push(buf);
	if (updating.length === 0) return Promise.resolve();
	return new Promise((resolve) => {
		let remaining = updating.length;
		const onUpdateEnd = () => {
			remaining--;
			if (remaining === 0) resolve();
		};
		for (const buf of updating) buf.addEventListener("updateend", onUpdateEnd, {
			once: true,
			signal
		});
		signal.addEventListener("abort", () => resolve(), { once: true });
	});
}

//#endregion
//#region ../spf/dist/dev/media/dom/mse/end-of-stream.js
/**
* Check if the temporally last segment of `expectedSegments` is present in
* `appendedSegments` and not marked partial.
*
* Compares by segment ID rather than by a pipeline flag, so the result
* stays correct across quality switches (different tracks have different
* segment IDs) and back-buffer flushes (flushed segment IDs are removed
* from the appended list).
*/
function isLastSegmentAppended(expectedSegments, appendedSegments) {
	if (expectedSegments.length === 0) return true;
	const lastSeg = expectedSegments[expectedSegments.length - 1];
	if (!lastSeg) return false;
	return appendedSegments?.some((s) => s.id === lastSeg.id && !s.partial) ?? false;
}

//#endregion
//#region ../spf/dist/dev/media/dom/mse/mediasource-setup.js
/**
* MediaSource Setup
*
* Utilities for creating and configuring MediaSource/ManagedMediaSource
* for MSE (Media Source Extensions) playback.
*
* Global ManagedMediaSource types are defined in ./mediasource.d.ts
*/
/**
* Check if MediaSource API is supported.
*/
function supportsMediaSource() {
	return typeof MediaSource !== "undefined";
}
/**
* Check if ManagedMediaSource API is supported.
* ManagedMediaSource is a newer Safari API with better lifecycle management.
*/
function supportsManagedMediaSource() {
	return typeof ManagedMediaSource !== "undefined";
}
/**
* Create a MediaSource or ManagedMediaSource instance.
*
* @param options - Creation options
* @returns A MediaSource or ManagedMediaSource instance
* @throws Error if no MediaSource API is available
*
* @example
* const mediaSource = createMediaSource();
* const mediaElement = document.querySelector('video');
* attachMediaSource(mediaSource, mediaElement);
*/
function createMediaSource(options = {}) {
	const { preferManaged = false } = options;
	if (preferManaged && supportsManagedMediaSource()) return new ManagedMediaSource();
	if (supportsMediaSource()) return new MediaSource();
	throw new Error("MediaSource API is not supported");
}
/**
* Attach a MediaSource to an HTMLMediaElement.
*
* Uses srcObject for ManagedMediaSource (Safari), or createObjectURL for regular MediaSource.
*
* @param mediaSource - The MediaSource to attach
* @param mediaElement - The media element to attach to
* @returns Object with URL and detach function
*
* @example
* const mediaSource = createMediaSource();
* const { detach } = attachMediaSource(mediaSource, videoElement);
* // Use mediaSource...
* // Later, to clean up:
* detach();
*/
function attachMediaSource(mediaSource, mediaElement) {
	if (supportsManagedMediaSource() && mediaSource instanceof ManagedMediaSource) {
		mediaElement.disableRemotePlayback = true;
		mediaElement.srcObject = mediaSource;
		const detach = () => {
			mediaElement.srcObject = null;
			mediaElement.load();
		};
		return {
			url: "",
			detach
		};
	}
	const url = URL.createObjectURL(mediaSource);
	mediaElement.src = url;
	const detach = () => {
		mediaElement.removeAttribute("src");
		mediaElement.load();
		URL.revokeObjectURL(url);
	};
	return {
		url,
		detach
	};
}
/**
* Create a SourceBuffer on a MediaSource.
*
* @param mediaSource - The MediaSource (must be in 'open' state)
* @param mimeCodec - MIME type with codecs (e.g., 'video/mp4; codecs="avc1.42E01E"')
* @returns The created SourceBuffer
* @throws Error if MediaSource is not open or codec is unsupported
*
* @example
* const buffer = createSourceBuffer(mediaSource, 'video/mp4; codecs="avc1.42E01E"');
*/
function createSourceBuffer(mediaSource, mimeCodec) {
	if (mediaSource.readyState !== "open") throw new Error("MediaSource is not open");
	if (!isCodecSupported(mimeCodec)) throw new Error(`Codec not supported: ${mimeCodec}`);
	return mediaSource.addSourceBuffer(mimeCodec);
}
/**
* Build a MIME codec string from a track's `mimeType` + `codecs`. Works
* on partially-resolved tracks — both fields come from the multivariant
* playlist and are available before media-playlist resolution.
*
* @param track - Track carrying `mimeType` and `codecs`
* @returns MIME codec string suitable for `MediaSource.addSourceBuffer`
*
* @example
* buildMimeCodec({ mimeType: 'video/mp4', codecs: ['avc1.42E01E'] })
* // => 'video/mp4; codecs="avc1.42E01E"'
*/
function buildMimeCodec(track) {
	const codecString = track.codecs?.join(",") ?? "";
	return `${track.mimeType}; codecs="${codecString}"`;
}
/**
* Check if a codec is supported.
*
* @param mimeCodec - MIME type with codecs string
* @returns True if the codec is supported
*
* @example
* if (isCodecSupported('video/mp4; codecs="avc1.42E01E"')) {
*   // Create source buffer
* }
*/
function isCodecSupported(mimeCodec) {
	if (!supportsMediaSource()) return false;
	return MediaSource.isTypeSupported(mimeCodec);
}
/**
* Observe `mediaSource.readyState` changes via DOM events.
*
* Listens to `sourceopen`, `sourceended`, and `sourceclose` and invokes
* `onChange` with the current `readyState` after each event. Listeners
* are automatically removed when `abortSignal` is aborted.
*
* @param mediaSource - The MediaSource to observe
* @param abortSignal - AbortSignal that controls listener lifetime
* @param onChange - Called with the current readyState after each change
*
* @example
* const controller = new AbortController();
* onMediaSourceReadyStateChange(mediaSource, controller.signal, (state) => {
*   if (state === 'open') { ... }
* });
* // Later: controller.abort();
*/
function onMediaSourceReadyStateChange(mediaSource, abortSignal, onChange) {
	const update = () => onChange(mediaSource.readyState);
	const options = { signal: abortSignal };
	mediaSource.addEventListener("sourceopen", update, options);
	mediaSource.addEventListener("sourceended", update, options);
	mediaSource.addEventListener("sourceclose", update, options);
}
/**
* Wait until `mediaSource.readyState` transitions away from `'closed'`
* (the next `sourceopen`/`sourceended`/`sourceclose` event), or until
* `signal` aborts — whichever fires first.
*
* Resolves immediately when `readyState` is already `'open'` or `'ended'`
* (no further transition is coming, so there's nothing to wait for). The
* caller is expected to re-check `readyState` after the await to
* distinguish `'open'` from terminal states.
*
* Companion to `onMediaSourceReadyStateChange` for one-shot use in async
* sequences (e.g., a behavior's reactor entry that needs to wait for the
* MediaSource to attach before performing a spec-conforming mutation).
*
* @example
* await waitForMediaSourceOpen(mediaSource, signal);
* if (signal.aborted || mediaSource.readyState !== 'open') return;
* // safe to perform 'open'-state-only work
*/
function waitForMediaSourceOpen(mediaSource, signal) {
	if (signal.aborted) return Promise.resolve();
	if (mediaSource.readyState !== "closed") return Promise.resolve();
	return new Promise((resolve) => {
		const done = () => resolve();
		const options = {
			once: true,
			signal
		};
		mediaSource.addEventListener("sourceopen", done, options);
		mediaSource.addEventListener("sourceended", done, options);
		mediaSource.addEventListener("sourceclose", done, options);
		signal.addEventListener("abort", done, { once: true });
	});
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/end-of-stream.js
/**
* **Drive each `open → ended` transition of the MediaSource.** Calls
* `MediaSource.endOfStream()` once the temporally last segments of every
* active buffer actor's currently-loading track are fully appended and
* the user has reached them — letting the browser finalize duration and
* fire `ended` on the media element.
*
* Re-fires on every subsequent `open → ended → open` cycle. Per the MSE
* spec, `appendBuffer()` after `endOfStream()` transitions the MediaSource
* back to `'open'`; seek-back replays and back-buffer refills that re-load
* earlier segments take this path, so the behavior must call
* `endOfStream()` again once the last segments are reappended. The reactor's
* `'preconditions-unmet'` ↔ `'eos-ready'` cycle *is* the re-arm mechanism:
* a successful `endOfStream()` flips `mediaSource.readyState` to `'ended'`,
* which (via the local `msIsOpen` signal) exits `'eos-ready'`; the next
* `'open'` re-evaluates preconditions and may re-enter.
*
* # Tracking what's in the buffer
*
* Each `SourceBufferActor` knows which track it's currently loading via
* `initTrackId` (set on the most recent `append-init` message) and which
* segments it has appended. `deriveState` iterates over the available
* buffer actors (`[videoBufferActor, audioBufferActor].filter(Boolean)`),
* resolves each to its track via `findTrackById`, and checks that track's
* last segment is appended. This means audio-only / video-only / mixed
* configurations compose uniformly — the body iterates whatever's in
* scope. No reliance on `selectedTrackId` slots: the actor's view IS the
* source of truth for "what's being loaded into this buffer."
*
* **Two-fire on mid-end ABR switches**: when a quality switch occurs near
* end-of-stream, the actor's `initTrackId` still reflects the *old* track
* until the new init segment is appended. The reactor may fire
* `endOfStream()` against the old track's last-segment-appended state,
* then the new init's `appendBuffer()` re-opens the MS, which re-arms us
* to fire again once the new track's last segment lands. Functionally
* correct (the browser re-fires `ended` after the re-arm), at the cost of
* one extra call. Accepted as the price of dropping `selectedTrackId`
* dependence.
*
* # Buffer-actor idle gate
*
* Each actor must be `'idle'` (no queued or in-flight tasks) before we
* fire. `buffer.updating === false` alone is insufficient: for chunked
* fMP4 streaming, `updateend` fires after each chunk while the actor's
* for-await loop synchronously enqueues the next `appendBuffer()` — so
* the buffer flips `!updating → updating` across a microtask boundary.
* The actor's `'updating'` state spans the entire multi-chunk append, so
* `actor.snapshot.value === 'idle'` is the canonical "no more pending or
* in-flight work" oracle.
*
* This bet (actor models the pipeline; `mediaSource.sourceBuffers` models
* DOM state) is durable as long as the segment loader is the sole writer
* of `appendBuffer()` calls. A future loop-mode that auto-fetches earlier
* segments mid-`ended` would require a broader coordination story
* between SourceBuffers and the MediaSource — possibly via a
* `MediaSourceActor` — and that's where this gate would want
* re-evaluating.
*
* # currentTime gate
*
* `currentTime` must have reached at least one active track's last
* segment startTime. Prevents `'eos-ready'` entry when a back-buffer
* `remove()` / `appendBuffer()` briefly re-opens the MediaSource while
* the user is mid-stream. HLS rendition time-alignment means any active
* track works as the reference.
*
* # MS readyState — local subscription
*
* The behavior subscribes to `mediaSource`'s readyState changes inside
* its setup (via `onMediaSourceReadyStateChange`) and mirrors `'open'` to
* a behavior-local `msIsOpen` signal that `deriveState` reads. No shared
* `mediaSourceReadyState` slot dependency. Anticipates a future
* `MediaSourceActor` whose snapshot would expose the same signal — the
* consumer code wouldn't change.
*
* # `endOfStream` entry sequence
*
* 1. Wait for all SourceBuffers to be idle (DOM-level — defensive,
*    should already hold given the actor-idle gate in `deriveState`).
* 2. Set `mediaSource.duration` from `getMaxBufferedEnd` to match actual
*    container timestamps (`endOfStream()` only clamps up implicitly;
*    setting it explicitly here keeps the final value deterministic).
* 3. Call `mediaSource.endOfStream()`.
*
* State-exit cleanup (`controller.abort()`) cancels any in-flight wait
* on source unload, presentation replace, or behavior destroy.
*
* # Coordination with `updateMediaSourceDuration`
*
* `updateMediaSourceDuration` writes the initial `mediaSource.duration`
* from `presentation.duration` once per source; this behavior writes the
* final value from the buffered end. Decision domains don't overlap.
*/
function deriveState$4(presentation, mediaSource, msIsOpen, videoBufferActor, audioBufferActor, currentTime) {
	if (!mediaSource || !presentation || !msIsOpen) return "preconditions-unmet";
	const actors = [videoBufferActor, audioBufferActor].filter((a) => a !== void 0);
	if (actors.length === 0) return "preconditions-unmet";
	let lastSegStart;
	for (const actor of actors) {
		const snapshot = actor.snapshot.get();
		if (snapshot.value !== "idle") return "preconditions-unmet";
		const { initTrackId, segments: appended } = snapshot.context;
		if (!initTrackId) return "preconditions-unmet";
		const track = findTrackById(presentation, initTrackId);
		if (!track || !isResolvedTrack(track)) return "preconditions-unmet";
		if (!isLastSegmentAppended(track.segments, appended)) return "preconditions-unmet";
		if (track.segments.length > 0) {
			const start = track.segments[track.segments.length - 1].startTime;
			if (lastSegStart === void 0 || start > lastSegStart) lastSegStart = start;
		}
	}
	if (lastSegStart !== void 0 && (currentTime ?? 0) < lastSegStart) return "preconditions-unmet";
	return "eos-ready";
}
function endOfStreamSetup({ state, context }) {
	const msIsOpen = signal(false);
	const cleanupMsListener = effect(() => {
		const mediaSource = context.mediaSource.get();
		if (!mediaSource) {
			msIsOpen.set(false);
			return;
		}
		msIsOpen.set(mediaSource.readyState === "open");
		const controller = new AbortController();
		onMediaSourceReadyStateChange(mediaSource, controller.signal, (rs) => {
			msIsOpen.set(rs === "open");
		});
		return () => controller.abort();
	});
	const derivedStateSignal = computed(() => deriveState$4(state.presentation.get(), context.mediaSource.get(), msIsOpen.get(), context.videoBufferActor.get(), context.audioBufferActor.get(), state.currentTime.get()));
	const reactor = createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			"eos-ready": { entry: () => {
				const mediaSource = context.mediaSource.get();
				const controller = new AbortController();
				const endStreamWhenReady = async () => {
					await waitForSourceBuffersReady(mediaSource.sourceBuffers, controller.signal);
					if (controller.signal.aborted) return;
					const bufferedEnd = getMaxBufferedEnd(mediaSource.sourceBuffers);
					if (bufferedEnd > 0) mediaSource.duration = bufferedEnd;
					mediaSource.endOfStream();
				};
				endStreamWhenReady().catch((err) => console.error("Failed to call endOfStream:", err));
				return () => controller.abort();
			} }
		}
	});
	return () => {
		cleanupMsListener();
		reactor.destroy();
	};
}
const endOfStream = defineBehavior({
	stateKeys: ["presentation", "currentTime"],
	contextKeys: [
		"mediaSource",
		"videoBufferActor",
		"audioBufferActor"
	],
	setup: endOfStreamSetup
});

//#endregion
//#region ../spf/dist/dev/network/ewma.js
/**
* Exponentially Weighted Moving Average (EWMA)
*
* Pure functional implementation of EWMA calculations.
* Based on Shaka Player's EWMA algorithm.
*/
/**
* Calculate alpha (decay factor) from half-life.
*
* Alpha determines how quickly old data "expires":
* - alpha close to 1 = slow decay (long memory)
* - alpha close to 0 = fast decay (short memory)
*
* @param halfLife - The quantity of prior samples (by weight) that make up
*   half of the new estimate. Must be positive.
* @returns Alpha value between 0 and 1
*
* @example
* const alpha = calculateAlpha(2); // ≈ 0.7071 for 2-second half-life
*/
function calculateAlpha(halfLife) {
	return Math.exp(Math.log(.5) / halfLife);
}
/**
* Calculate exponentially weighted moving average.
*
* Updates an estimate by blending a new value with the previous estimate,
* weighted by the sample duration. Longer samples have more influence.
*
* @param prevEstimate - Previous EWMA estimate
* @param value - New sample value to incorporate
* @param weight - Sample weight (typically duration in seconds)
* @param halfLife - Half-life for decay (typically 2-5 seconds)
* @returns Updated EWMA estimate
*
* @example
* let estimate = 0;
* estimate = calculateEwma(estimate, 1_000_000, 1, 2); // First sample
* estimate = calculateEwma(estimate, 2_000_000, 1, 2); // Second sample
*/
function calculateEwma(prevEstimate, value, weight, halfLife) {
	const adjAlpha = calculateAlpha(halfLife) ** weight;
	return value * (1 - adjAlpha) + adjAlpha * prevEstimate;
}
/**
* Apply zero-factor correction to EWMA estimate.
*
* The zero-factor correction compensates for bias when starting from zero.
* Without this correction, early estimates would be artificially low.
*
* As totalWeight increases, the correction factor approaches 1, meaning
* the estimate becomes more reliable and needs less correction.
*
* @param estimate - Raw EWMA estimate (uncorrected)
* @param totalWeight - Accumulated weight from all samples
* @param halfLife - Half-life used in EWMA calculation
* @returns Corrected estimate, or 0 if totalWeight is 0
*
* @example
* const raw = calculateEwma(0, 1_000_000, 1, 2);
* const corrected = applyZeroFactor(raw, 1, 2); // ≈ 1_000_000
*/
function applyZeroFactor(estimate, totalWeight, halfLife) {
	if (totalWeight === 0) return 0;
	return estimate / (1 - calculateAlpha(halfLife) ** totalWeight);
}

//#endregion
//#region ../spf/dist/dev/network/bandwidth-estimator.js
/**
* Dual EWMA Bandwidth Estimator
*
* Estimates available bandwidth using two EWMA calculations with different
* half-lives, taking the minimum of both. This approach (from Shaka Player):
*
* - **Fast EWMA** (2s half-life): Reacts quickly to bandwidth drops
* - **Slow EWMA** (5s half-life): Provides stability during fluctuations
* - **min(fast, slow)**: Adapts down quickly, up slowly
*
* This naturally provides asymmetric behavior needed for good QoE:
* avoiding stalls (quick downgrade) while preventing oscillation (slow upgrade).
*/
/**
* Default bandwidth estimator configuration.
*
* Values match Shaka Player defaults based on experimentation.
*/
const DEFAULT_BANDWIDTH_CONFIG = {
	fastHalfLife: 2,
	slowHalfLife: 5,
	minTotalBytes: 128e3,
	minBytes: 16e3,
	minDuration: 5
};
/**
* Add a bandwidth sample from a segment download.
*
* Samples are filtered based on:
* - Minimum bytes (filters TTFB-dominated small segments)
* - Minimum duration (filters cached responses)
*
* Valid samples update both fast and slow EWMA estimates.
*
* @param state - Current estimator state
* @param durationMs - Download duration in milliseconds
* @param numBytes - Number of bytes downloaded
* @param config - Optional estimator configuration (uses defaults if not provided)
* @returns New estimator state with sample incorporated (or unchanged if filtered)
*
* @example
* let state = { fastEstimate: 0, fastTotalWeight: 0, ... };
* // Sample: 1MB in 1 second
* state = sampleBandwidth(state, 1000, 1_000_000);
*/
function sampleBandwidth(state, durationMs, numBytes, config = DEFAULT_BANDWIDTH_CONFIG) {
	const updatedBytesSampled = state.bytesSampled + numBytes;
	if (numBytes < config.minBytes) return {
		...state,
		bytesSampled: updatedBytesSampled
	};
	if (durationMs < config.minDuration) return {
		...state,
		bytesSampled: updatedBytesSampled
	};
	const bandwidth = 8e3 * numBytes / durationMs;
	const weight = durationMs / 1e3;
	return {
		fastEstimate: calculateEwma(state.fastEstimate, bandwidth, weight, config.fastHalfLife),
		fastTotalWeight: state.fastTotalWeight + weight,
		slowEstimate: calculateEwma(state.slowEstimate, bandwidth, weight, config.slowHalfLife),
		slowTotalWeight: state.slowTotalWeight + weight,
		bytesSampled: updatedBytesSampled
	};
}
/**
* Get the current bandwidth estimate.
*
* Returns the **minimum** of the fast and slow EWMA estimates.
* This provides the key asymmetric behavior:
* - When bandwidth drops, fast EWMA reacts first and dominates (quick adaptation)
* - When bandwidth rises, slow EWMA lags behind and dominates (slow adaptation)
*
* Uses default estimate until enough data has been sampled — and when no
* estimator state exists at all (`state === undefined`).
*
* @param state - Current estimator state, or `undefined` before any samples have been collected
* @param defaultEstimate - Fallback estimate before sufficient samples (bps)
* @param config - Optional estimator configuration (uses defaults if not provided)
* @returns Bandwidth estimate in bits per second
*
* @example
* const estimate = getBandwidthEstimate(state, 5_000_000); // 5 Mbps default
*/
function getBandwidthEstimate(state, defaultEstimate, config = DEFAULT_BANDWIDTH_CONFIG) {
	if (!state || state.bytesSampled < config.minTotalBytes) return defaultEstimate;
	const fastEstimate = applyZeroFactor(state.fastEstimate, state.fastTotalWeight, config.fastHalfLife);
	const slowEstimate = applyZeroFactor(state.slowEstimate, state.slowTotalWeight, config.slowHalfLife);
	return Math.min(fastEstimate, slowEstimate);
}

//#endregion
//#region ../spf/dist/dev/network/chunked-stream-iterable.js
const DEFAULT_MIN_CHUNK_SIZE = 2 ** 17;
/**
* Adapts a `ReadableStream<Uint8Array>` (e.g. `response.body`) into an
* `AsyncIterable<Uint8Array>` that yields chunks no smaller than
* `minChunkSize` bytes. Smaller network chunks are accumulated and yielded
* together once the threshold is met. Any remainder is flushed on stream end.
*
* Errors from the underlying stream propagate naturally — the reader lock is
* always released via `finally`.
*/
var ChunkedStreamIterable = class {
	minChunkSize;
	#readableStream;
	constructor(readableStream, { minChunkSize = DEFAULT_MIN_CHUNK_SIZE } = {}) {
		this.#readableStream = readableStream;
		this.minChunkSize = minChunkSize;
	}
	async *[Symbol.asyncIterator]() {
		let pending;
		const reader = this.#readableStream.getReader();
		try {
			while (true) {
				const { done, value } = await reader.read();
				if (done) {
					if (pending) yield pending;
					break;
				}
				pending = pending ? concat(pending, value) : value;
				if (pending.length >= this.minChunkSize) {
					yield pending;
					pending = void 0;
				}
			}
		} finally {
			reader.releaseLock();
		}
	}
};
function concat(a, b) {
	const result = new Uint8Array(a.length + b.length);
	result.set(a);
	result.set(b, a.length);
	return result;
}

//#endregion
//#region ../spf/dist/dev/network/fetch.js
/**
* HTTP Fetch Wrapper
*
* Composable building blocks:
* - fetchResolvable() — fetch a Resource (handles byte ranges); returns Response
* - getResponseText() — extract text from Response
* - fetchResolvableStream() — single-stage async generator over body chunks
* - fetchStream() — two-stage: await connection establishment, then lazily
*   iterate body chunks. Use when timing the connection start independently
*   of body consumption matters (e.g., observable fetch timing for ABR).
* - createTrackedFetch() — factory for a fetchStream-shape function that
*   samples bandwidth (via EWMA) per chunk and notifies via callback.
*/
/**
* Fetch resolvable from a Resource.
*
* Handles byte range requests if byteRange is present.
* Returns native fetch Response for composability (can extract text, stream, etc.).
*
* @param addressable - Resource to fetch (url + optional byteRange)
* @returns Promise resolving to Response
*
* @example
* const response = await fetchResolvable({ url: 'https://example.com/segment.m4s' });
* const text = await getResponseText(response);
*
* @example
* // With byte range
* const response = await fetchResolvable({
*   url: 'https://example.com/file.mp4',
*   byteRange: { start: 1000, end: 1999 }
* });
*/
async function fetchResolvable(addressable, options) {
	const headers = new Headers(options?.headers);
	if (addressable.byteRange) {
		const { start, end } = addressable.byteRange;
		headers.set("Range", `bytes=${start}-${end}`);
	}
	const request = new Request(addressable.url, {
		method: "GET",
		headers,
		...options
	});
	return fetch(request);
}
/**
* Extract text from Response.
*
* Accepts minimal Response-like object (just needs text() method).
* Returns promise from response.text().
*
* @param response - Response-like object with text() method
* @returns Promise resolving to text content
*
* @example
* const response = await fetchResolvable(addressable);
* const text = await getResponseText(response);
*/
function getResponseText(response) {
	return response.text();
}
async function fetchStream(addressable, options) {
	const { minChunkSize, ...fetchOptions } = options ?? {};
	const response = await fetchResolvable(addressable, fetchOptions);
	if (!response.body) throw new Error("Response has no body");
	return new ChunkedStreamIterable(response.body, ...minChunkSize !== void 0 ? [{ minChunkSize }] : []);
}
/**
* Returns a {@link FetchBytes} function that samples bandwidth via EWMA
* per body chunk. The factory captures the running bandwidth state
* internally; per chunk it computes the next state and notifies the
* supplied `onSample` callback.
*
* The factory's internal accumulator is seeded from `initial` and updated
* on every chunk; callers don't need to thread it back in. `onSample`
* receives the *new* state after each chunk — typical use is to bridge
* samples back into engine state for ABR consumers.
*
* @param initial - Starting `BandwidthState` (commonly zeros or the
*   engine's current accumulator).
* @param onSample - Called with the new `BandwidthState` after each chunk.
*/
function createTrackedFetch(initial, onSample) {
	let state = initial;
	return async (addressable, options) => {
		const { minChunkSize, ...fetchOptions } = options ?? {};
		const response = await fetchResolvable(addressable, fetchOptions);
		if (!response.body) throw new Error("Response has no body");
		const body = response.body;
		return { [Symbol.asyncIterator]: async function* () {
			let chunkStart = performance.now();
			for await (const chunk of new ChunkedStreamIterable(body, ...minChunkSize !== void 0 ? [{ minChunkSize }] : [])) {
				const elapsed = performance.now() - chunkStart;
				state = sampleBandwidth(state, elapsed, chunk.byteLength);
				onSample(state);
				yield chunk;
				chunkStart = performance.now();
			}
		} };
	};
}

//#endregion
//#region ../spf/dist/dev/media/buffer/back-buffer.js
/**
* Default back buffer configuration.
*/
const DEFAULT_BACK_BUFFER_CONFIG = { keepSegments: 2 };
/**
* Calculate back buffer flush point.
*
* Determines where to flush old segments from the back buffer.
* Keeps a fixed number of segments behind the current playback position.
*
* Algorithm:
* 1. Find segments before currentTime
* 2. Count back N segments (keepSegments)
* 3. Return startTime of segment N+1 back (flush everything before this)
*
* @param segments - Available segments (should be sorted by startTime)
* @param currentTime - Current playback position in seconds
* @param config - Optional back buffer configuration
* @returns Time in seconds to flush up to (flush range: [0, flushEnd))
*
* @example
* const segments = [
*   { startTime: 0, duration: 6, ... },
*   { startTime: 6, duration: 6, ... },
*   { startTime: 12, duration: 6, ... },
*   { startTime: 18, duration: 6, ... },
* ];
*
* // Playing at 18s, keep 2 segments
* const flushEnd = calculateBackBufferFlushPoint(segments, 18);
* // Returns 6 (flush [0, 6), keep [6-18))
*/
function calculateBackBufferFlushPoint(segments, currentTime, config = DEFAULT_BACK_BUFFER_CONFIG) {
	if (segments.length === 0) return 0;
	const segmentsBefore = segments.filter((seg) => seg.startTime < currentTime);
	if (segmentsBefore.length === 0) return 0;
	const segmentsToFlush = segmentsBefore.length - config.keepSegments;
	if (segmentsToFlush <= 0) return 0;
	if (segmentsToFlush >= segmentsBefore.length) return currentTime;
	return segmentsBefore[segmentsToFlush].startTime;
}

//#endregion
//#region ../spf/dist/dev/playback/actors/dom/segment-loader.js
/**
* Resolves when the SourceBufferActor snapshot reaches 'idle'.
* Rejects if the signal is aborted or the actor is destroyed.
*
* Used to sequence SourceBufferActor operations without awaiting send()
* directly — send() is fire-and-forget; callers observe completion via
* state transition.
*/
function waitForIdle(snapshot, signal) {
	return new Promise((resolve, reject) => {
		if (snapshot.get().value === "idle") {
			resolve();
			return;
		}
		if (snapshot.get().value === "destroyed") {
			reject(new DOMException("Aborted", "AbortError"));
			return;
		}
		if (signal.aborted) {
			reject(signal.reason);
			return;
		}
		let stop;
		const cleanup = (fn) => {
			stop?.();
			signal.removeEventListener("abort", onAbort);
			fn();
		};
		const onAbort = () => cleanup(() => reject(signal.reason));
		stop = effect(() => {
			const value = snapshot.get().value;
			if (value === "idle") cleanup(resolve);
			else if (value === "destroyed") cleanup(() => reject(new DOMException("Aborted", "AbortError")));
		});
		signal.addEventListener("abort", onAbort, { once: true });
	});
}
/**
* Wraps a LoadTask descriptor into a Task that fetches (if needed) and
* forwards to SourceBufferActor. Updates in-flight context around async
* operations so the loading handler can make accurate continue/preempt
* decisions at any point.
*/
function makeLoadTask(op, { getContext, setContext, fetchBytes, sourceBufferActor }) {
	return new Task(async (taskSignal) => {
		if (taskSignal.aborted) return;
		if (op.type === "remove") {
			sourceBufferActor.send(op);
			await waitForIdle(sourceBufferActor.snapshot, taskSignal);
			return;
		}
		if (op.type === "append-init") {
			setContext({
				...getContext(),
				inFlightInitTrackId: op.meta.trackId
			});
			try {
				const data = await fetchBytes(op, {
					signal: taskSignal,
					minChunkSize: Infinity
				});
				if (!taskSignal.aborted) {
					sourceBufferActor.send({
						type: "append-init",
						data,
						meta: op.meta
					});
					await waitForIdle(sourceBufferActor.snapshot, taskSignal);
				}
			} finally {
				setContext({
					...getContext(),
					inFlightInitTrackId: null
				});
			}
			return;
		}
		setContext({
			...getContext(),
			inFlightSegmentId: op.meta.id
		});
		try {
			const stream = await fetchBytes(op, { signal: taskSignal });
			if (!taskSignal.aborted) {
				sourceBufferActor.send({
					type: "append-segment",
					data: stream,
					meta: op.meta
				});
				await waitForIdle(sourceBufferActor.snapshot, taskSignal);
			}
		} finally {
			setContext({
				...getContext(),
				inFlightSegmentId: null
			});
		}
	});
}
/**
* Creates a SegmentLoaderActor for one track type (video or audio).
*
* Receives load assignments via `send()` and owns all execution: planning,
* removes, fetches, and appends. Coordinates with the SourceBufferActor for
* all physical SourceBuffer operations.
*
* Planning (Cases 1–3) happens in the `load` handler on every incoming
* message, producing an ordered LoadTask list. The runner drains that list
* sequentially via SerialRunner. When a new message arrives mid-run, the
* handler replans and either continues the in-flight operation (abortPending
* + schedule new remainder) or preempts it (abortAll + cancel SourceBuffer
* if needed + schedule new plan).
*
* @param sourceBufferActor - Shared SourceBufferActor reference (not owned)
* @param fetchBytes - Tracked fetch closure (owns throughput sampling for segments).
*   Accepts an optional `minChunkSize` in options; init segments pass `Infinity`
*   so the entire body accumulates as one chunk before appending.
*/
function createSegmentLoaderActor(sourceBufferActor, fetchBytes, config = {}) {
	const forwardBufferConfig = {
		...DEFAULT_FORWARD_BUFFER_CONFIG,
		...config.forwardBuffer
	};
	const backBufferConfig = {
		...DEFAULT_BACK_BUFFER_CONFIG,
		...config.backBuffer
	};
	const getBufferedSegments = (allSegments) => {
		const bufferedIds = new Set(peek(sourceBufferActor.snapshot).context.segments.filter((s) => !s.partial).map((s) => s.id));
		return allSegments.filter((s) => bufferedIds.has(s.id));
	};
	/**
	* Translate a load message into an ordered LoadTask list based on committed
	* actor state. In-flight awareness is handled separately in the load handler.
	*
	* @todo Rename alongside LoadTask (e.g. planOps).
	*
	* Case 1 — Removes: forward and back buffer flush points, segment-aligned.
	*   No flush on track switch: appending new content overwrites existing buffer
	*   ranges, and the actor's time-aligned deduplication keeps the segment model
	*   accurate as new segments arrive.
	*
	* Case 2 — Init: schedule if not yet committed for this track.
	*
	* Case 3 — Segments: all segments in the load window not yet committed.
	*/
	const planTasks = (message) => {
		const { track, range } = message;
		const actorCtx = peek(sourceBufferActor.snapshot).context;
		const bufferedSegments = getBufferedSegments(track.segments);
		const currentTime = range?.start ?? 0;
		const tasks = [];
		if (range) {
			const forwardFlushStart = calculateForwardFlushPoint(bufferedSegments, currentTime, forwardBufferConfig);
			if (forwardFlushStart < Infinity) tasks.push({
				type: "remove",
				start: forwardFlushStart,
				end: Infinity
			});
			const backFlushEnd = calculateBackBufferFlushPoint(bufferedSegments, currentTime, backBufferConfig);
			if (backFlushEnd > 0) tasks.push({
				type: "remove",
				start: 0,
				end: backFlushEnd
			});
		}
		if (actorCtx.initTrackId !== track.id) tasks.push({
			type: "append-init",
			meta: { trackId: track.id },
			url: track.initialization.url,
			...track.initialization.byteRange !== void 0 && { byteRange: track.initialization.byteRange }
		});
		if (range) {
			const segmentsToLoad = getSegmentsToLoad(track.segments, bufferedSegments, currentTime, forwardBufferConfig).filter((seg) => {
				const existing = actorCtx.segments.find((s) => Math.abs(s.startTime - seg.startTime) < SEGMENT_TIME_EPSILON);
				if (existing?.partial) return true;
				if (!existing?.trackBandwidth || !track.bandwidth) return true;
				return track.bandwidth > existing.trackBandwidth;
			});
			for (const segment of segmentsToLoad) tasks.push({
				type: "append-segment",
				meta: {
					id: segment.id,
					startTime: segment.startTime,
					duration: segment.duration,
					trackId: track.id,
					trackBandwidth: track.bandwidth
				},
				url: segment.url,
				...segment.byteRange !== void 0 && { byteRange: segment.byteRange }
			});
		}
		return tasks;
	};
	const scheduleAll = (tasks, { getContext, setContext, runner }) => {
		tasks.forEach((op) => {
			runner.schedule(makeLoadTask(op, {
				getContext,
				setContext,
				fetchBytes,
				sourceBufferActor
			})).then(void 0, (e) => {
				if (e instanceof Error && e.name === "AbortError") return;
				console.error("Unexpected error in segment loader:", e);
				runner.abortPending();
			});
		});
	};
	return createMachineActor({
		runner: () => new SerialRunner(),
		initial: "idle",
		context: {
			inFlightInitTrackId: null,
			inFlightSegmentId: null
		},
		states: {
			idle: { on: { load: (msg, ctx) => {
				const allTasks = planTasks(msg);
				if (allTasks.length === 0) return;
				ctx.transition("loading");
				scheduleAll(allTasks, ctx);
			} } },
			loading: {
				onSettled: "idle",
				on: { load: (msg, ctx) => {
					const { context, runner } = ctx;
					const allTasks = planTasks(msg);
					if (context.inFlightSegmentId !== null && allTasks.some((t) => t.type === "append-segment" && t.meta.id === context.inFlightSegmentId) || context.inFlightInitTrackId !== null && allTasks.some((t) => t.type === "append-init" && t.meta.trackId === context.inFlightInitTrackId)) {
						runner.abortPending();
						scheduleAll(allTasks.filter((t) => !(t.type === "append-segment" && t.meta.id === context.inFlightSegmentId) && !(t.type === "append-init" && t.meta.trackId === context.inFlightInitTrackId)), ctx);
					} else {
						runner.abortAll();
						if (context.inFlightSegmentId !== null || context.inFlightInitTrackId !== null && allTasks.some((t) => t.type === "append-init" && t.meta.trackId !== context.inFlightInitTrackId)) sourceBufferActor.send({ type: "cancel" });
						scheduleAll(allTasks, ctx);
					}
				} }
			}
		}
	});
}

//#endregion
//#region ../spf/dist/dev/media/dom/mse/append-segment.js
/**
* Append media data to a SourceBuffer.
*
* Accepts either a full ArrayBuffer (single append) or an AsyncIterable of
* Uint8Array chunks (one append per chunk, in order). Waits for `updateend`
* between each call so appends are serialized correctly.
*
* Errors from the SourceBuffer (`error` event) or from the iterable are
* propagated as rejections.
*/
async function appendSegment(sourceBuffer, data, signal) {
	if (data instanceof ArrayBuffer) await appendChunk(sourceBuffer, data);
	else try {
		for await (const chunk of data) {
			if (signal?.aborted) throw signal.reason ?? new DOMException("Aborted", "AbortError");
			await appendChunk(sourceBuffer, chunk);
		}
	} catch (e) {
		if (e instanceof DOMException && e.name === "AbortError" && !sourceBuffer.updating) try {
			sourceBuffer.abort();
		} catch {}
		throw e;
	}
}
async function appendChunk(sourceBuffer, data) {
	if (sourceBuffer.updating) await new Promise((resolve) => {
		const onUpdateEnd = () => {
			sourceBuffer.removeEventListener("updateend", onUpdateEnd);
			resolve();
		};
		sourceBuffer.addEventListener("updateend", onUpdateEnd);
	});
	return new Promise((resolve, reject) => {
		const onUpdateEnd = () => {
			cleanup();
			resolve();
		};
		const onError = (event) => {
			cleanup();
			reject(/* @__PURE__ */ new Error(`SourceBuffer append error: ${event.type}`));
		};
		const cleanup = () => {
			sourceBuffer.removeEventListener("updateend", onUpdateEnd);
			sourceBuffer.removeEventListener("error", onError);
		};
		sourceBuffer.addEventListener("updateend", onUpdateEnd);
		sourceBuffer.addEventListener("error", onError);
		try {
			sourceBuffer.appendBuffer(data);
		} catch (error) {
			cleanup();
			reject(error);
		}
	});
}

//#endregion
//#region ../spf/dist/dev/media/dom/mse/buffer-flusher.js
/**
* Buffer flusher helper (P12)
*
* Removes a time range from a SourceBuffer to manage memory.
*/
/**
* Remove a time range from a SourceBuffer.
*
* Waits for the SourceBuffer to be ready (not updating), then removes
* the specified range. Returns a promise that resolves when removal completes.
*
* @param sourceBuffer - The SourceBuffer to remove data from
* @param start - Start of the time range to remove (seconds)
* @param end - End of the time range to remove (seconds)
* @returns Promise that resolves when removal completes
*
* @example
* await flushBuffer(videoSourceBuffer, 0, 30);
*/
async function flushBuffer(sourceBuffer, start, end) {
	if (sourceBuffer.updating) await new Promise((resolve) => {
		const onUpdateEnd = () => {
			sourceBuffer.removeEventListener("updateend", onUpdateEnd);
			resolve();
		};
		sourceBuffer.addEventListener("updateend", onUpdateEnd);
	});
	return new Promise((resolve, reject) => {
		const onUpdateEnd = () => {
			cleanup();
			resolve();
		};
		const onError = (event) => {
			cleanup();
			reject(/* @__PURE__ */ new Error(`SourceBuffer remove error: ${event.type}`));
		};
		const cleanup = () => {
			sourceBuffer.removeEventListener("updateend", onUpdateEnd);
			sourceBuffer.removeEventListener("error", onError);
		};
		sourceBuffer.addEventListener("updateend", onUpdateEnd);
		sourceBuffer.addEventListener("error", onError);
		try {
			sourceBuffer.remove(start, end);
		} catch (error) {
			cleanup();
			reject(error);
		}
	});
}

//#endregion
//#region ../spf/dist/dev/playback/actors/dom/source-buffer.js
function snapshotBuffered(buffered) {
	const ranges = [];
	for (let i = 0; i < buffered.length; i++) ranges.push({
		start: buffered.start(i),
		end: buffered.end(i)
	});
	return ranges;
}
function appendInitTask(message, { getContext, sourceBuffer }) {
	return new Task(async (taskSignal) => {
		const ctx = getContext();
		if (taskSignal.aborted) return ctx;
		await appendSegment(sourceBuffer, message.data);
		return {
			...ctx,
			initTrackId: message.meta.trackId
		};
	});
}
function appendSegmentTask(message, { getContext, sourceBuffer, setContext }) {
	return new Task(async (taskSignal) => {
		const ctx = getContext();
		if (taskSignal.aborted) return ctx;
		const { meta } = message;
		const filtered = ctx.segments.filter((s) => Math.abs(s.startTime - meta.startTime) >= SEGMENT_TIME_EPSILON);
		if (!(message.data instanceof ArrayBuffer)) setContext({
			...ctx,
			segments: [...filtered, {
				id: meta.id,
				startTime: meta.startTime,
				duration: meta.duration,
				trackId: meta.trackId,
				...meta.trackBandwidth !== void 0 && { trackBandwidth: meta.trackBandwidth },
				partial: true
			}],
			bufferedRanges: ctx.bufferedRanges
		});
		await appendSegment(sourceBuffer, message.data, taskSignal);
		return {
			...ctx,
			segments: [...filtered, {
				id: meta.id,
				startTime: meta.startTime,
				duration: meta.duration,
				trackId: meta.trackId,
				...meta.trackBandwidth !== void 0 && { trackBandwidth: meta.trackBandwidth }
			}],
			bufferedRanges: snapshotBuffered(sourceBuffer.buffered)
		};
	});
}
function removeTask(message, { getContext, sourceBuffer }) {
	return new Task(async (taskSignal) => {
		const ctx = getContext();
		if (taskSignal.aborted) return ctx;
		await flushBuffer(sourceBuffer, message.start, message.end);
		const bufferedRanges = snapshotBuffered(sourceBuffer.buffered);
		const filtered = ctx.segments.filter((s) => {
			const midpoint = s.startTime + s.duration / 2;
			return bufferedRanges.some((r) => midpoint >= r.start && midpoint < r.end);
		});
		return {
			...ctx,
			segments: filtered,
			bufferedRanges
		};
	});
}
const messageTaskFactories = {
	"append-init": appendInitTask,
	"append-segment": appendSegmentTask,
	remove: removeTask
};
function messageToTask(message, options) {
	const factory = messageTaskFactories[message.type];
	return factory(message, options);
}
function createSourceBufferActor(sourceBuffer, initialContext) {
	const handleError = (e) => {
		if (!(e instanceof Error && e.name === "AbortError")) console.error("SourceBuffer operation failed:", e);
	};
	const onMessage = (msg, { transition, setContext, getContext, runner }) => {
		transition("updating");
		const task = messageToTask(msg, {
			getContext,
			sourceBuffer,
			setContext
		});
		runner.schedule(task).then(setContext, handleError);
	};
	return createMachineActor({
		runner: () => new SerialRunner(),
		initial: "idle",
		context: {
			segments: [],
			bufferedRanges: [],
			initTrackId: void 0,
			...initialContext
		},
		states: {
			idle: { on: {
				"append-init": onMessage,
				"append-segment": onMessage,
				remove: onMessage,
				batch: (msg, { transition, setContext, getContext, runner }) => {
					const { messages } = msg;
					if (messages.length === 0) return;
					transition("updating");
					messages.forEach((msg) => {
						const task = messageToTask(msg, {
							getContext,
							sourceBuffer,
							setContext
						});
						runner.schedule(task).then(setContext, handleError);
					});
				}
			} },
			updating: {
				onSettled: "idle",
				on: { cancel: (_, { runner }) => {
					runner.abortAll();
				} }
			}
		}
	});
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/setup-buffer-actors.js
/**
* **Per-type buffer + segment-loader actor setup.** Per available track
* type (video / audio), when `mediaSource` is attached and the selected
* track of that type is present in the presentation with codecs (partial
* resolution from the multivariant playlist is enough — codecs live on
* the `EXT-X-STREAM-INF` line, not in the per-type media playlist),
* creates a `SourceBuffer`, a `SourceBufferActor`, and a
* `SegmentLoaderActor` bound to that buffer-actor; publishes the per-type
* actor slots. On `mediaSource` detach or behavior destroy, destroys both
* actors in reverse order and clears the per-type slots so the next
* source starts fresh.
*
* Each per-type variant (`setupVideoBufferActors` /
* `setupAudioBufferActors`) is a single-positive-state reactor
* (`'preconditions-unmet'` ↔ `'buffer-ready'`) gating only on its own
* type. No cross-type coupling in `stateKeys` —
* `setupVideoBufferActors` carries only `selectedVideoTrackId` (plus
* `bandwidthState`, written by its trackedFetch), and audio mirrors.
*
* # Firefox `mozHasAudio` invariant
*
* Appending to a video `SourceBuffer` before the audio `SourceBuffer`
* exists causes `mozHasAudio` to be permanently false in Firefox. With
* the two per-type variants decoupled, the invariant is no longer
* structural to a single `entry` body (as it was when both buffers were
* created in one synchronous block inside a merged behavior). It's now
* preserved by a chain of assumptions about how this behavior composes
* with its upstream and downstream siblings:
*
* 1. **Upstream — default selections land in one `runPending`.**
*    `selectAudioTrack` (default audio) and `switchVideoQuality`
*    (default video) both subscribe to `state.presentation` flipping to
*    resolved; their effects run in the same `runPending` iteration and
*    write `selectedAudioTrackId` + `selectedVideoTrackId` within it.
* 2. **Self — both per-type monitors flip in one `runPending`.**
*    After (1), both monitors re-evaluate and flip to `'buffer-ready'`
*    in the next `runPending`. Both `entry` bodies run synchronously
*    within that iteration — both `addSourceBuffer` calls land before
*    the iteration ends.
* 3. **Downstream — `appendBuffer` is async.** `loadVideoSegments` /
*    `loadAudioSegments` read the per-type `xSegmentLoaderActor` slots;
*    their effects fire in the *next* `runPending` and the actual
*    `appendBuffer` requires a network round-trip via the
*    `SegmentLoaderActor` — many microtasks past both `addSourceBuffer`
*    calls.
*
* The cross-tick failure mode — a user-initiated audio track switch
* *after* video segments have begun appending — is out of scope for
* this behavior and would be addressed in the buffer/segment-loading
* path via `changeType`-aware logic.
*
* # Sole writer
*
* `setupVideoBufferActors` is sole writer of `videoBufferActor` +
* `videoSegmentLoaderActor` (and `bandwidthState` via its
* trackedFetch); `setupAudioBufferActors` is sole writer of
* `audioBufferActor` + `audioSegmentLoaderActor`. Both read
* `mediaSource` from `setupMediaSource`. Downstream MSE behaviors
* (`loadVideoSegments`, `loadAudioSegments`, `endOfStream`,
* `updateMediaSourceDuration`) only read these slots.
*/
function setupBufferActors({ state, context, config }) {
	const { type, selectedKey, actorKey, loaderKey, fetch, forwardBuffer, backBuffer } = config;
	const derivedStateSignal = computed(() => {
		if (!context.mediaSource.get()) return "preconditions-unmet";
		return hasCodecs(getSelectedTrack({
			presentation: state.presentation.get(),
			[selectedKey]: state[selectedKey].get()
		}, type)) ? "buffer-ready" : "preconditions-unmet";
	});
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			"buffer-ready": { entry: () => {
				const bufferActor = createSourceBufferActor(createSourceBuffer(context.mediaSource.get(), buildMimeCodec(getSelectedTrack({
					presentation: state.presentation.get(),
					[selectedKey]: state[selectedKey].get()
				}, type))));
				const segmentLoader = createSegmentLoaderActor(bufferActor, fetch, {
					forwardBuffer,
					backBuffer
				});
				context[actorKey].set(bufferActor);
				context[loaderKey].set(segmentLoader);
				return () => {
					segmentLoader.destroy();
					bufferActor.destroy();
					context[loaderKey].set(void 0);
					context[actorKey].set(void 0);
				};
			} }
		}
	});
}
/**
* Set up the video `SourceBufferActor` + `SegmentLoaderActor`. Fires
* when `mediaSource` is attached and the selected video track is
* present in the presentation with codecs. Gates only on video state —
* no cross-type coupling. Owns a bandwidth-sampling `trackedFetch` and
* is sole writer of `state.bandwidthState`.
*/
const setupVideoBufferActors = defineBehavior({
	stateKeys: [
		"presentation",
		"selectedVideoTrackId",
		"bandwidthState"
	],
	contextKeys: [
		"mediaSource",
		"videoBufferActor",
		"videoSegmentLoaderActor"
	],
	setup: ({ state, context, config = {} }) => {
		const trackedFetch = createTrackedFetch(state.bandwidthState.get() ?? {
			fastEstimate: 0,
			fastTotalWeight: 0,
			slowEstimate: 0,
			slowTotalWeight: 0,
			bytesSampled: 0
		}, (next) => state.bandwidthState.set(next));
		return setupBufferActors({
			state,
			context,
			config: {
				...VIDEO_TYPE_CONFIG,
				fetch: trackedFetch,
				...config
			}
		});
	}
});
/**
* Set up the audio `SourceBufferActor` + `SegmentLoaderActor`. Same
* shape as `setupVideoBufferActors`, narrowed to audio. Today supplies
* a non-sampling `fetchStream` (no audio ABR); adding audio ABR is a
* localized change to this setup body (swap `fetchStream` for a
* `createTrackedFetch` call + declare `bandwidthState` writable here)
* without touching the shared helper.
*/
const setupAudioBufferActors = defineBehavior({
	stateKeys: ["presentation", "selectedAudioTrackId"],
	contextKeys: [
		"mediaSource",
		"audioBufferActor",
		"audioSegmentLoaderActor"
	],
	setup: ({ state, context, config = {} }) => setupBufferActors({
		state,
		context,
		config: {
			...AUDIO_TYPE_CONFIG,
			fetch: fetchStream,
			...config
		}
	})
});

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/setup-mediasource.js
/**
* **Own the MediaSource lifecycle for the current source.** When a resolved
* presentation and a mediaElement are both in scope, creates a MediaSource,
* attaches it to the element, waits for `'open'`, and publishes it on
* `context.mediaSource`. On source change or behavior destroy, detaches the
* MediaSource and clears the slot so the next source starts fresh.
*
* Single-positive-state reactor (`'preconditions-unmet'` ↔ `'mediasource-attached'`):
* state derivation gates on `mediaElement + isResolvedPresentation`. Riding the
* resolver's resolved/unresolved lifecycle makes direct URL replacement
* structural — `resolvePresentation` routes the presentation back through
* unresolved on URL change, which drives this reactor through
* `'preconditions-unmet'` so the entry's state-exit cleanup detaches the old
* MediaSource before the new one is built.
*
* The entry resolves preconditions in sequence before publishing:
*
* 1. **Create + attach** — `createMediaSource` + `attachMediaSource` run
*    synchronously on entry. The `detach` closure returned by
*    `attachMediaSource` is captured for state-exit cleanup, so the cleanup
*    is always bound to its setup even if the wait below is aborted.
* 2. **Wait for `'open'`** — `waitForMediaSourceOpen` defers until the first
*    `sourceopen` event (or any readyState transition out of `'closed'`).
* 3. **Publish on `'open'`** — re-check `readyState === 'open'` after the
*    await (covers `'ended'` / `'closed'` race) before writing to
*    `context.mediaSource`. Downstream `setupVideoBufferActors` /
*    `setupAudioBufferActors` call `addSourceBuffer` directly, which
*    throws on non-open, so publish-only-when-open is the load-bearing
*    contract.
*
* State-exit cleanup aborts the in-flight wait, detaches the MediaSource,
* and clears `context.mediaSource`. Order: abort first (prevents a late
* publish racing the slot clear), then detach, then clear.
*
* Sole writer of `context.mediaSource`; other MSE behaviors
* (`setupVideoBufferActors`, `setupAudioBufferActors`,
* `updateMediaSourceDuration`, `endOfStream`, `loadVideoSegments`) only
* read.
*/
function deriveState$3(presentation, mediaElement) {
	if (!mediaElement || !isResolvedPresentation(presentation)) return "preconditions-unmet";
	return "mediasource-attached";
}
function setupMediaSourceSetup({ state, context }) {
	const derivedStateSignal = computed(() => deriveState$3(state.presentation.get(), context.mediaElement.get()));
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			"mediasource-attached": { entry: () => {
				const mediaElement = context.mediaElement.get();
				const controller = new AbortController();
				const mediaSource = createMediaSource({ preferManaged: true });
				const { detach } = attachMediaSource(mediaSource, mediaElement);
				const publishWhenOpen = async () => {
					await waitForMediaSourceOpen(mediaSource, controller.signal);
					if (controller.signal.aborted) return;
					if (mediaSource.readyState !== "open") {
						console.warn(`[setupMediaSource] MediaSource transitioned to '${mediaSource.readyState}' before first 'sourceopen' — slot left unpublished; recoverable on next source reset.`);
						return;
					}
					context.mediaSource.set(mediaSource);
				};
				publishWhenOpen().catch((err) => console.error("[setupMediaSource] failed to publish MediaSource:", err));
				return () => {
					controller.abort();
					detach();
					context.mediaSource.set(void 0);
				};
			} }
		}
	});
}
const setupMediaSource = defineBehavior({
	stateKeys: ["presentation"],
	contextKeys: ["mediaElement", "mediaSource"],
	setup: setupMediaSourceSetup
});

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/sync-text-tracks.js
/**
* **Own the text-track slots on the host media element, mirroring the SPF
* model.** When a presentation is resolved and a media element is
* available, allocate one slot in `mediaElement.textTracks` per model text
* track — via creating `<track>` children, since that's the only spec
* mechanism for adding *and* removing entries to `textTracks` (no
* `removeTextTrack` API exists). Once slots are provisioned, mirror
* `selectedTextTrackId` into their `mode`s, and propagate user-initiated
* DOM `change` events back to `selectedTextTrackId` so non-SPF consumers
* (host-page captions buttons, browser native UI, video.js store) remain
* the canonical write path for selection.
*
* Single-positive-state reactor (`'preconditions-unmet'` ↔ `'sync-active'`):
* the entry allocates the slots, applies the initial selection, attaches
* the `change` listener, and opens a brief Chromium settling-window guard —
* all transition-driven, fire-once on state entry, with paired cleanup on
* state exit. A single `effects:` mirrors subsequent
* `selectedTextTrackId` changes into `mode`s; that's the only
* continuous-reactivity concern.
*
* State-exit cleanup also sends a `'clear'` message to the
* `TextTracksActor` so its cue+segment cache (keyed by trackId) is
* dropped alongside the DOM `<track>` slots. The actor itself is owned
* by `setupTextTrackActors` and bound to mediaElement, not presentation,
* so it survives source resets; clearing its context here keeps the
* cache consistent with the DOM. Without this, a subsequent
* presentation reusing a trackId would have `getSegmentsToLoad` treat
* its segments as already-buffered and skip loading them.
*
* Multi-writer with `selectTextTrack` is intentional and orthogonal:
* `selectTextTrack` owns the *default-on-load / clear-on-unload* contract
* for `selectedTextTrackId`; this behavior writes only from DOM `change`
* events. They reflect distinct decision-making domains (config-driven
* intent vs DOM-driven user action), so this behavior never clears the
* slot on source unload — that's `selectTextTrack`'s contract. (It *does*
* write `undefined` when a user disables all tracks via native UI, since
* that's a real user action that should round-trip into SPF state.)
*/
function deriveState$2(presentation, mediaElement) {
	if (!mediaElement || !presentation) return "preconditions-unmet";
	return getTracksByType(presentation, "text").length > 0 ? "sync-active" : "preconditions-unmet";
}
function syncTextTracksSetup({ state, context, config }) {
	const { addSubtitlesTracksToMedia, getShowingSubtitlesTrackFromMedia, removeAllSubtitlesTracksFromMedia } = config;
	const derivedStateSignal = computed(() => deriveState$2(state.presentation.get(), context.mediaElement.get()));
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			"sync-active": {
				entry: () => {
					const mediaElement = context.mediaElement.get();
					addSubtitlesTracksToMedia(mediaElement, getTracksByType(state.presentation.get(), "text"));
					syncTextTrackModes(mediaElement.textTracks, state.selectedTextTrackId.get());
					let inSettlingWindow = true;
					const settlingTimeout = setTimeout(() => {
						inSettlingWindow = false;
					}, 0);
					const onChange = () => {
						if (inSettlingWindow) {
							syncTextTrackModes(mediaElement.textTracks, state.selectedTextTrackId.get());
							return;
						}
						const nextId = getShowingSubtitlesTrackFromMedia(mediaElement)?.id || void 0;
						if (nextId === state.selectedTextTrackId.get()) return;
						state.selectedTextTrackId.set(nextId);
					};
					const unlisten = listen(mediaElement.textTracks, "change", onChange);
					return () => {
						unlisten();
						clearTimeout(settlingTimeout);
						removeAllSubtitlesTracksFromMedia(mediaElement);
						peek(context.textTracksActor)?.send({ type: "clear" });
					};
				},
				effects: () => {
					syncTextTrackModes(peek(context.mediaElement).textTracks, state.selectedTextTrackId.get());
				}
			}
		}
	});
}
const syncTextTracks = defineBehavior({
	stateKeys: ["presentation", "selectedTextTrackId"],
	contextKeys: ["mediaElement", "textTracksActor"],
	setup: syncTextTracksSetup
});

//#endregion
//#region ../spf/dist/dev/playback/behaviors/dom/update-mediasource-duration.js
/**
* **Propagate `presentation.duration` to `mediaSource.duration` — exactly
* once per MediaSource.**
*
* When the presentation has a valid positive duration (including `Infinity`
* for live) and a MediaSource is in context, writes the value through to
* `mediaSource.duration` on initial setup — once, while `mediaSource.duration`
* is still `NaN`. Once any non-NaN value is present (set by us, or by
* `endOfStream` from the buffered end), the behavior leaves the property
* alone; re-syncing a drift against `presentation.duration` would race with
* concurrent `appendBuffer()` calls.
*
* The entry resolves three async preconditions in order before writing:
*
* 1. **MediaSource open** — `waitForMediaSourceOpen` defers until the first
*    `sourceopen` event (or any readyState transition, since `'ended'` /
*    `'closed'` mean we've missed the window and should bail).
* 2. **All SourceBuffers idle** — `waitForSourceBuffersReady` defers per
*    the MSE-spec rule that `duration` cannot be set while any buffer has
*    `updating === true`.
* 3. **Buffered-range clamp** — `getMaxBufferedEnd` ensures the written
*    duration is at least the highest buffered end (MSE spec disallows
*    a smaller `duration` than any buffered range).
*
* The buffer-set helpers operate across `mediaSource.sourceBuffers` (the
* canonical aggregate), so the behavior composes uniformly across
* audio-only, video-only, and mixed configurations.
*
* Single-positive-state reactor (`'preconditions-unmet'` ↔ `'duration-writable'`):
* state derivation is purely signal-driven (presentation validity +
* mediaSource existence). MediaSource lifecycle state (`readyState`) and
* the `duration` itself are non-signal DOM properties resolved inside the
* entry's async sequence. The state-exit cleanup aborts the in-flight
* wait, so source resets and behavior destroy structurally cancel the
* pending write. A post-await re-check of `mediaSource.readyState ===
* 'open'` covers the narrow race where `endOfStream()` synchronously
* transitions readyState to `'ended'` between our `waitForMediaSourceOpen`
* resolution and the `mediaSource.duration` write.
*
* Downstream of `calculatePresentationDuration` (which writes
* `presentation.duration`); concurrent with `endOfStream` (which may later
* write `mediaSource.duration` — the "exactly once" contract keeps us out
* of that path).
*/
function deriveState$1(presentation, mediaSource) {
	return shouldUpdateDuration(presentation, mediaSource) ? "duration-writable" : "preconditions-unmet";
}
function updateMediaSourceDurationSetup({ state, context }) {
	const derivedStateSignal = computed(() => deriveState$1(state.presentation.get(), context.mediaSource.get()));
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			"duration-writable": { entry: () => {
				const presentation = state.presentation.get();
				const mediaSource = context.mediaSource.get();
				if (!Number.isNaN(mediaSource.duration)) return;
				const controller = new AbortController();
				const writeWhenReady = async () => {
					await waitForMediaSourceOpen(mediaSource, controller.signal);
					if (controller.signal.aborted) return;
					if (mediaSource.readyState !== "open") return;
					await waitForSourceBuffersReady(mediaSource.sourceBuffers, controller.signal);
					if (controller.signal.aborted) return;
					if (mediaSource.readyState !== "open") return;
					const maxBufferedEnd = getMaxBufferedEnd(mediaSource.sourceBuffers);
					mediaSource.duration = maxBufferedEnd > presentation.duration ? maxBufferedEnd : presentation.duration;
				};
				writeWhenReady();
				return () => controller.abort();
			} }
		}
	});
}
const updateMediaSourceDuration = defineBehavior({
	stateKeys: ["presentation"],
	contextKeys: ["mediaSource"],
	setup: updateMediaSourceDurationSetup
});

//#endregion
//#region ../spf/dist/dev/media/abr/quality-selection.js
/**
* Default quality selection configuration.
* Values match Shaka Player upgrade threshold (0.85 = 15% headroom).
*/
const DEFAULT_QUALITY_CONFIG = {
	safetyMargin: .85,
	upgradeMargin: 1.15
};
/**
* Select the track to apply now, given current bandwidth, a current
* selection (optional), and tuning. Returns:
*
* - The bandwidth-fitting optimal when no `currentTrack` is supplied.
* - The optimal when it's a downgrade vs. `currentTrack` (downgrades
*   apply immediately — no hysteresis).
* - The optimal when it clears `currentTrack.bandwidth * upgradeMargin`
*   (upgrade clears hysteresis).
* - `currentTrack` itself when an upgrade doesn't clear the margin
*   (stay put — caller checks identity to no-op).
*
* "Optimal" is the highest-bandwidth track where the available bandwidth
* meets the safety requirement (`currentBandwidth >= track.bandwidth / safetyMargin`).
* Falls back to the lowest-bandwidth track when nothing fits the safety
* margin (preserves a definitive pick under under-bandwidth conditions).
*
* @example
* const tracks = [low, mid, high];
* selectQuality(tracks, 5_000_000, { currentTrack: low });
* // Returns `high` if 5 Mbps clears safety AND high.bandwidth >= low.bandwidth * upgradeMargin.
* // Returns `low` (no-op signal) otherwise.
*/
function selectQuality(tracks, currentBandwidth, opts = {}) {
	if (tracks.length === 0) return;
	const safetyMargin = opts.safetyMargin ?? DEFAULT_QUALITY_CONFIG.safetyMargin;
	const upgradeMargin = opts.upgradeMargin ?? DEFAULT_QUALITY_CONFIG.upgradeMargin;
	const { currentTrack } = opts;
	const sortedTracks = tracks.slice().sort((a, b) => a.bandwidth - b.bandwidth);
	let chosen;
	for (const track of sortedTracks) if (currentBandwidth >= track.bandwidth / safetyMargin) {
		if (!chosen || track.bandwidth > chosen.bandwidth || track.bandwidth === chosen.bandwidth && hasHigherResolution(track, chosen)) chosen = track;
	}
	const optimal = chosen ?? sortedTracks[0];
	if (!optimal) return void 0;
	if (!currentTrack) return optimal;
	if (optimal.bandwidth < currentTrack.bandwidth) return optimal;
	if (optimal.bandwidth >= currentTrack.bandwidth * upgradeMargin) return optimal;
	return currentTrack;
}
/**
* Select the lowest-bandwidth track from the candidate set.
*
* Used as a safety-net fallback by callers that need a definitive pick when
* a primary selection algorithm declines to choose. Pairs naturally with
* `selectQuality` — when ABR has no good answer (e.g., all candidates
* exceed available bandwidth and the algorithm doesn't fall back
* internally), the lowest bitrate is the safest default.
*
* @param tracks - Candidate tracks (can be unsorted)
* @returns The lowest-bandwidth track, or `undefined` if `tracks` is empty
*
* @example
* const fallback = selectLowestQuality(tracks);
*/
function selectLowestQuality(tracks) {
	if (tracks.length === 0) return void 0;
	return tracks.reduce((min, t) => t.bandwidth < min.bandwidth ? t : min);
}
/**
* Check if track A has higher resolution than track B.
* Compares by total pixel count (width × height).
*
* @param trackA - First track to compare
* @param trackB - Second track to compare
* @returns True if trackA has more pixels than trackB
*/
function hasHigherResolution(trackA, trackB) {
	return (trackA.width ?? 0) * (trackA.height ?? 0) > (trackB.width ?? 0) * (trackB.height ?? 0);
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/quality-switching.js
function setupQualitySwitching({ state, config }) {
	const safetyMargin = config.quality?.safetyMargin ?? DEFAULT_QUALITY_CONFIG.safetyMargin;
	const upgradeMargin = config.quality?.upgradeMargin ?? DEFAULT_QUALITY_CONFIG.upgradeMargin;
	const initialBandwidth = config.initialBandwidth ?? 5e6;
	const bandwidthConfig = {
		...DEFAULT_BANDWIDTH_CONFIG,
		...config.bandwidth
	};
	const { selectionKey, userSelectionKey, getTracks, selectOptimal } = config;
	const derivedStateSignal = computed(() => isResolvedPresentation(state.presentation.get()) ? "presentation-resolved" : "presentation-unresolved");
	return createMachineReactor({
		initial: "presentation-unresolved",
		monitor: () => derivedStateSignal.get(),
		states: {
			"presentation-unresolved": {},
			"presentation-resolved": {
				entry: () => () => state[selectionKey].set(void 0),
				effects: [() => {
					const presentation = peek(state.presentation);
					if (!presentation) return;
					const allTracks = getTracks(presentation);
					const [firstAllTrack] = allTracks;
					if (!firstAllTrack) return;
					const userFilter = state[userSelectionKey].get();
					const matching = userFilter ? allTracks.filter((track) => {
						for (const key in userFilter) {
							const filterValue = userFilter[key];
							if (filterValue !== void 0 && track[key] !== filterValue) return false;
						}
						return true;
					}) : allTracks;
					const candidates = matching.length > 0 ? matching : allTracks;
					if (!candidates.length) return;
					const selectedId = state[selectionKey].get();
					if (candidates.length === 1) {
						if (candidates[0].id !== selectedId) state[selectionKey].set(candidates[0].id);
						return;
					}
					const bandwidth = getBandwidthEstimate(state.bandwidthState.get(), initialBandwidth, bandwidthConfig);
					if (!selectedId && config.picker) {
						const id = config.picker(presentation, config);
						if (id) {
							state[selectionKey].set(id);
							return;
						}
					}
					const optimal = selectOptimal(candidates, bandwidth, {
						safetyMargin,
						upgradeMargin,
						currentTrack: candidates.find((t) => t.id === selectedId)
					}) ?? selectLowestQuality(candidates);
					if (optimal.id !== selectedId) state[selectionKey].set(optimal.id);
				}]
			}
		}
	});
}
/**
* Manage `selectedVideoTrackId`: pick a default on src load, dynamically
* adjust based on bandwidth, clear on src unload. Honors
* `userVideoTrackSelection` as a partial-track constraint on candidates;
* short-circuits ABR when the constraint narrows to a single track.
*
* @example
* const reactor = switchVideoQuality.setup({ state });
*/
const switchVideoQuality = defineBehavior({
	stateKeys: [
		"presentation",
		"bandwidthState",
		"selectedVideoTrackId",
		"userVideoTrackSelection"
	],
	contextKeys: [],
	setup: ({ state, config }) => setupQualitySwitching({
		state,
		config: {
			...config,
			selectionKey: "selectedVideoTrackId",
			userSelectionKey: "userVideoTrackSelection",
			getTracks: (presentation) => getTracksByType(presentation, "video"),
			selectOptimal: selectQuality
		}
	})
});

//#endregion
//#region ../spf/dist/dev/media/utils/preload.js
function isStandardPreload(value) {
	return value === "auto" || value === "metadata" || value === "none";
}
/**
* Default `preload` value used as the fallback across behaviors
* (`syncPreload`, `resolvePresentation`, `isBlockingPreload`). Matches the
* `<video>`/`<audio>` element's implicit default.
*/
const DEFAULT_PRELOAD = "metadata";
/**
* True when the preload value blocks initial resolution / loading.
* Falsy values (undefined, empty) fall back to `defaultPreload` (default
* `DEFAULT_PRELOAD`); the resolved value blocks iff it is `'none'`.
*/
function isBlockingPreload(preload, defaultPreload = DEFAULT_PRELOAD) {
	return (preload || defaultPreload) === "none";
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/resolve-presentation.js
/**
* **Resolve an unresolved presentation by fetching and parsing its manifest.**
*
* Reads `state.presentation`; when it holds `{ url }` (unresolved) and the
* preload / load-activation gate is met, fetches the manifest, parses it via
* the **required** `config.parsePresentation`, and writes the resolved
* `Presentation` back to the same slot. The behavior is format-neutral: the
* composing engine wires in its parser (e.g. the HLS engine supplies the
* multivariant-playlist parser).
*
* Source-identity-driven, expressed as a 4-state machine:
*
* ```
* 'preconditions-unmet' → 'idle' → 'resolving' → 'resolved'
* ```
*
* - `'preconditions-unmet'`: no presentation, or presentation has no URL.
* - `'idle'`: URL present, unresolved, gate unmet (blocking preload + no
*   load-activation). Waits for the gate to open.
* - `'resolving'`: URL present, unresolved, gate met. Entry starts the fetch
*   and returns the AbortController — the reactor calls `.abort()` on state
*   exit, so source change / gate-close / destroy all cancel cleanly.
* - `'resolved'`: `state.presentation` holds a resolved `Presentation`.
*
* Gate semantics: `state.preload` (or `config.defaultPreload`, default
* `'metadata'`, when state.preload is unset) blocks resolution when its
* value is `'none'` (see `isBlockingPreload` in `media/utils/preload`).
* `state.loadActivated` is an override — true bypasses the preload gate
* entirely.
*
* Multi-writer with the engine adapter, which writes the initial unresolved
* `{ url }` to `state.presentation` from src input. Different domains
* (config-input vs. derived state via fetch) — legitimate multi-writer.
*/
function deriveState(presentation, preload, loadActivated, defaultPreload) {
	if (!presentation?.url) return "preconditions-unmet";
	if (isResolvedPresentation(presentation)) return "resolved";
	return !!loadActivated || !isBlockingPreload(preload, defaultPreload) ? "resolving" : "idle";
}
function resolvePresentationSetup({ state, config }) {
	const { parsePresentation } = config;
	const defaultPreload = config.defaultPreload ?? "metadata";
	const derivedStateSignal = computed(() => deriveState(state.presentation.get(), state.preload.get(), state.loadActivated.get(), defaultPreload));
	return createMachineReactor({
		initial: "preconditions-unmet",
		monitor: () => derivedStateSignal.get(),
		states: {
			"preconditions-unmet": {},
			idle: {},
			resolving: { entry: () => {
				const presentation = state.presentation.get();
				const ac = new AbortController();
				fetchResolvable(presentation, { signal: ac.signal }).then((response) => getResponseText(response)).then((text) => {
					const parsed = parsePresentation(text, presentation);
					state.presentation.set(parsed);
				}).catch((error) => {
					if (error instanceof Error && error.name === "AbortError") return;
					console.error("[resolvePresentation] manifest fetch/parse failed:", error);
				});
				return ac;
			} },
			resolved: {}
		}
	});
}
const resolvePresentation = defineBehavior({
	stateKeys: [
		"presentation",
		"preload",
		"loadActivated"
	],
	contextKeys: [],
	setup: resolvePresentationSetup
});

//#endregion
//#region ../spf/dist/dev/media/hls/parse-media-playlist.js
/**
* Parse HLS media playlist and resolve track with segments.
*
* Takes an unresolved track (from multivariant playlist) and media playlist text,
* returns a HAM-compliant resolved track with segments.
*
* @param text - Media playlist text content
* @param unresolved - Unresolved track from parseMultivariantPlaylist
* @returns Resolved track with segments (type inferred from input)
*/
function parseMediaPlaylist(text, unresolved) {
	const lines = text.split(/\r?\n/);
	const baseUrl = unresolved.url;
	const segments = [];
	let initSegmentUrl;
	let initSegmentByteRange;
	let currentDuration = 0;
	let currentByteRange;
	let currentTime = 0;
	let segmentIndex = 0;
	let previousByteRangeEnd;
	for (const line of lines) {
		const trimmed = line.trim();
		if (!trimmed || trimmed.startsWith("#") && !trimmed.startsWith("#EXT")) continue;
		if (trimmed === "#EXTM3U" || trimmed.startsWith("#EXT-X-VERSION:") || trimmed.startsWith("#EXT-X-TARGETDURATION:") || trimmed.startsWith("#EXT-X-PLAYLIST-TYPE:") || trimmed.startsWith("#EXT-X-INDEPENDENT-SEGMENTS")) continue;
		const mapAttrs = matchTag(trimmed, "EXT-X-MAP");
		if (mapAttrs) {
			const uri = mapAttrs.get("URI");
			if (uri) {
				initSegmentUrl = resolveUrl(uri, baseUrl);
				const byteRangeStr = mapAttrs.get("BYTERANGE");
				if (byteRangeStr) initSegmentByteRange = parseByteRange(byteRangeStr, 0) ?? void 0;
			}
			continue;
		}
		if (trimmed.startsWith("#EXTINF:")) {
			currentDuration = parseExtInfDuration(trimmed.slice(8));
			continue;
		}
		if (trimmed.startsWith("#EXT-X-BYTERANGE:")) {
			currentByteRange = parseByteRange(trimmed.slice(17), previousByteRangeEnd) ?? void 0;
			continue;
		}
		if (trimmed === "#EXT-X-ENDLIST") continue;
		if (!trimmed.startsWith("#") && currentDuration > 0) {
			const segment = {
				id: `segment-${segmentIndex}`,
				url: resolveUrl(trimmed, baseUrl),
				duration: currentDuration,
				startTime: currentTime
			};
			if (currentByteRange) {
				segment.byteRange = currentByteRange;
				previousByteRangeEnd = currentByteRange.end + 1;
			} else previousByteRangeEnd = void 0;
			segments.push(segment);
			currentTime += currentDuration;
			segmentIndex++;
			currentDuration = 0;
			currentByteRange = void 0;
		}
	}
	const totalDuration = currentTime;
	const initialization = unresolved.type === "text" && !initSegmentUrl ? void 0 : initSegmentUrl ? {
		url: initSegmentUrl,
		...initSegmentByteRange ? { byteRange: initSegmentByteRange } : {}
	} : { url: "" };
	return {
		...unresolved,
		startTime: 0,
		duration: totalDuration,
		segments,
		initialization
	};
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/resolve-track.js
function setupTrackResolution({ state, config: { selectedKey, findTrackToResolve } }) {
	const runner = new ConcurrentRunner();
	const derivedStateSignal = computed(() => isResolvedPresentation(state.presentation.get()) ? "presentation-resolved" : "presentation-unresolved");
	return createMachineReactor({
		initial: "presentation-unresolved",
		monitor: () => derivedStateSignal.get(),
		states: {
			"presentation-unresolved": {},
			"presentation-resolved": {
				entry: () => () => runner.abortAll(),
				effects: [() => {
					const presentation = peek(state.presentation);
					const trackId = state[selectedKey].get();
					if (!presentation || !trackId) return;
					const track = findTrackToResolve(presentation, trackId);
					if (!track || isResolvedTrack(track)) return;
					runner.schedule(new Task(async (signal) => {
						const mediaTrack = parseMediaPlaylist(await getResponseText(await fetchResolvable(track, { signal })), track);
						update(state.presentation, (current) => isResolvedPresentation(current) ? updateTrackInPresentation(current, mediaTrack) : current);
					}, { id: track.id }));
				}]
			}
		}
	});
}
const VIDEO_TRACK_RESOLUTION_CONFIG = {
	...VIDEO_TYPE_CONFIG,
	findTrackToResolve: (presentation, trackId) => findTrack(presentation, "video", trackId)
};
const AUDIO_TRACK_RESOLUTION_CONFIG = {
	...AUDIO_TYPE_CONFIG,
	findTrackToResolve: (presentation, trackId) => findTrack(presentation, "audio", trackId)
};
const TEXT_TRACK_RESOLUTION_CONFIG = {
	...TEXT_TYPE_CONFIG,
	findTrackToResolve: (presentation, trackId) => findTrack(presentation, "text", trackId)
};
/**
* Resolve unresolved video tracks. Schedules a fetch task whenever the
* selected video track is partially resolved, parses the manifest, and
* writes the resolved track back into `state.presentation`.
*/
const resolveVideoTrack = defineBehavior({
	stateKeys: ["presentation", "selectedVideoTrackId"],
	contextKeys: [],
	setup: ({ state, config = {} }) => setupTrackResolution({
		state,
		config: {
			...VIDEO_TRACK_RESOLUTION_CONFIG,
			...config
		}
	})
});
/**
* Resolve unresolved audio tracks. Same shape as `resolveVideoTrack`,
* narrowed to audio.
*/
const resolveAudioTrack = defineBehavior({
	stateKeys: ["presentation", "selectedAudioTrackId"],
	contextKeys: [],
	setup: ({ state, config = {} }) => setupTrackResolution({
		state,
		config: {
			...AUDIO_TRACK_RESOLUTION_CONFIG,
			...config
		}
	})
});
/**
* Resolve unresolved text tracks. Same shape as `resolveVideoTrack`,
* narrowed to text.
*/
const resolveTextTrack = defineBehavior({
	stateKeys: ["presentation", "selectedTextTrackId"],
	contextKeys: [],
	setup: ({ state, config = {} }) => setupTrackResolution({
		state,
		config: {
			...TEXT_TRACK_RESOLUTION_CONFIG,
			...config
		}
	})
});

//#endregion
//#region ../spf/dist/dev/media/primitives/select-tracks.js
/**
* Pick the first track of the given type from a presentation.
*
* Returns the first track in the first switching set of the matching
* selection set, or `undefined` if either is missing. POC-shaped
* default-pick — `pickVideoTrack` / `pickAudioTrack` honor bandwidth +
* language preferences and will replace this once selection callers are
* ready.
*/
function pickFirstTrackId(presentation, type) {
	return presentation.selectionSets?.find((set) => set.type === type)?.switchingSets[0]?.tracks[0]?.id;
}
/**
* Pick text track to activate. Conforms to the `TrackPicker` contract so it
* can be used directly as a default picker for `selectTextTrack` without an
* adapter wrapper.
*
* Selection priority (if enabled):
* 1. User preference (preferredSubtitleLanguage)
* 2. DEFAULT track (if enableDefaultTrack is true and track has DEFAULT=YES + AUTOSELECT=YES)
* 3. No auto-selection (user opt-in)
*
* By default, FORCED tracks are excluded per Apple's HLS spec.
*/
function pickTextTrack(presentation, config) {
	const textSet = presentation.selectionSets?.find((set) => set.type === "text");
	if (!textSet?.switchingSets?.[0]?.tracks.length) return void 0;
	const tracks = textSet.switchingSets[0].tracks;
	const availableTracks = config?.includeForcedTracks ? tracks : tracks.filter((track) => !track.forced);
	if (availableTracks.length === 0) return void 0;
	const { preferredSubtitleLanguage, enableDefaultTrack = false } = config ?? {};
	if (preferredSubtitleLanguage) {
		const languageMatch = availableTracks.find((track) => track.language === preferredSubtitleLanguage);
		if (languageMatch) return languageMatch.id;
	}
	if (enableDefaultTrack) {
		const defaultTrack = availableTracks.find((track) => track.default === true);
		if (defaultTrack) return defaultTrack.id;
	}
}

//#endregion
//#region ../spf/dist/dev/playback/behaviors/select-tracks.js
/**
* **Default audio/text track selection on src load / unselect on src unload.**
* When a presentation is resolved, sets `selectedAudioTrackId` /
* `selectedTextTrackId` to a per-type-picker default if no selection already
* exists. When the presentation is unset/reset (transitions back to
* unresolved), clears the selection so a stale id from the previous source
* doesn't persist.
*
* Lifecycle-driven: each transition fires its work once. Does not police
* the selection between transitions; external writes (user picks) are left
* alone.
*
* Picker is config-driven: each per-type export wires a sensible default
* (`pickFirstTrackId` for audio, `pickTextTrack` for text) and the caller
* can supply their own via `config.picker` for custom selection logic
* (language preferences, default-track handling, etc.). The behavior's
* `config` is forwarded to the picker as its second argument, so options
* like `preferredAudioLanguage` / `preferredSubtitleLanguage` reach the
* picker without an intermediate wrapping layer.
*
* Compose `selectVideoTrack` for the simple "pick a default video track"
* behavior, or `switchVideoQuality` (`./quality-switching.ts`) for the
* ABR-driven variant — they're alternatives, not stackable (both write
* `selectedVideoTrackId`). The simple variant tree-shakes out the
* ABR machinery (bandwidth-estimator, quality-selection algorithms).
*/
function setupTrackSelection({ state, config: { selectedKey, picker, pickerConfig } }) {
	const derivedStateSignal = computed(() => isResolvedPresentation(state.presentation.get()) ? "presentation-resolved" : "presentation-unresolved");
	return createMachineReactor({
		initial: "presentation-unresolved",
		monitor: () => derivedStateSignal.get(),
		states: {
			"presentation-unresolved": {},
			"presentation-resolved": { entry: () => {
				if (!state[selectedKey].get()) {
					const id = picker(state.presentation.get(), pickerConfig);
					if (id) state[selectedKey].set(id);
				}
				return () => state[selectedKey].set(void 0);
			} }
		}
	});
}
/** Default video picker: first track in the video selection set. */
const defaultVideoPicker = (presentation) => pickFirstTrackId(presentation, "video");
const VIDEO_TRACK_SELECTION_CONFIG = {
	...VIDEO_TYPE_CONFIG,
	picker: defaultVideoPicker
};
/** Default audio picker: first track in the audio selection set. */
const defaultAudioPicker = (presentation) => pickFirstTrackId(presentation, "audio");
const AUDIO_TRACK_SELECTION_CONFIG = {
	...AUDIO_TYPE_CONFIG,
	picker: defaultAudioPicker
};
const TEXT_TRACK_SELECTION_CONFIG = {
	...TEXT_TYPE_CONFIG,
	picker: pickTextTrack
};
defineBehavior({
	stateKeys: ["presentation", "selectedVideoTrackId"],
	contextKeys: [],
	setup: ({ state, config }) => setupTrackSelection({
		state,
		config: {
			...VIDEO_TRACK_SELECTION_CONFIG,
			picker: config?.picker ?? VIDEO_TRACK_SELECTION_CONFIG.picker,
			pickerConfig: config
		}
	})
});
/**
* Select an audio track when a presentation loads. Clears the selection
* on src unload.
*
* @example
* const reactor = selectAudioTrack.setup({ state });
*
* @example
* // Custom picker with language preference
* const reactor = selectAudioTrack.setup({
*   state,
*   config: { preferredAudioLanguage: 'en', picker: myLanguageAwarePicker },
* });
*/
const selectAudioTrack = defineBehavior({
	stateKeys: ["presentation", "selectedAudioTrackId"],
	contextKeys: [],
	setup: ({ state, config }) => setupTrackSelection({
		state,
		config: {
			...AUDIO_TRACK_SELECTION_CONFIG,
			picker: config?.picker ?? AUDIO_TRACK_SELECTION_CONFIG.picker,
			pickerConfig: config
		}
	})
});
/**
* Select a text track based on user preferences (preferred language,
* default-track auto-select, forced-track filtering). Clears the selection
* on src unload.
*
* Unlike audio selection, the default text picker returns `undefined` when
* no preference matches, leaving the selection unset — text-track
* selection is user opt-in.
*
* @example
* const reactor = selectTextTrack.setup({ state, config: { preferredSubtitleLanguage: 'en' } });
*/
const selectTextTrack = defineBehavior({
	stateKeys: ["presentation", "selectedTextTrackId"],
	contextKeys: [],
	setup: ({ state, config }) => setupTrackSelection({
		state,
		config: {
			...TEXT_TRACK_SELECTION_CONFIG,
			picker: config?.picker ?? TEXT_TRACK_SELECTION_CONFIG.picker,
			pickerConfig: config
		}
	})
});

//#endregion
//#region ../spf/dist/dev/playback/behaviors/sync-preload.js
/**
* **Bidirectional sync between `state.preload` and `mediaElement.preload`.**
*
* Two effects:
* - **Read (DOM → state)** — on `context.mediaElement` swap or
*   `state.presentation.url` change, copies `mediaElement.preload` into
*   `state.preload` if it's a W3C value and `state.preload` isn't holding
*   an extended (non-W3C) value. When the DOM has no W3C opinion and
*   `state.preload` is undefined, backfills from `config.defaultPreload`
*   (default-default `'metadata'`) so `state.preload` is never undefined
*   in steady state.
* - **Write (state → DOM)** — on `state.preload` change or
*   `context.mediaElement` swap, writes `state.preload` back to
*   `mediaElement.preload` if the value is W3C.
*
* Extended values (e.g. `'canplay'`) written externally to `state.preload`
* are sticky: read won't overwrite them, write won't push them to the DOM.
* All writes are deduped to break echo loops and avoid spurious re-triggers
* downstream (notably `resolvePresentation`, which reads `state.preload`).
*/
function syncPreloadSetup({ state, context, config }) {
	const defaultPreload = config?.defaultPreload ?? "metadata";
	const presentationUrl = computed(() => state.presentation.get()?.url);
	const cleanupRead = effect(() => {
		presentationUrl.get();
		const mediaElement = context.mediaElement.get();
		const current = peek(state.preload);
		if (current !== void 0 && !isStandardPreload(current)) return;
		const target = mediaElement && isStandardPreload(mediaElement.preload) ? mediaElement.preload : current === void 0 ? defaultPreload : void 0;
		if (target === void 0 || target === current) return;
		state.preload.set(target);
	});
	const cleanupWrite = effect(() => {
		const next = state.preload.get();
		const mediaElement = context.mediaElement.get();
		if (!mediaElement) return;
		if (!isStandardPreload(next)) return;
		if (mediaElement.preload === next) return;
		mediaElement.preload = next;
	});
	return () => {
		cleanupRead();
		cleanupWrite();
	};
}
const syncPreload = defineBehavior({
	stateKeys: ["preload", "presentation"],
	contextKeys: ["mediaElement"],
	setup: syncPreloadSetup
});

//#endregion
//#region ../spf/dist/dev/playback/engines/hls/engine.js
/**
* Generic `shareSignals` instantiated against the HLS engine's full state
* and context — captures composition signal refs into the consumer's
* `onSignalsReady` callback at setup time.
*/
const shareSignals = makeShareSignals();
/**
* Create an HLS playback engine.
*
* Composes SPF behaviors into a reactive pipeline for HLS playback over MSE:
* manifest resolution, track selection, ABR, segment loading, and
* end-of-stream coordination.
*
* @example
* ```ts
* let signals: SimpleHlsEngineSignals;
* const engine = createSimpleHlsEngine({
*   initialBandwidth: 2_000_000,
*   preferredAudioLanguage: 'en',
*   onSignalsReady: (refs) => {
*     signals = refs;
*   },
* });
*
* signals.context.mediaElement.set(videoEl);
* signals.state.presentation.set({ url: 'https://example.com/stream.m3u8' });
*
* videoEl.play();
*
* await engine.destroy();
* ```
*/
function createSimpleHlsEngine(config = {}) {
	const finalConfig = {
		...config,
		resolveTextTrackSegment: config.resolveTextTrackSegment ?? resolveVttSegment,
		resolveDuration: config.resolveDuration ?? getResolvedSelectedTrackDuration,
		parsePresentation: config.parsePresentation ?? parseMultivariantPlaylist,
		addSubtitlesTracksToMedia: config.addSubtitlesTracksToMedia ?? addSubtitlesTracksToMedia,
		getShowingSubtitlesTrackFromMedia: config.getShowingSubtitlesTrackFromMedia ?? getShowingSubtitlesTrackFromMedia,
		removeAllSubtitlesTracksFromMedia: config.removeAllSubtitlesTracksFromMedia ?? removeAllSubtitlesTracksFromMedia
	};
	return createComposition([
		syncPreload,
		trackLoadTriggers,
		resolvePresentation,
		selectAudioTrack,
		selectTextTrack,
		resolveVideoTrack,
		resolveAudioTrack,
		resolveTextTrack,
		calculatePresentationDuration,
		setupMediaSource,
		updateMediaSourceDuration,
		setupVideoBufferActors,
		setupAudioBufferActors,
		trackCurrentTime,
		switchVideoQuality,
		loadVideoSegments,
		loadAudioSegments,
		endOfStream,
		syncTextTracks,
		setupTextTrackActors,
		loadTextTrackSegments,
		shareSignals
	], {
		config: finalConfig,
		initialState: { bandwidthState: {
			fastEstimate: 0,
			fastTotalWeight: 0,
			slowEstimate: 0,
			slowTotalWeight: 0,
			bytesSampled: 0
		} }
	});
}

//#endregion
//#region ../spf/dist/dev/playback/engines/hls/adapter.js
const simpleHlsMediaDefaultProps = {
	src: "",
	preload: ""
};
/**
* Mixin that adds SPF playback engine behavior to any base class.
*
* Implements the src/play() contract per the WHATWG HTML spec so that SPF can
* be used anywhere a media element API is expected.
*
* A new engine is created on every src assignment — this fully tears down all
* state, SourceBuffers, and in-flight requests from the previous source before
* the next one begins. The media element reference is preserved across src
* changes and re-applied to the new engine automatically.
*
* @example
* class SimpleHlsMedia extends SimpleHlsMediaMixin(HTMLVideoElementHost) {}
*
* const media = new SimpleHlsMedia();
* media.attach(document.querySelector('video'));
* media.src = 'https://stream.mux.com/abc123.m3u8';
*/
function SimpleHlsMediaMixin(BaseClass) {
	class SimpleHlsMediaImpl extends BaseClass {
		#engine;
		#config;
		#signals;
		#preload = simpleHlsMediaDefaultProps.preload;
		/** Pending loadstart listener from a deferred play() retry, if any. */
		#loadstartListener = null;
		constructor(...args) {
			super(...args);
			const { config } = args?.[0] ?? {};
			this.#config = config;
			this.#engine = this.#createEngine();
		}
		get engine() {
			return this.#engine;
		}
		attach(mediaElement) {
			super.attach?.(mediaElement);
			this.#signals.context.mediaElement.set(mediaElement);
		}
		detach() {
			this.#cancelPendingPlay();
			this.#signals.context.mediaElement.set(void 0);
			super.detach?.();
		}
		destroy() {
			this.#cancelPendingPlay();
			this.#engine.destroy();
		}
		get preload() {
			return this.#preload;
		}
		set preload(value) {
			this.#preload = value;
			if (value) this.#signals.state.preload.set(value);
		}
		get src() {
			return this.#signals.state.presentation.get()?.url ?? "";
		}
		set src(value) {
			const prevMediaElement = this.#signals.context.mediaElement.get();
			this.#cancelPendingPlay();
			this.#engine.destroy();
			this.#engine = this.#createEngine();
			if (this.#preload) this.#signals.state.preload.set(this.#preload);
			if (prevMediaElement) this.#signals.context.mediaElement.set(prevMediaElement);
			if (value) this.#signals.state.presentation.set({ url: value });
		}
		play() {
			const mediaElement = this.#signals.context.mediaElement.get();
			if (!mediaElement) return Promise.reject(/* @__PURE__ */ new Error("SimpleHlsMediaElement: no media element attached"));
			this.#signals.state.loadActivated.set(true);
			return mediaElement.play().catch((err) => {
				if (this.src) return new Promise((resolve, reject) => {
					const listener = () => {
						this.#loadstartListener = null;
						mediaElement.play().then(resolve, reject);
					};
					this.#loadstartListener = listener;
					mediaElement.addEventListener("loadstart", listener, { once: true });
				});
				throw err;
			});
		}
		#createEngine() {
			return createSimpleHlsEngine({
				...this.#config,
				onSignalsReady: (signals) => {
					this.#signals = signals;
				}
			});
		}
		#cancelPendingPlay() {
			if (!this.#loadstartListener) return;
			this.#signals.context.mediaElement.get()?.removeEventListener("loadstart", this.#loadstartListener);
			this.#loadstartListener = null;
		}
	}
	return SimpleHlsMediaImpl;
}
/** Standalone SPF media adapter with no base class. */
var SimpleHlsMediaElement = class extends SimpleHlsMediaMixin(class {}) {};

//#endregion
//#region ../core/dist/dev/dom/media/simple-hls/index.js
var SimpleHlsMedia = class extends SimpleHlsMediaMixin(HTMLVideoElementHost) {};

//#endregion
//#region src/media/simple-hls-video/index.ts
var SimpleHlsVideo = class extends MediaAttachMixin(CustomMediaElement("video", SimpleHlsMedia)) {};

//#endregion
//#region src/define/media/simple-hls-video.ts
var SimpleHlsVideoElement = class extends SimpleHlsVideo {
	static {
		this.tagName = "simple-hls-video";
	}
};
safeDefine(SimpleHlsVideoElement);

//#endregion
