// @videojs/html/dist/dev/ui/playback-rate-menu/playback-rate-menu-trigger-element.js
import { playerContext } from "../../player/context.js";
import { MediaElement } from "../media-element.js";
import { PlayerController } from "../../player/player-controller.js";
import { applyElementProps, applyStateDataAttrs, logMissingFeature, selectPlaybackRate } from "@videojs/core/dom";
import { PlaybackRateMenuCore, PlaybackRateMenuDataAttrs } from "@videojs/core";
//#region src/ui/playback-rate-menu/playback-rate-menu-trigger-element.ts
var PlaybackRateMenuTriggerElement = class extends MediaElement {
	constructor(..._args) {
		super(..._args);
		this.label = "";
		this.disabled = false;
		this.commandfor = void 0;
		this.formatRate = PlaybackRateMenuCore.defaultProps.formatRate;
		this.#core = new PlaybackRateMenuCore();
		this.#mediaState = new PlayerController(this, playerContext, selectPlaybackRate);
		this.#disconnect = null;
		this.#handleClick = (event) => {
			if (this.#mediaState.value && !this.#core.state.current.disabled) return;
			event.preventDefault();
			event.stopImmediatePropagation();
		};
		this.#handleKeyDown = (event) => {
			if (event.target !== event.currentTarget) return;
			if (!this.#mediaState.value || this.#core.state.current.disabled) {
				if (event.key !== "Tab") event.preventDefault();
				return;
			}
			if (event.key === "Enter" || event.key === " ") {
				event.preventDefault();
				this.click();
			}
		};
	}
	static {
		this.tagName = "media-playback-rate-menu-trigger";
	}
	static {
		this.properties = {
			label: { type: String },
			disabled: { type: Boolean },
			commandfor: { type: String }
		};
	}
	#core;
	#mediaState;
	#disconnect;
	connectedCallback() {
		super.connectedCallback();
		if (this.destroyed) return;
		this.#disconnect = new AbortController();
		applyElementProps(this, {
			onClick: this.#handleClick,
			onKeyDown: this.#handleKeyDown
		}, { signal: this.#disconnect.signal });
		if (!this.#mediaState.value && this.#mediaState.displayName) logMissingFeature(this.localName, this.#mediaState.displayName);
	}
	disconnectedCallback() {
		super.disconnectedCallback();
		this.#disconnect?.abort();
		this.#disconnect = null;
	}
	/** Returns the trigger's current label derived from media state. */
	getLabel() {
		return this.#core.state.current.label || void 0;
	}
	update(changed) {
		super.update(changed);
		const media = this.#mediaState.value;
		if (!media) return;
		this.#core.setProps(this);
		this.#core.setMedia(media);
		const state = this.#core.getState();
		applyElementProps(this, {
			role: "button",
			tabIndex: 0,
			...this.#core.getAttrs(state)
		});
		applyStateDataAttrs(this, state, PlaybackRateMenuDataAttrs);
	}
	#handleClick;
	#handleKeyDown;
};
//#endregion
export { PlaybackRateMenuTriggerElement };
