// @videojs/core/dist/dev/core/ui/input-feedback/volume-indicator-core.js
import { IndicatorCloseController, getIndicatorCloseDelay } from "./indicator-lifecycle.js";
import { DEFAULT_INPUT_INDICATOR_LABELS, deriveVolumeStatus, isVolumeIndicatorAction, predictVolumeActionOutcome } from "./status.js";
import { createState } from "@videojs/store";
//#region src/core/ui/input-feedback/volume-indicator-core.ts
const BOUNDARY_CLEAR_DELAY = 300;
const INITIAL_STATE = {
	open: false,
	generation: 0,
	level: null,
	value: null,
	fill: null,
	min: false,
	max: false,
	transitionStarting: false,
	transitionEnding: false
};
var VolumeIndicatorCore = class {
	state = createState({ ...INITIAL_STATE });
	#props = {};
	#boundaryTimer = null;
	#boundaryRestartTimer = null;
	#close = new IndicatorCloseController(() => this.state.patch({
		open: false,
		level: null,
		value: null,
		fill: null,
		min: false,
		max: false
	}), () => getIndicatorCloseDelay(this.#props));
	setProps(props) {
		this.#props = props;
	}
	destroy() {
		this.#close.destroy();
		this.#clearBoundaryTimers();
	}
	close() {
		this.#clearBoundaryTimers();
		this.#close.close();
	}
	processEvent(event, snapshot) {
		if (!isVolumeIndicatorAction(event.action)) return false;
		const current = this.state.current;
		const prediction = predictVolumeActionOutcome(event, snapshot);
		const details = deriveVolumeStatus(event, snapshot, DEFAULT_INPUT_INDICATOR_LABELS, prediction);
		const boundary = getVolumeBoundary(event, prediction.snapshotVolume, prediction.nextVolume);
		const repeatedBoundary = boundary !== null && current[boundary] === true;
		if (!boundary) this.#clearBoundaryTimers();
		this.state.patch({
			open: true,
			generation: current.generation + 1,
			level: details.volumeLevel,
			value: details.value,
			fill: details.value,
			min: boundary === "min" && !repeatedBoundary,
			max: boundary === "max" && !repeatedBoundary
		});
		if (boundary) if (repeatedBoundary) this.#restartBoundary(boundary);
		else this.#scheduleBoundaryClear();
		this.#close.arm();
		return true;
	}
	#scheduleBoundaryClear() {
		this.#clearBoundaryTimer();
		this.#boundaryTimer = setTimeout(() => {
			this.#boundaryTimer = null;
			this.state.patch({
				min: false,
				max: false
			});
		}, BOUNDARY_CLEAR_DELAY);
	}
	#restartBoundary(boundary) {
		this.#clearBoundaryTimers();
		this.state.patch({
			min: false,
			max: false
		});
		this.#boundaryRestartTimer = setTimeout(() => {
			this.#boundaryRestartTimer = null;
			this.state.patch({ [boundary]: true });
			this.#scheduleBoundaryClear();
		}, 0);
	}
	#clearBoundaryTimer() {
		if (this.#boundaryTimer === null) return;
		clearTimeout(this.#boundaryTimer);
		this.#boundaryTimer = null;
	}
	#clearBoundaryRestartTimer() {
		if (this.#boundaryRestartTimer === null) return;
		clearTimeout(this.#boundaryRestartTimer);
		this.#boundaryRestartTimer = null;
	}
	#clearBoundaryTimers() {
		this.#clearBoundaryTimer();
		this.#clearBoundaryRestartTimer();
	}
};
function getVolumeBoundary(event, currentVolume, nextVolume) {
	if (event.action !== "volumeStep" || event.value === void 0 || event.value === 0) return null;
	if (nextVolume !== currentVolume) return null;
	return event.value < 0 ? "min" : "max";
}
//#endregion
export { VolumeIndicatorCore };
