// @videojs/core/dist/dev/dom/media/google-cast/index.js
import { getDefaultCastOptions, loadCastFramework, requiresCastFramework } from "./utils.js";
import { GoogleCastProvider } from "./google-cast-provider.js";
import { RemotePlayback } from "./remote-playback.js";
//#region src/dom/media/google-cast/index.ts
const GoogleCastMixin = (superclass) => {
	class GoogleCastMedia extends superclass {
		#castOptions = getDefaultCastOptions();
		#castCustomData;
		#castSrc;
		#castContentType;
		#castStreamType;
		#castReceiver;
		#remote;
		#provider;
		#destroyed = false;
		get remote() {
			if (this.#remote) return this.#remote;
			if (this.#destroyed) return super.remote;
			if (requiresCastFramework()) {
				if (!this.disableRemotePlayback) loadCastFramework();
				this.#provider = new GoogleCastProvider(this, {
					duration: () => super.duration,
					currentTime: () => super.currentTime,
					paused: () => super.paused,
					muted: () => super.muted,
					pause: () => super.pause()
				});
				return this.#remote = new RemotePlayback(this.#provider);
			}
			return super.remote;
		}
		attach(target) {
			super.attach(target);
			if (requiresCastFramework() && !this.disableRemotePlayback) loadCastFramework();
		}
		detach() {
			super.detach();
		}
		destroy() {
			this.#provider?.destroy();
			this.#provider = null;
			this.#remote = null;
			this.#destroyed = true;
			super.destroy();
		}
		async load() {
			if (!this.#provider?.isCasting) return super.load();
			return this.#provider.load();
		}
		play() {
			if (this.#provider?.isCasting) return this.#provider.play();
			return super.play();
		}
		pause() {
			if (this.#provider?.isCasting) {
				this.#provider.pause();
				return;
			}
			super.pause();
		}
		/** Resolved options passed to the Google Cast framework when it initializes. */
		get castOptions() {
			return this.#castOptions;
		}
		/** Cast receiver application ID. Defaults to Google's default media receiver. */
		get castReceiver() {
			return this.#castReceiver;
		}
		set castReceiver(val) {
			const next = val || void 0;
			if (this.#castReceiver === next) return;
			this.#castReceiver = next;
			if (next) this.#castOptions.receiverApplicationId = next;
		}
		/** Source URL loaded on the Cast receiver. Falls back to a `<source>` child, `src`, then `currentSrc`. */
		get castSrc() {
			return this.#castSrc ?? this.querySelector("source")?.src ?? this.src ?? this.currentSrc;
		}
		set castSrc(val) {
			const next = val || void 0;
			if (this.#castSrc === next) return;
			this.#castSrc = next;
			if (this.#provider?.isCasting) this.load();
		}
		/** MIME type of the Cast source. When unset, the receiver infers it from the URL. */
		get castContentType() {
			return this.#castContentType;
		}
		set castContentType(val) {
			this.#castContentType = val || void 0;
		}
		/** Stream type (`'on-demand'` or `'live'`) used on the Cast receiver. Falls back to `streamType`. */
		get castStreamType() {
			return this.#castStreamType ?? this.streamType;
		}
		set castStreamType(val) {
			if (this.#castStreamType === val) return;
			this.#castStreamType = val;
			if (this.#provider?.isCasting) this.load();
		}
		/** Custom data sent to the Cast receiver with the load request. */
		get castCustomData() {
			return this.#castCustomData;
		}
		set castCustomData(val) {
			const valType = typeof val;
			if (!["object", "undefined"].includes(valType)) {
				console.error(`castCustomData must be nullish or an object but value was of type ${valType}`);
				return;
			}
			this.#castCustomData = val;
		}
		get seeking() {
			if (this.#provider?.isCasting) return this.#provider.seeking;
			return super.seeking;
		}
		get readyState() {
			if (this.#provider?.isCasting) return this.#provider.readyState;
			return super.readyState;
		}
		get paused() {
			if (this.#provider?.isCasting) return this.#provider.paused;
			return super.paused;
		}
		get ended() {
			if (this.#provider?.isCasting) return this.#provider.ended;
			return super.ended;
		}
		get muted() {
			if (this.#provider?.isCasting) return this.#provider.muted;
			return super.muted;
		}
		set muted(val) {
			if (this.#provider?.isCasting) {
				this.#provider.muted = val;
				return;
			}
			super.muted = val;
		}
		get volume() {
			if (this.#provider?.isCasting) return this.#provider.volume;
			return super.volume;
		}
		set volume(val) {
			if (this.#provider?.isCasting) {
				this.#provider.volume = val;
				return;
			}
			super.volume = val;
		}
		get playbackRate() {
			if (this.#provider?.isCasting) return this.#provider.playbackRate;
			return super.playbackRate;
		}
		set playbackRate(val) {
			if (this.#provider?.isCasting) {
				this.#provider.playbackRate = val;
				return;
			}
			super.playbackRate = val;
		}
		get duration() {
			if (this.#provider?.isCasting) return this.#provider.duration;
			return super.duration;
		}
		get currentTime() {
			if (this.#provider?.isCasting) return this.#provider.currentTime;
			return super.currentTime;
		}
		set currentTime(val) {
			if (this.#provider?.isCasting) {
				this.#provider.currentTime = val;
				return;
			}
			super.currentTime = val;
		}
	}
	return GoogleCastMedia;
};
//#endregion
export { GoogleCastMixin };
