// @videojs/core/dist/default/dom/media/native-hls/errors.js
import { MediaError } from "../../../core/media/media-error.js";
//#region src/dom/media/native-hls/errors.ts
function NativeHlsMediaErrorsMixin(BaseClass) {
	class NativeHlsMediaErrors extends BaseClass {
		#disconnect = null;
		#error = null;
		get error() {
			return this.#error;
		}
		attach(target) {
			super.attach?.(target);
			this.#init(target);
		}
		detach() {
			this.#destroy();
			super.detach?.();
		}
		destroy() {
			this.#destroy();
			super.destroy?.();
		}
		#destroy() {
			this.#disconnect?.abort();
			this.#disconnect = null;
			this.#error = null;
		}
		#init(target) {
			this.#destroy();
			this.#disconnect = new AbortController();
			const signal = this.#disconnect.signal;
			target.addEventListener("error", (event) => {
				event.stopImmediatePropagation();
				const native = target.error;
				if (!native) return;
				const error = new MediaError(native.message, native.code, true);
				this.#error = error;
				this.dispatchEvent(new ErrorEvent("error", {
					error,
					message: error.message
				}));
			}, {
				signal,
				capture: true
			});
			target.addEventListener("emptied", () => {
				this.#error = null;
			}, { signal });
		}
	}
	return NativeHlsMediaErrors;
}
//#endregion
export { NativeHlsMediaErrorsMixin };
