// @videojs/html/dist/default/define/ui/compounds.js
import { AlertDialogCloseElement } from "../../ui/alert-dialog/alert-dialog-close-element.js";
import { AlertDialogDescriptionElement } from "../../ui/alert-dialog/alert-dialog-description-element.js";
import { AlertDialogTitleElement } from "../../ui/alert-dialog/alert-dialog-title-element.js";
import { ControlsElement } from "../../ui/controls/controls-element.js";
import { ControlsGroupElement } from "../../ui/controls/controls-group-element.js";
import { ErrorDialogElement } from "../../ui/error-dialog/error-dialog-element.js";
import { MenuBackElement } from "../../ui/menu/menu-back-element.js";
import { MenuCheckboxItemElement } from "../../ui/menu/menu-checkbox-item-element.js";
import { MenuElement } from "../../ui/menu/menu-element.js";
import { MenuGroupElement } from "../../ui/menu/menu-group-element.js";
import { MenuItemElement } from "../../ui/menu/menu-item-element.js";
import { MenuItemIndicatorElement } from "../../ui/menu/menu-item-indicator-element.js";
import { MenuLabelElement } from "../../ui/menu/menu-label-element.js";
import { MenuRadioGroupElement } from "../../ui/menu/menu-radio-group-element.js";
import { MenuRadioItemElement } from "../../ui/menu/menu-radio-item-element.js";
import { MenuSeparatorElement } from "../../ui/menu/menu-separator-element.js";
import { MenuViewElement } from "../../ui/menu/menu-view-element.js";
import { SeekIndicatorElement } from "../../ui/seek-indicator/seek-indicator-element.js";
import { SeekIndicatorValueElement } from "../../ui/seek-indicator/seek-indicator-value-element.js";
import { SliderBufferElement } from "../../ui/slider/slider-buffer-element.js";
import { SliderElement } from "../../ui/slider/slider-element.js";
import { SliderFillElement } from "../../ui/slider/slider-fill-element.js";
import { SliderPreviewElement } from "../../ui/slider/slider-preview-element.js";
import { SliderThumbElement } from "../../ui/slider/slider-thumb-element.js";
import { SliderThumbnailElement } from "../../ui/slider/slider-thumbnail-element.js";
import { SliderTrackElement } from "../../ui/slider/slider-track-element.js";
import { SliderValueElement } from "../../ui/slider/slider-value-element.js";
import { StatusAnnouncerElement } from "../../ui/status-announcer/status-announcer-element.js";
import { StatusIndicatorElement } from "../../ui/status-indicator/status-indicator-element.js";
import { StatusIndicatorValueElement } from "../../ui/status-indicator/status-indicator-value-element.js";
import { TimeElement } from "../../ui/time/time-element.js";
import { TimeGroupElement } from "../../ui/time/time-group-element.js";
import { TimeSeparatorElement } from "../../ui/time/time-separator-element.js";
import { TimeSliderElement } from "../../ui/time-slider/time-slider-element.js";
import { VolumeIndicatorElement } from "../../ui/volume-indicator/volume-indicator-element.js";
import { VolumeIndicatorFillElement } from "../../ui/volume-indicator/volume-indicator-fill-element.js";
import { VolumeIndicatorValueElement } from "../../ui/volume-indicator/volume-indicator-value-element.js";
import { VolumeSliderElement } from "../../ui/volume-slider/volume-slider-element.js";
import { safeDefine } from "../safe-define.js";
//#region src/define/ui/compounds.ts
function defineMenu() {
	safeDefine(MenuElement);
	safeDefine(MenuBackElement);
	safeDefine(MenuItemElement);
	safeDefine(MenuLabelElement);
	safeDefine(MenuSeparatorElement);
	safeDefine(MenuGroupElement);
	safeDefine(MenuRadioGroupElement);
	safeDefine(MenuRadioItemElement);
	safeDefine(MenuCheckboxItemElement);
	safeDefine(MenuItemIndicatorElement);
	safeDefine(MenuViewElement);
}
function defineControls() {
	safeDefine(ControlsElement);
	safeDefine(ControlsGroupElement);
}
function defineErrorDialog() {
	safeDefine(ErrorDialogElement);
	safeDefine(AlertDialogCloseElement);
	safeDefine(AlertDialogDescriptionElement);
	safeDefine(AlertDialogTitleElement);
}
function defineInputIndicators() {
	safeDefine(StatusAnnouncerElement);
	safeDefine(StatusIndicatorElement);
	safeDefine(StatusIndicatorValueElement);
	safeDefine(VolumeIndicatorElement);
	safeDefine(VolumeIndicatorFillElement);
	safeDefine(VolumeIndicatorValueElement);
	safeDefine(SeekIndicatorElement);
	safeDefine(SeekIndicatorValueElement);
}
/** Shared slider sub-elements used by all slider types. */
function defineSliderParts() {
	safeDefine(SliderFillElement);
	safeDefine(SliderPreviewElement);
	safeDefine(SliderThumbElement);
	safeDefine(SliderTrackElement);
	safeDefine(SliderValueElement);
}
function defineSlider() {
	safeDefine(SliderElement);
	defineSliderParts();
}
function defineTime() {
	safeDefine(TimeElement);
	safeDefine(TimeGroupElement);
	safeDefine(TimeSeparatorElement);
}
function defineTimeSlider() {
	safeDefine(TimeSliderElement);
	defineSliderParts();
	safeDefine(SliderBufferElement);
	safeDefine(SliderThumbnailElement);
}
function defineVolumeSlider() {
	safeDefine(VolumeSliderElement);
	defineSliderParts();
}
//#endregion
export { defineControls, defineErrorDialog, defineInputIndicators, defineMenu, defineSlider, defineSliderParts, defineTime, defineTimeSlider, defineVolumeSlider };
