// @videojs/core/dist/dev/dom/media/native-hls/stream-type.js
import { MediaStreamTypes } from "../../../core/media/types.js";
//#region src/dom/media/native-hls/stream-type.ts
function NativeHlsMediaStreamTypeMixin(BaseClass) {
	class NativeHlsMediaStreamType extends BaseClass {
		#streamType = MediaStreamTypes.UNKNOWN;
		#isUserStreamType = false;
		#disconnect = null;
		get streamType() {
			return this.#streamType;
		}
		set streamType(value) {
			if (value === MediaStreamTypes.UNKNOWN) {
				this.#isUserStreamType = false;
				this.#setDetected(this.#detect());
				return;
			}
			this.#isUserStreamType = true;
			this.#update(value);
		}
		attach(target) {
			super.attach?.(target);
			this.#init(target);
		}
		detach() {
			this.#destroy();
			this.#setDetected(MediaStreamTypes.UNKNOWN);
			super.detach?.();
		}
		destroy() {
			this.#destroy();
			super.destroy?.();
		}
		#destroy() {
			this.#disconnect?.abort();
			this.#disconnect = null;
		}
		#init(target) {
			this.#destroy();
			this.#disconnect = new AbortController();
			const { signal } = this.#disconnect;
			const detect = () => this.#setDetected(this.#detect(target));
			target.addEventListener("durationchange", detect, { signal });
			target.addEventListener("loadedmetadata", detect, { signal });
			target.addEventListener("emptied", () => this.#setDetected(MediaStreamTypes.UNKNOWN), { signal });
			detect();
		}
		#detect(target = this.target) {
			if (!target) return MediaStreamTypes.UNKNOWN;
			const { duration } = target;
			if (duration === Infinity) return MediaStreamTypes.LIVE;
			if (Number.isFinite(duration) && duration > 0) return MediaStreamTypes.ON_DEMAND;
			return MediaStreamTypes.UNKNOWN;
		}
		#setDetected(value) {
			if (this.#isUserStreamType) return;
			this.#update(value);
		}
		#update(value) {
			if (this.#streamType === value) return;
			this.#streamType = value;
			this.dispatchEvent(new Event("streamtypechange"));
		}
	}
	return NativeHlsMediaStreamType;
}
//#endregion
export { NativeHlsMediaStreamTypeMixin };
