// vidstack/prod/vidstack.js
import { DOMEvent, EventsController, appendTriggerEvent, findTriggerEvent, hasTriggerEvent, isKeyboardClick, isKeyboardEvent, isPointerEvent, isString, useState, walkTriggerEventChain } from "./chunks/vidstack-D9mvYVIz.js";
import { AudioProviderLoader, AudioTrackList, DASHProviderLoader, FullscreenController, HLSProviderLoader, List, LocalMediaStorage, Logger, MEDIA_KEY_SHORTCUTS, MediaControls, MediaPlayer, MediaProvider, MediaRemoteControl, ScreenOrientationController, TextRenderers, TextTrackList, VideoProviderLoader, VideoQualityList, VimeoProviderLoader, YouTubeProviderLoader, boundTime, canFullscreen, isAudioProvider, isDASHProvider, isGoogleCastProvider, isHLSProvider, isHTMLAudioElement, isHTMLIFrameElement, isHTMLMediaElement, isHTMLVideoElement, isVideoProvider, isVideoQualitySrc, isVimeoProvider, isYouTubeProvider, mediaState, softResetMediaState } from "./chunks/vidstack-Cf87pnA0.js";
import { AUDIO_EXTENSIONS, AUDIO_TYPES, DASH_VIDEO_EXTENSIONS, DASH_VIDEO_TYPES, HLS_VIDEO_EXTENSIONS, HLS_VIDEO_TYPES, VIDEO_EXTENSIONS, VIDEO_TYPES, canChangeVolume, canGoogleCastSrc, canOrientScreen, canPlayHLSNatively, canRotateScreen, canUsePictureInPicture, canUseVideoPresentation, isAudioSrc, isDASHSrc, isHLSSrc, isMediaStream, isVideoSrc } from "./chunks/vidstack-Dtnbh_MF.js";
import { TimeRange, getTimeRangesEnd, getTimeRangesStart, normalizeTimeIntervals, updateTimeIntervals } from "./chunks/vidstack-CdfyZegh.js";
import { getDownloadFile } from "./chunks/vidstack-mvRkf9tF.js";
import { TextTrack, TextTrackSymbol, isTrackCaptionKind, parseJSONCaptionsFile } from "./chunks/vidstack-DYPyTdYw.js";
import { findActiveCue, isCueActive, watchActiveTextTrack, watchCueTextChange } from "./chunks/vidstack-B1Z6f96t.js";
import { mediaContext } from "./chunks/vidstack-CYE89tBD.js";
import { ARIAKeyShortcuts, AirPlayButton, AudioRadioGroup, CaptionButton, CaptionsRadioGroup, DEFAULT_PLAYBACK_RATES, FullscreenButton, LiveButton, Menu, MenuButton, MenuItem, MenuItems, MenuPortal, MuteButton, PIPButton, PlayButton, QualityRadioGroup, Radio, SeekButton, Slider, SliderController, SliderPreview, SliderValue, SpeedRadioGroup, Thumbnail, ThumbnailsLoader, Time, TimeSlider, VolumeSlider, formatSpokenTime, formatTime, menuPortalContext, sliderContext, sliderState, sortVideoQualities, updateSliderPreviewPlacement } from "./chunks/vidstack-D0Mmfcu7.js";
import { AudioGainRadioGroup, AudioGainSlider, Captions, ChaptersRadioGroup, Controls, ControlsGroup, DEFAULT_AUDIO_GAINS, Gesture, GoogleCastButton, MediaAnnouncer, QualitySlider, RadioGroup, SliderChapters, SliderVideo, SpeedSlider, ToggleButton, Tooltip, TooltipContent, TooltipTrigger } from "./chunks/vidstack-CCGISNeA.js";
import { Poster } from "./chunks/vidstack-KSHlFG6B.js";
import { usePlyrLayoutClasses } from "./chunks/vidstack-DMGsupdO.js";
//#region src/core/tracks/text/render/libass-text-renderer.ts
var LibASSTextRenderer = class {
	loader;
	config;
	priority = 1;
	#instance = null;
	#track = null;
	#typeRE = /(ssa|ass)$/;
	constructor(loader, config) {
		this.loader = loader;
		this.config = config;
	}
	canRender(track, video) {
		return !!video && !!track.src && (isString(track.type) && this.#typeRE.test(track.type) || this.#typeRE.test(track.src));
	}
	attach(video) {
		if (!video) return;
		this.loader().then(async (mod) => {
			this.#instance = new mod.default({
				...this.config,
				video,
				subUrl: this.#track?.src || ""
			});
			new EventsController(this.#instance).add("ready", () => {
				const canvas = this.#instance?._canvas;
				if (canvas) canvas.style.pointerEvents = "none";
			}).add("error", (event) => {
				if (!this.#track) return;
				this.#track[TextTrackSymbol.readyState] = 3;
				this.#track.dispatchEvent(new DOMEvent("error", {
					trigger: event,
					detail: event.error
				}));
			});
		});
	}
	changeTrack(track) {
		if (!track || track.readyState === 3) this.#freeTrack();
		else if (this.#track !== track) {
			this.#instance?.setTrackByUrl(track.src);
			this.#track = track;
		}
	}
	detach() {
		this.#freeTrack();
	}
	#freeTrack() {
		this.#instance?.freeTrack();
		this.#track = null;
	}
};
//#endregion
//#region src/components/ui/sliders/slider-thumbnail.ts
/**
* Used to display preview thumbnails when the user is hovering or dragging the time slider.
* The time ranges in the WebVTT file will automatically be matched based on the current slider
* pointer position.
*
* @attr data-loading - Whether thumbnail image is loading.
* @attr data-error - Whether an error occurred loading thumbnail.
* @attr data-hidden - Whether thumbnail is not available or failed to load.
* @docs {@link https://www.vidstack.io/docs/player/components/sliders/slider-thumbnail}
*/
var SliderThumbnail = class extends Thumbnail {
	#slider;
	onAttach(el) {
		this.#slider = useState(Slider.state);
	}
	getTime() {
		const { duration, clipStartTime } = this.media.$state;
		return clipStartTime() + this.#slider.pointerRate() * duration();
	}
};
//#endregion
export { ARIAKeyShortcuts, AUDIO_EXTENSIONS, AUDIO_TYPES, AirPlayButton, AudioGainRadioGroup, AudioGainSlider, AudioProviderLoader, AudioRadioGroup, AudioTrackList, CaptionButton, Captions, CaptionsRadioGroup, ChaptersRadioGroup, Controls, ControlsGroup, DASHProviderLoader, DASH_VIDEO_EXTENSIONS, DASH_VIDEO_TYPES, DEFAULT_AUDIO_GAINS, DEFAULT_PLAYBACK_RATES, FullscreenButton, FullscreenController, Gesture, GoogleCastButton, HLSProviderLoader, HLS_VIDEO_EXTENSIONS, HLS_VIDEO_TYPES, LibASSTextRenderer, List, LiveButton, LocalMediaStorage, Logger, MEDIA_KEY_SHORTCUTS, MediaAnnouncer, MediaControls, MediaPlayer, MediaProvider, MediaRemoteControl, Menu, MenuButton, MenuItem, MenuItems, MenuPortal, MuteButton, PIPButton, PlayButton, Poster, QualityRadioGroup, QualitySlider, Radio, RadioGroup, ScreenOrientationController, SeekButton, Slider, SliderChapters, SliderController, SliderPreview, SliderThumbnail, SliderValue, SliderVideo, SpeedRadioGroup, SpeedSlider, TextRenderers, TextTrack, TextTrackList, Thumbnail, ThumbnailsLoader, Time, TimeRange, TimeSlider, ToggleButton, Tooltip, TooltipContent, TooltipTrigger, VIDEO_EXTENSIONS, VIDEO_TYPES, VideoProviderLoader, VideoQualityList, VimeoProviderLoader, VolumeSlider, YouTubeProviderLoader, appendTriggerEvent, boundTime, canChangeVolume, canFullscreen, canGoogleCastSrc, canOrientScreen, canPlayHLSNatively, canRotateScreen, canUsePictureInPicture, canUseVideoPresentation, findActiveCue, findTriggerEvent, formatSpokenTime, formatTime, getDownloadFile, getTimeRangesEnd, getTimeRangesStart, hasTriggerEvent, isAudioProvider, isAudioSrc, isCueActive, isDASHProvider, isDASHSrc, isGoogleCastProvider, isHLSProvider, isHLSSrc, isHTMLAudioElement, isHTMLIFrameElement, isHTMLMediaElement, isHTMLVideoElement, isKeyboardClick, isKeyboardEvent, isMediaStream, isPointerEvent, isTrackCaptionKind, isVideoProvider, isVideoQualitySrc, isVideoSrc, isVimeoProvider, isYouTubeProvider, mediaContext, mediaState, menuPortalContext, normalizeTimeIntervals, parseJSONCaptionsFile, sliderContext, sliderState, softResetMediaState, sortVideoQualities, updateSliderPreviewPlacement, updateTimeIntervals, usePlyrLayoutClasses, walkTriggerEventChain, watchActiveTextTrack, watchCueTextChange };
