// @videojs/core/dist/dev/dom/store/features/controls.js
import { definePlayerFeature } from "../../feature.js";
import { isMediaPauseCapable, isMediaRemotePlaybackCapable } from "../../media/predicate.js";
import { findGestureCoordinator } from "../../gesture/coordinator.js";
import { isRemotePlaybackConnected, isRemotePlaybackConnecting } from "../../presentation/remote-playback.js";
import { isNull } from "@videojs/utils/predicate";
import { listen } from "@videojs/utils/dom";
//#region src/dom/store/features/controls.ts
const IDLE_DELAY = 2e3;
const TAP_THRESHOLD = 250;
const controlsFeature = definePlayerFeature({
	name: "controls",
	state: ({ get, set }) => ({
		userActive: true,
		controlsVisible: true,
		toggleControls() {
			const next = !get().userActive;
			set({
				userActive: next,
				controlsVisible: next
			});
			return next;
		}
	}),
	attach({ target, signal, get, set }) {
		const { media, container } = target;
		if (!isMediaPauseCapable(media) || isNull(container)) {
			if (isNull(container)) console.warn("[vjs] controlsFeature requires a container element for activity tracking.");
			return;
		}
		const computeVisible = (userActive) => {
			return userActive || media.paused || isRemotePlaybackConnected(media) || isRemotePlaybackConnecting(media);
		};
		let idleTimer;
		function clearIdle() {
			clearTimeout(idleTimer);
			idleTimer = void 0;
		}
		function scheduleIdle() {
			clearIdle();
			idleTimer = setTimeout(setInactive, IDLE_DELAY);
		}
		function setActive() {
			if (!get().userActive) set({
				userActive: true,
				controlsVisible: true
			});
			scheduleIdle();
		}
		function setInactive() {
			clearIdle();
			set({
				userActive: false,
				controlsVisible: computeVisible(false)
			});
		}
		set({ toggleControls() {
			if (get().controlsVisible) setInactive();
			else setActive();
			return get().controlsVisible;
		} });
		let pointerDownTime = 0;
		function onPointerDown() {
			pointerDownTime = Date.now();
		}
		function onPointerUp(event) {
			if (event.pointerType === "touch" && Date.now() - pointerDownTime < TAP_THRESHOLD) {
				if (findGestureCoordinator(container)?.bindings.some((b) => b.type === "tap" && b.action === "toggleControls" && (!b.pointer || b.pointer === "touch"))) return;
				const isMediaOrContainer = [media, container].includes(event.target);
				if (get().controlsVisible && isMediaOrContainer) setInactive();
				else setActive();
			} else setActive();
		}
		const onPlaybackChange = () => {
			const { userActive } = get();
			set({ controlsVisible: computeVisible(userActive) });
			if (!media.paused && userActive) scheduleIdle();
		};
		listen(container, "pointermove", setActive, { signal });
		listen(container, "pointerdown", onPointerDown, { signal });
		listen(container, "pointerup", onPointerUp, { signal });
		listen(container, "keyup", setActive, { signal });
		listen(container, "focusin", setActive, { signal });
		listen(container, "mouseleave", setInactive, { signal });
		listen(media, "play", onPlaybackChange, { signal });
		listen(media, "pause", onPlaybackChange, { signal });
		listen(media, "ended", onPlaybackChange, { signal });
		if (isMediaRemotePlaybackCapable(media)) {
			const onCastChange = () => {
				const { userActive } = get();
				set({ controlsVisible: computeVisible(userActive) });
			};
			listen(media.remote, "connect", onCastChange, { signal });
			listen(media.remote, "connecting", onCastChange, { signal });
			listen(media.remote, "disconnect", onCastChange, { signal });
		}
		signal.addEventListener("abort", clearIdle, { once: true });
		scheduleIdle();
	}
});
//#endregion
export { controlsFeature };
