// @videojs/core/dist/dev/dom/store/features/remote-playback.js
import { definePlayerFeature } from "../../feature.js";
import { isMediaRemotePlaybackCapable } from "../../media/predicate.js";
import { isRemotePlaybackConnected, requestRemotePlayback } from "../../presentation/remote-playback.js";
import { exitFullscreen, isFullscreen } from "../../presentation/fullscreen.js";
import { listen } from "@videojs/utils/dom";
//#region src/dom/store/features/remote-playback.ts
const remotePlaybackFeature = definePlayerFeature({
	name: "remotePlayback",
	state: ({ target }) => ({
		remotePlaybackState: "disconnected",
		remotePlaybackAvailability: "unsupported",
		async toggleRemotePlayback() {
			const { media, container } = target();
			if (isRemotePlaybackConnected(media)) return requestRemotePlayback(media);
			if (isFullscreen(container, media)) await exitFullscreen(media);
			return requestRemotePlayback(media);
		}
	}),
	attach({ target, signal, set }) {
		const { media } = target;
		if (!isMediaRemotePlaybackCapable(media)) return;
		const syncState = () => set({ remotePlaybackState: media.remote.state });
		syncState();
		listen(media.remote, "connect", syncState, { signal });
		listen(media.remote, "connecting", syncState, { signal });
		listen(media.remote, "disconnect", syncState, { signal });
		media.remote.watchAvailability((available) => {
			set({ remotePlaybackAvailability: available ? "available" : "unavailable" });
		}).catch(() => {
			set({ remotePlaybackAvailability: "unsupported" });
		});
		signal.addEventListener("abort", () => {
			media.remote?.cancelWatchAvailability().catch(() => {});
		});
	}
});
//#endregion
export { remotePlaybackFeature };
