// vidstack/server/vidstack-elements.js
//#region src/elements/index.ts
function defineCustomElement() {}
var MediaCaptionsElement = class {};
var MediaGestureElement = class {};
var MediaProviderElement = class {};
var MediaPlayerElement = class {};
var MediaAnnouncerElement = class {};
var MediaPosterElement = class {};
var MediaThumbnailElement = class {};
var MediaTimeElement = class {};
var MediaControlsElement = class {};
var MediaControlsGroupElement = class {};
var MediaTitleElement = class {};
var MediaChapterTitleElement = class {};
var MediaSpinnerElement = class {};
var MediaLayoutElement = class {};
var MediaAudioLayoutElement = class {};
var MediaVideoLayoutElement = class {};
var MediaPlyrLayoutElement = class {};
var MediaAirPlayButtonElement = class {};
var MediaCaptionButtonElement = class {};
var MediaFullscreenButtonElement = class {};
var MediaGoogleCastButtonElement = class {};
var MediaLiveButtonElement = class {};
var MediaMuteButtonElement = class {};
var MediaPIPButtonElement = class {};
var MediaPlayButtonElement = class {};
var MediaSeekButtonElement = class {};
var MediaToggleButtonElement = class {};
var MediaTooltipElement = class {};
var MediaTooltipTriggerElement = class {};
var MediaTooltipContentElement = class {};
var MediaAudioRadioGroupElement = class {};
var MediaCaptionsRadioGroupElement = class {};
var MediaMenuElement = class {};
var MediaMenuButtonElement = class {};
var MediaMenuPortalElement = class {};
var MediaMenuItemElement = class {};
var MediaMenuItemsElement = class {};
var MediaChaptersRadioGroupElement = class {};
var MediaSpeedRadioGroupElement = class {};
var MediaAudioGainRadioGroupElement = class {};
var MediaQualityRadioGroupElement = class {};
var MediaRadioElement = class {};
var MediaRadioGroupElement = class {};
var MediaSliderElement = class {};
var MediaSliderThumbnailElement = class {};
var MediaSliderValueElement = class {};
var MediaSliderVideoElement = class {};
var MediaTimeSliderElement = class {};
var MediaSliderPreviewElement = class {};
var MediaVolumeSliderElement = class {};
var MediaAudioGainSliderElement = class {};
var MediaSpeedSliderElement = class {};
var MediaQualitySliderElement = class {};
var MediaSliderChaptersElement = class {};
var MediaSliderStepsElement = class {};
//#endregion
export { MediaAirPlayButtonElement, MediaAnnouncerElement, MediaAudioGainRadioGroupElement, MediaAudioGainSliderElement, MediaAudioLayoutElement, MediaAudioRadioGroupElement, MediaCaptionButtonElement, MediaCaptionsElement, MediaCaptionsRadioGroupElement, MediaChapterTitleElement, MediaChaptersRadioGroupElement, MediaControlsElement, MediaControlsGroupElement, MediaFullscreenButtonElement, MediaGestureElement, MediaGoogleCastButtonElement, MediaLayoutElement, MediaLiveButtonElement, MediaMenuButtonElement, MediaMenuElement, MediaMenuItemElement, MediaMenuItemsElement, MediaMenuPortalElement, MediaMuteButtonElement, MediaPIPButtonElement, MediaPlayButtonElement, MediaPlayerElement, MediaPlyrLayoutElement, MediaPosterElement, MediaProviderElement, MediaQualityRadioGroupElement, MediaQualitySliderElement, MediaRadioElement, MediaRadioGroupElement, MediaSeekButtonElement, MediaSliderChaptersElement, MediaSliderElement, MediaSliderPreviewElement, MediaSliderStepsElement, MediaSliderThumbnailElement, MediaSliderValueElement, MediaSliderVideoElement, MediaSpeedRadioGroupElement, MediaSpeedSliderElement, MediaSpinnerElement, MediaThumbnailElement, MediaTimeElement, MediaTimeSliderElement, MediaTitleElement, MediaToggleButtonElement, MediaTooltipContentElement, MediaTooltipElement, MediaTooltipTriggerElement, MediaVideoLayoutElement, MediaVolumeSliderElement, defineCustomElement };
